<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Complete end-to-end LC-MS/MS Metabolomic Data analysis • metabonaut</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><script src="a-end-to-end-untargeted-metabolomics_files/libs/quarto-html/popper.min.js"></script><script src="a-end-to-end-untargeted-metabolomics_files/libs/quarto-html/tippy.umd.min.js"></script><link href="a-end-to-end-untargeted-metabolomics_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="a-end-to-end-untargeted-metabolomics_files/libs/quarto-html/light-border.css" rel="stylesheet">
<!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Complete end-to-end LC-MS/MS Metabolomic Data analysis">
<meta property="og:image" content="https://rformassspectrometry.github.io/metabonaut/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">metabonaut</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.6</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/install_v0.html">Installation instruction</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-workflows" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Workflows</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-workflows">
<li><a class="dropdown-item" href="../articles/dataset-investigation.html">Dataset investigation: What to do when you get your data</a></li>
    <li><a class="dropdown-item" href="../articles/a-end-to-end-untargeted-metabolomics.html">End-to-end workflow for untargeted metabolomics data analysis in R</a></li>
    <li><a class="dropdown-item" href="../articles/alignment-to-external-dataset.html">Seamless Alignment: Merging New data with Existing Preprocessed Datasets</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/rformassspectrometry/metabonaut/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-quarto">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1></h1>


      <p class="date">2024-12-04</p>

      <small class="dont-index">Source: <a href="https://github.com/rformassspectrometry/metabonaut/blob/main/vignettes/a-end-to-end-untargeted-metabolomics.qmd" class="external-link"><code>vignettes/a-end-to-end-untargeted-metabolomics.qmd</code></a></small>
      <div class="d-none name"><code></code></div>
    </div>






    <header id="title-block-header" class="quarto-title-block"><h1 class="title">Complete end-to-end LC-MS/MS Metabolomic Data analysis</h1></header><div class="cell">
<style type="text/css">
.verysmall .table{
  font-size: 8px;
}
</style>
</div>
<section class="section level2"><h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The present workflow describes all steps for the analysis of an LC-MS/MS experiment, which includes the preprocessing of the raw data to generate the <em>abundance</em> matrix for the <em>features</em> in the various samples, followed by data normalization, differential abundance analysis and finally the annotation of features to metabolites. Note that also alternative analysis options and R packages could be used for different steps and some examples are mentioned throughout the workflow.</p>
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="images/workflow.png"></p>
<figcaption>Steps of the end-to-end workflow and possible alternatives</figcaption></figure>
</div>
</section><section class="section level2"><h2 id="data-description">Data description<a class="anchor" aria-label="anchor" href="#data-description"></a>
</h2>
<p>See the <a href="https://rformassspectrometry.github.io/metabonaut/articles/dataset-investigation.html">data description</a> vignette for detailed explanation of the dataset we go through in this workflow and general tips on what should be done when you first get your data.</p>
</section><section class="section level2"><h2 id="packages-needed">Packages needed<a class="anchor" aria-label="anchor" href="#packages-needed"></a>
</h2>
<p>Our workflow is therefore based on the following dependencies:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Data Import and handling</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readxl.tidyverse.org" class="external-link">readxl</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/MsExperiment" class="external-link">MsExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/MsIO" class="external-link">MsIO</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">alabaster.se</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/MsBackendMetaboLights" class="external-link">MsBackendMetaboLights</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Preprocessing of LC-MS data</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/sneumann/xcms" class="external-link">xcms</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/Spectra" class="external-link">Spectra</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/MetaboCoreUtils" class="external-link">MetaboCoreUtils</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Statistical analysis</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/limma/" class="external-link">limma</a></span><span class="op">)</span> <span class="co"># Differential abundance</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HenrikBengtsson/matrixStats" class="external-link">matrixStats</a></span><span class="op">)</span> <span class="co"># Summaries over matrices</span></span>
<span></span>
<span><span class="co">## Visualisation</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rapporter.github.io/pander/" class="external-link">pander</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/TomKellyGenetics/vioplot" class="external-link">vioplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/sinhrks/ggfortify" class="external-link">ggfortify</a></span><span class="op">)</span>   <span class="co"># Plot PCA</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">gridExtra</span><span class="op">)</span>   <span class="co"># To arrange multiple ggplots into single plots</span></span>
<span></span>
<span><span class="co">## Annotation</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">AnnotationHub</span><span class="op">)</span> <span class="co"># Annotation resources</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/CompoundDb" class="external-link">CompoundDb</a></span><span class="op">)</span>    <span class="co"># Access small compound annotation data.</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/MetaboAnnotation" class="external-link">MetaboAnnotation</a></span><span class="op">)</span> <span class="co"># Functionality for metabolite annotation.</span></span></code></pre></div>
</div>
</section><section class="section level2"><h2 id="data-import">Data import<a class="anchor" aria-label="anchor" href="#data-import"></a>
</h2>
<p>Note that different equipment will generate various file extensions, so a conversion step might be needed beforehand, though it does not apply to this dataset. The Spectra package supports a variety of ways to store and retrieve MS data, including mzML, mzXML, CDF files, simple flat files, or database systems. If necessary, several tools, such as ProteoWizard’s MSConvert, can be used to convert files to the .mzML format <span class="citation" data-cites="chambers_cross-platform_2012">(<a href="#ref-chambers_cross-platform_2012" role="doc-biblioref">Chambers et al. 2012</a>)</span>.</p>
<p>Below we will show how to extract our dataset from the MetaboLigths database and load it as an <code>MsExperiment</code> object. For more information on how to load your data from the MetaboLights database, refer to the <a href="https://rformassspectrometry.github.io/MsIO/articles/MsIO.html#loading-data-from-metabolights" class="external-link">vignette</a>. For other type of data loading, check out this xcms <a href="https://jorainer.github.io/xcmsTutorials/articles/xcms-preprocessing.html#data-import-and-exploration" class="external-link">vignette</a> A more specific vignette will be created for data import soon.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">param</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MsIO/man/MetaboLightsParam.html" class="external-link">MetaboLightsParam</a></span><span class="op">(</span>mtblsId <span class="op">=</span> <span class="st">"MTBLS8735"</span>,</span>
<span>                           assayName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"a_MTBLS8735_LC-MS_positive_"</span>,</span>
<span>                                              <span class="st">"hilic_metabolite_profiling.txt"</span><span class="op">)</span>,</span>
<span>                           filePattern <span class="op">=</span> <span class="st">".mzML"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">lcms1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MsIO/man/saveMsObject.html" class="external-link">readMsObject</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">MsExperiment</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                      <span class="va">param</span>,</span>
<span>                      keepOntology <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                      keepProtocol <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                      simplify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<p>We next configure the parallel processing setup. Most functions from the <em>xcms</em> package allow per-sample parallel processing, which can improve the performance of the analysis, especially for large data sets. <em>xcms</em> and all packages from the <em>RforMassSpectrometry</em> package ecosystem use the parallel processing setup configured through the <em><a href="https://bioconductor.org/packages/3.20/BiocParallel" class="external-link">BiocParallel</a></em> Bioconductor package. With the code below we use a <em>fork-based</em> parallel processing on unix system, and a <em>socket-based</em> parallel processing on the Windows operating system.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Set up parallel processing using 2 cores</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">.Platform</span><span class="op">$</span><span class="va">OS.type</span> <span class="op">==</span> <span class="st">"unix"</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html" class="external-link">register</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html" class="external-link">register</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
</section><section class="section level2"><h2 id="data-organisation">Data organisation<a class="anchor" aria-label="anchor" href="#data-organisation"></a>
</h2>
<p>The experimental data is now represented by a <code>MsExperiment</code> object from the <em><a href="https://bioconductor.org/packages/3.20/MsExperiment" class="external-link">MsExperiment</a></em> package. The <code>MsExperiment</code> object is a container for metadata and spectral data that provides and manages also the linkage between samples and spectra.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lcms1</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Object of class MsExperiment
 Spectra: MS1 (17210)
 Experiment data: 10 sample(s)
 Sample data links:
  - spectra: 10 sample(s) to 17210 element(s).</code></pre>
</div>
</div>
<p>Below we provide a brief overview of the data structure and content. The <code><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData()</a></code> function extracts sample information from the object. We next extract this data and use the <em>pander</em> package to render and show this information in Table 1 below. Throughout the document we use the R pipe operator (<code>|&gt;</code>) to avoid nested function calls and hence improve code readability.</p>
<p>The <code><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData()</a></code> output from <em>MetaboLights</em> is not always ideal for direct and easy access to the data. We therefore rename it and transform it in a more user-friendly way. The user can add, transform and remove any column they want using base R functionalities.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Derived_Spectral_Data_File"</span>,</span>
<span>                      <span class="st">"Characteristics[Sample type]"</span>,</span>
<span>                      <span class="st">"Factor Value[Phenotype]"</span>,</span>
<span>                      <span class="st">"Sample Name"</span>,</span>
<span>                      <span class="st">"Factor Value[Age]"</span><span class="op">)</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">kable</span><span class="op">(</span>format <span class="op">=</span> <span class="st">"pipe"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<table class="table">
<caption>Table 1. Samples from the data set.</caption>
<colgroup>
<col style="width: 25%">
<col style="width: 26%">
<col style="width: 21%">
<col style="width: 10%">
<col style="width: 16%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;">Derived_Spectral_Data_File</th>
<th style="text-align: left;">Characteristics.Sample.type.</th>
<th style="text-align: left;">Factor.Value.Phenotype.</th>
<th style="text-align: left;">Sample.Name</th>
<th style="text-align: right;">Factor.Value.Age.</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">FILES/MS_QC_POOL_1_POS.mzML</td>
<td style="text-align: left;">pool</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">POOL</td>
<td style="text-align: right;">NA</td>
</tr>
<tr class="even">
<td style="text-align: left;">FILES/MS_A_POS.mzML</td>
<td style="text-align: left;">experimental sample</td>
<td style="text-align: left;">CVD</td>
<td style="text-align: left;">A</td>
<td style="text-align: right;">53</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FILES/MS_B_POS.mzML</td>
<td style="text-align: left;">experimental sample</td>
<td style="text-align: left;">CTR</td>
<td style="text-align: left;">B</td>
<td style="text-align: right;">30</td>
</tr>
<tr class="even">
<td style="text-align: left;">FILES/MS_QC_POOL_2_POS.mzML</td>
<td style="text-align: left;">pool</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">POOL</td>
<td style="text-align: right;">NA</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FILES/MS_C_POS.mzML</td>
<td style="text-align: left;">experimental sample</td>
<td style="text-align: left;">CTR</td>
<td style="text-align: left;">C</td>
<td style="text-align: right;">66</td>
</tr>
<tr class="even">
<td style="text-align: left;">FILES/MS_D_POS.mzML</td>
<td style="text-align: left;">experimental sample</td>
<td style="text-align: left;">CVD</td>
<td style="text-align: left;">D</td>
<td style="text-align: right;">36</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FILES/MS_QC_POOL_3_POS.mzML</td>
<td style="text-align: left;">pool</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">POOL</td>
<td style="text-align: right;">NA</td>
</tr>
<tr class="even">
<td style="text-align: left;">FILES/MS_E_POS.mzML</td>
<td style="text-align: left;">experimental sample</td>
<td style="text-align: left;">CTR</td>
<td style="text-align: left;">E</td>
<td style="text-align: right;">66</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FILES/MS_F_POS.mzML</td>
<td style="text-align: left;">experimental sample</td>
<td style="text-align: left;">CVD</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">44</td>
</tr>
<tr class="even">
<td style="text-align: left;">FILES/MS_QC_POOL_4_POS.mzML</td>
<td style="text-align: left;">pool</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">POOL</td>
<td style="text-align: right;">NA</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Let's rename the column for easier access</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample_name"</span>, <span class="st">"derived_spectra_data_file"</span>,</span>
<span>                                <span class="st">"metabolite_asssignment_file"</span>,</span>
<span>                                <span class="st">"source_name"</span>,</span>
<span>                                <span class="st">"organism"</span>,</span>
<span>                                <span class="st">"blood_sample_type"</span>,</span>
<span>                                <span class="st">"sample_type"</span>, <span class="st">"age"</span>, <span class="st">"unit"</span>, <span class="st">"phenotype"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add "QC" to the phenotype of the QC samples</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">$</span><span class="va">phenotype</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">$</span><span class="va">sample_name</span> <span class="op">==</span> <span class="st">"POOL"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"QC"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">$</span><span class="va">sample_name</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">$</span><span class="va">sample_name</span> <span class="op">==</span> <span class="st">"POOL"</span> <span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"POOL1"</span>, <span class="st">"POOL2"</span>, <span class="st">"POOL3"</span>, <span class="st">"POOL4"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#  Add injection index column</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">$</span><span class="va">injection_index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#let's look at the updated sample data</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"derived_spectra_data_file"</span>,</span>
<span>                     <span class="st">"phenotype"</span>, <span class="st">"sample_name"</span>, <span class="st">"age"</span>,</span>
<span>                     <span class="st">"injection_index"</span><span class="op">)</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">kable</span><span class="op">(</span>format <span class="op">=</span> <span class="st">"pipe"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<table style="width:100%;" class="table">
<caption>Table 2. Simplified sample data.</caption>
<colgroup>
<col style="width: 39%">
<col style="width: 14%">
<col style="width: 17%">
<col style="width: 5%">
<col style="width: 22%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;">derived_spectra_data_file</th>
<th style="text-align: left;">phenotype</th>
<th style="text-align: left;">sample_name</th>
<th style="text-align: right;">age</th>
<th style="text-align: right;">injection_index</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">FILES/MS_QC_POOL_1_POS.mzML</td>
<td style="text-align: left;">QC</td>
<td style="text-align: left;">POOL1</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">FILES/MS_A_POS.mzML</td>
<td style="text-align: left;">CVD</td>
<td style="text-align: left;">A</td>
<td style="text-align: right;">53</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FILES/MS_B_POS.mzML</td>
<td style="text-align: left;">CTR</td>
<td style="text-align: left;">B</td>
<td style="text-align: right;">30</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">FILES/MS_QC_POOL_2_POS.mzML</td>
<td style="text-align: left;">QC</td>
<td style="text-align: left;">POOL2</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FILES/MS_C_POS.mzML</td>
<td style="text-align: left;">CTR</td>
<td style="text-align: left;">C</td>
<td style="text-align: right;">66</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="even">
<td style="text-align: left;">FILES/MS_D_POS.mzML</td>
<td style="text-align: left;">CVD</td>
<td style="text-align: left;">D</td>
<td style="text-align: right;">36</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FILES/MS_QC_POOL_3_POS.mzML</td>
<td style="text-align: left;">QC</td>
<td style="text-align: left;">POOL3</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">7</td>
</tr>
<tr class="even">
<td style="text-align: left;">FILES/MS_E_POS.mzML</td>
<td style="text-align: left;">CTR</td>
<td style="text-align: left;">E</td>
<td style="text-align: right;">66</td>
<td style="text-align: right;">8</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FILES/MS_F_POS.mzML</td>
<td style="text-align: left;">CVD</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">44</td>
<td style="text-align: right;">9</td>
</tr>
<tr class="even">
<td style="text-align: left;">FILES/MS_QC_POOL_4_POS.mzML</td>
<td style="text-align: left;">QC</td>
<td style="text-align: left;">POOL4</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">10</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>There are 11 samples in this data set. Below are abbreviations essential for proper interpretation of this metadata information:</p>
<ul>
<li>
<em>phenotype</em>: The sample groups of the experiment:
<ul>
<li>
<code>"QC"</code>: Quality control sample (pool of serum samples from an external, large cohort).</li>
<li>
<code>"CVD"</code>: Sample from an individual with a cardiovascular disease.</li>
<li>
<code>"CTR"</code>: Sample from a presumably healthy control.</li>
</ul>
</li>
<li>
<em>sample_name</em>: An arbitrary name/identifier of the sample.</li>
<li>
<em>age</em>: The (rounded) age of the individuals.</li>
<li>
<em>injection_index</em>: An index representing the order (position) in which an individual sample was measured (injected) within the LC-MS measurement run of the experiment.</li>
</ul>
<p>We will define colors for each of the sample groups based on their sample group using the <em>RColorBrewer</em> package:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Define colors for the different phenotypes</span></span>
<span><span class="va">col_phenotype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fl">9</span>, name <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">9</span>, <span class="fl">5</span>, <span class="fl">4</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">col_phenotype</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"QC"</span>, <span class="co"># grey</span></span>
<span>                          <span class="st">"CVD"</span>, <span class="co"># orange</span></span>
<span>                          <span class="st">"CTR"</span><span class="op">)</span> <span class="co"># purple</span></span>
<span><span class="va">col_sample</span> <span class="op">&lt;-</span> <span class="va">col_phenotype</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">$</span><span class="va">phenotype</span><span class="op">]</span></span></code></pre></div>
</div>
<p>The MS data of this experiment is stored as a <code>Spectra</code> object (from the <em><a href="https://bioconductor.org/packages/3.20/Spectra" class="external-link">Spectra</a></em> Bioconductor package) within the <code>MsExperiment</code> object and can be accessed using <code><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra()</a></code> function. Each element in this object is a spectrum - they are organised linearly and are all combined in the same <code>Spectra</code> object one after the other (ordered by retention time and samples).</p>
<p>Below we access the dataset’s <code>Spectra</code> object which will summarize its available information and provide, among other things, the total number of spectra of the data set.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Access Spectra Object</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>MSn data (Spectra) with 17210 spectra in a MsBackendMetaboLights backend:
        msLevel     rtime scanIndex
      &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;
1             1     0.274         1
2             1     0.553         2
3             1     0.832         3
4             1     1.111         4
5             1     1.390         5
...         ...       ...       ...
17206         1   479.052      1717
17207         1   479.331      1718
17208         1   479.610      1719
17209         1   479.889      1720
17210         1   480.168      1721
 ... 36 more variables/columns.

file(s):
MS_QC_POOL_1_POS.mzML
MS_A_POS.mzML
MS_B_POS.mzML
 ... 7 more files</code></pre>
</div>
</div>
</section><section class="section level2"><h2 id="data-visualization-and-general-quality-assessment">Data visualization and general quality assessment<a class="anchor" aria-label="anchor" href="#data-visualization-and-general-quality-assessment"></a>
</h2>
<p>Effective visualization is paramount for inspecting and assessing the quality of MS data. For a general overview of our LC-MS data, we can:</p>
<ul>
<li>Combine all mass peaks from all (MS1) spectra of a sample into a single spectrum in which each mass peak then represents the maximum signal of all mass peaks with a similar <em>m/z</em>. This spectrum might then be called Base Peak Spectrum (BPS), providing information on the most abundant ions of a sample.</li>
<li>Aggregate mass peak intensities for each spectrum, resulting in the Base Peak Chromatogram (BPC). The BPC shows the highest measured intensity for each distinct retention time (hence spectrum) and is thus orthogonal to the BPS.</li>
<li>Sum the mass peak intensities for each spectrum to create a Total Ion Chromatogram (TIC).</li>
<li>Compare the BPS of all samples in an experiment to evaluate similarity of their ion content.</li>
<li>Compare the BPC of all samples in an experiment to identify samples with similar or dissimilar chromatographic signal.</li>
</ul>
<p>In addition to such general data evaluation and visualization, it is also crucial to investigate specific signal of e.g. internal standards or compounds/ions known to be present in the samples. By providing a reliable reference, internal standards help achieve consistent and accurate analytical results.</p>
<p>As LC signal is the most variable because of the unstable nature of the LC, we will only show the BPC and their similarity. More in depth inspection and discussion of spectra and chromatrograhic data cna be found in the <a href="https://rformassspectrometry.github.io/metabonaut/articles/data-investigation.html">data exploration vignette</a>.</p>
<section class="level2"><h3 id="chromatographic-data-visualization-bpc-and-tic">Chromatographic Data Visualization: BPC and TIC<a class="anchor" aria-label="anchor" href="#chromatographic-data-visualization-bpc-and-tic"></a>
</h3>
<p>The <code><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram()</a></code> function facilitates the extraction of intensities along the retention time. However, access to chromatographic information is currently not as efficient and seamless as it is for spectral information. Work is underway to develop/improve the infrastructure for chromatographic data through a new <code>Chromatograms</code> object aimed to be as flexible and user-friendly as the <code>Spectra</code> object.</p>
<p>For visualizing LC-MS data, a BPC or TIC serves as a valuable tool to assess the performance of liquid chromatography across various samples in an experiment. In our case, we extract the BPC from our data to create such a plot. The BPC captures the maximum peak signal from each spectrum in a data file and plots this information against the retention time for that spectrum on the y-axis. The BPC can be extracted using the <code>chromatogram</code> function.</p>
<p>By setting the parameter <code>aggregationFun = "max"</code>, we instruct the function to report the maximum signal per spectrum. Conversely, when setting <code>aggregationFun = "sum"</code>, it sums up all intensities of a spectrum, thereby creating a TIC.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Extract and plot BPC for full data</span></span>
<span><span class="va">bpc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="va">lcms1</span>, aggregationFun <span class="op">=</span> <span class="st">"max"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">bpc</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="fl">80</span><span class="op">)</span>, main <span class="op">=</span> <span class="st">"BPC"</span>, lwd <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, col <span class="op">=</span> <span class="va">col_phenotype</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">col_phenotype</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">2</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/bpc1-1.png" width="672"></p>
<figcaption>Figure 5. BPC of all samples colored by phenotype.</figcaption></figure>
</div>
</div>
</div>
<p>After about 240 seconds no signal seems to be measured. Thus, we filter the data removing that part as well as the first 10 seconds measured in the LC run.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Filter the data based on retention time</span></span>
<span><span class="va">lcms1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterRt</a></span><span class="op">(</span><span class="va">lcms1</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">240</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Filter spectra</code></pre>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bpc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="va">lcms1</span>, aggregationFun <span class="op">=</span> <span class="st">"max"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Plot after filtering</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">bpc</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="fl">80</span><span class="op">)</span>,</span>
<span>     main <span class="op">=</span> <span class="st">"BPC after filtering retention time"</span>, lwd <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, col <span class="op">=</span> <span class="va">col_phenotype</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">col_phenotype</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">2</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span>, bty <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/bpc2-1.png" width="672"></p>
<figcaption>Figure 6. BPC after filtering retention time.</figcaption></figure>
</div>
</div>
</div>
<p>Initially, we examined the entire BPC and subsequently filtered it based on the desired retention times. This not only results in a smaller file size but also facilitates a more straightforward interpretation of the BPC.</p>
<p>The final plot illustrates the BPC for each sample colored by phenotype, providing insights on the signal measured along the retention times of each sample. It reveals the points at which compounds eluted from the LC column. In essence, a BPC condenses the three-dimensional LC-MS data (<em>m/z</em> by retention time by intensity) into two dimensions (retention time by intensity).</p>
<p>We can also here compare similarities of the BPCs in a heatmap. The retention times will however not be identical between different samples. Thus we <em>bin()</em> the chromatographic signal per sample along the retention time axis into bins of two seconds resulting in data with the same number of bins/data points. We can then calculate pairwise similarities between these data vectors using the <code><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor()</a></code> function and visualize the result using <code><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap()</a></code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Total ion chromatogram</span></span>
<span><span class="va">tic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="va">lcms1</span>, aggregationFun <span class="op">=</span> <span class="st">"sum"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">bin</a></span><span class="op">(</span>binSize <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Calculate similarity (Pearson correlation) between BPCs</span></span>
<span><span class="va">ticmap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">cbind</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">tic</span>, <span class="va">intensity</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ticmap</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">ticmap</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">$</span><span class="va">sample_name</span></span>
<span><span class="va">ann</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>phenotype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">[</span>, <span class="st">"phenotype"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ann</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ticmap</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Plot heatmap</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">ticmap</span>, annotation_col <span class="op">=</span> <span class="va">ann</span>,</span>
<span>         annotation_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>phenotype <span class="op">=</span> <span class="va">col_phenotype</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/heatmap1-1.png" width="672"></p>
<figcaption>Figure 7. Heatmap of BPC similarities.</figcaption></figure>
</div>
</div>
</div>
<p>The heatmap above reinforces what our exploration of spectra data showed, which is a strong separation between the QC and study samples. This is important to bear in mind for later analyses.</p>
</section><section class="level2"><h3 id="chromatographic-data-visualization-eics">Chromatographic Data Visualization: EICs<a class="anchor" aria-label="anchor" href="#chromatographic-data-visualization-eics"></a>
</h3>
<p>Throughout the entire process, it is crucial to have reference points within the dataset, such as well-known ions. Most experiments nowadays include internal standards (IS), and it was the case here. We strongly recommend using them for visualization throughout the entire analysis. For this experiment, a set of 15 IS was spiked to all samples. After reviewing signal of all of them, we selected two to guide this analysis process. However, we also advise to plot and evaluate all the ions after each steps.</p>
<p>To illustrate this, we generate Extracted Ion Chromatograms (EIC) for these selected <em>test ions</em>. By restricting the MS data to intensities within a restricted, small <em>m/z</em> range and a selected retention time window, EICs are expected to contain only signal from a single type of ion. The expected <em>m/z</em> and retention times for our set of IS was determined in a different experiment. Additionally, in cases where internal standards are not available, commonly present ions in the sample matrix can serve as suitable alternatives. Ideally, these compounds should be distributed across the entire retention time range of the experiment.</p>
<div class="cell" data-tbl-style="font-size: 0.8em" data-tbl-colwidths="[20,20,20]">
<div class="sourceCode cell-code" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Load our list of standard</span></span>
<span><span class="va">intern_standard</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="st">"intern_standard_list.txt"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract EICs for the list</span></span>
<span><span class="va">eic_is</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span></span>
<span>    <span class="va">lcms1</span>,</span>
<span>    rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">intern_standard</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rtmin"</span>, <span class="st">"rtmax"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,</span>
<span>    mz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">intern_standard</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mzmin"</span>, <span class="st">"mzmax"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Add internal standard metadata</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">eic_is</span><span class="op">)</span><span class="op">$</span><span class="va">mz</span> <span class="op">&lt;-</span> <span class="va">intern_standard</span><span class="op">$</span><span class="va">mz</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">eic_is</span><span class="op">)</span><span class="op">$</span><span class="va">rt</span> <span class="op">&lt;-</span> <span class="va">intern_standard</span><span class="op">$</span><span class="va">RT</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">eic_is</span><span class="op">)</span><span class="op">$</span><span class="va">name</span> <span class="op">&lt;-</span> <span class="va">intern_standard</span><span class="op">$</span><span class="va">name</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">eic_is</span><span class="op">)</span><span class="op">$</span><span class="va">abbreviation</span> <span class="op">&lt;-</span> <span class="va">intern_standard</span><span class="op">$</span><span class="va">abbreviation</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">eic_is</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">intern_standard</span><span class="op">$</span><span class="va">abbreviation</span></span>
<span></span>
<span><span class="va">fdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">eic_is</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Summary of IS information</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">eic_is</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"name"</span>, <span class="st">"mz"</span>, <span class="st">"rt"</span><span class="op">)</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">kable</span><span class="op">(</span>format <span class="op">=</span> <span class="st">"pipe"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<table style="width:80%;" class="table">
<caption>Table 4. Internal standard list with respective m/z and expected retention time [s].</caption>
<colgroup>
<col style="width: 20%">
<col style="width: 20%">
<col style="width: 20%">
<col style="width: 20%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">name</th>
<th style="text-align: right;">mz</th>
<th style="text-align: right;">rt</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">carnitine_d3</td>
<td style="text-align: left;">Carnitine (D3)</td>
<td style="text-align: right;">165.13128</td>
<td style="text-align: right;">61</td>
</tr>
<tr class="even">
<td style="text-align: left;">creatinine_methyl_d3</td>
<td style="text-align: left;">Creatinine (N-Methhyl_D3)</td>
<td style="text-align: right;">117.08498</td>
<td style="text-align: right;">126</td>
</tr>
<tr class="odd">
<td style="text-align: left;">glucose_d2</td>
<td style="text-align: left;">Glucose (6,6-D2)</td>
<td style="text-align: right;">205.06512</td>
<td style="text-align: right;">166</td>
</tr>
<tr class="even">
<td style="text-align: left;">alanine_13C_15N</td>
<td style="text-align: left;">L-Alanine (13C3, 99%; 15N, 99%)</td>
<td style="text-align: right;">94.06208</td>
<td style="text-align: right;">167</td>
</tr>
<tr class="odd">
<td style="text-align: left;">arginine_13C_15N</td>
<td style="text-align: left;">L-Arginine HCl (13C6, 99%; 15N4, 99%)</td>
<td style="text-align: right;">185.12718</td>
<td style="text-align: right;">183</td>
</tr>
<tr class="even">
<td style="text-align: left;">aspartic_13C_15N</td>
<td style="text-align: left;">L-Aspartic acid (13C4, 99%; 15N, 99%)</td>
<td style="text-align: right;">139.05528</td>
<td style="text-align: right;">179</td>
</tr>
<tr class="odd">
<td style="text-align: left;">cystine_13C_15N</td>
<td style="text-align: left;">L-Cystine (13C6, 99%; 15N2, 99%)</td>
<td style="text-align: right;">249.04528</td>
<td style="text-align: right;">209</td>
</tr>
<tr class="even">
<td style="text-align: left;">glutamic_13C_15N</td>
<td style="text-align: left;">L-Glutamic acid (13C5, 99%; 15N, 99%)</td>
<td style="text-align: right;">154.07428</td>
<td style="text-align: right;">171</td>
</tr>
<tr class="odd">
<td style="text-align: left;">glycine_13C_15N</td>
<td style="text-align: left;">Glycine (13C2, 99%; 15N, 99%)</td>
<td style="text-align: right;">79.04308</td>
<td style="text-align: right;">168</td>
</tr>
<tr class="even">
<td style="text-align: left;">histidine_13C_15N</td>
<td style="text-align: left;">L-Histidine HCl H2O (13C6; 15N3, 99%)</td>
<td style="text-align: right;">165.08798</td>
<td style="text-align: right;">185</td>
</tr>
<tr class="odd">
<td style="text-align: left;">isoleucine_13C_15N</td>
<td style="text-align: left;">L-Isoleucine (13C6, 99%; 15N, 99%)</td>
<td style="text-align: right;">139.11908</td>
<td style="text-align: right;">154</td>
</tr>
<tr class="even">
<td style="text-align: left;">leucine_13C_15N</td>
<td style="text-align: left;">L-Leucine (13C6, 99%; 15N, 99%)</td>
<td style="text-align: right;">139.11908</td>
<td style="text-align: right;">151</td>
</tr>
<tr class="odd">
<td style="text-align: left;">lysine_13C_15N</td>
<td style="text-align: left;">L-Lysine 2HCl (13C6, 99%; 15N2, 99%)</td>
<td style="text-align: right;">155.12698</td>
<td style="text-align: right;">184</td>
</tr>
<tr class="even">
<td style="text-align: left;">methionine_13C_15N</td>
<td style="text-align: left;">L-Methionine (13C5, 99%; 15N, 99%)</td>
<td style="text-align: right;">156.07218</td>
<td style="text-align: right;">161</td>
</tr>
<tr class="odd">
<td style="text-align: left;">phenylalanine_13C_15N</td>
<td style="text-align: left;">L-Phenylalanine (13C9, 99%; 15N, 99%)</td>
<td style="text-align: right;">176.11348</td>
<td style="text-align: right;">153</td>
</tr>
<tr class="even">
<td style="text-align: left;">proline_13C_15N</td>
<td style="text-align: left;">L-Proline (13C5, 99%; 15N, 99%)</td>
<td style="text-align: right;">122.08438</td>
<td style="text-align: right;">167</td>
</tr>
<tr class="odd">
<td style="text-align: left;">serine_13C_15N</td>
<td style="text-align: left;">L-Serine (13C3, 99%; 15N, 99%)</td>
<td style="text-align: right;">110.05698</td>
<td style="text-align: right;">178</td>
</tr>
<tr class="even">
<td style="text-align: left;">threonine_13C_15N</td>
<td style="text-align: left;">L-Threonine (13C4, 99%; 15N, 99%)</td>
<td style="text-align: right;">125.07598</td>
<td style="text-align: right;">171</td>
</tr>
<tr class="odd">
<td style="text-align: left;">tyrosine_13C_15N</td>
<td style="text-align: left;">L-Tyrosine (13C9, 99%; 15N, 99%)</td>
<td style="text-align: right;">192.10838</td>
<td style="text-align: right;">166</td>
</tr>
<tr class="even">
<td style="text-align: left;">valine_13C_15N</td>
<td style="text-align: left;">L-Valine (13C5, 99%; 15N, 99%)</td>
<td style="text-align: right;">124.10008</td>
<td style="text-align: right;">164</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We below plot the EICs for isotope labeled cystine and methionine.</p>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Extract the two IS from the chromatogram object.</span></span>
<span><span class="va">eic_cystine</span> <span class="op">&lt;-</span> <span class="va">eic_is</span><span class="op">[</span><span class="st">"cystine_13C_15N"</span><span class="op">]</span></span>
<span><span class="va">eic_met</span> <span class="op">&lt;-</span> <span class="va">eic_is</span><span class="op">[</span><span class="st">"methionine_13C_15N"</span><span class="op">]</span></span>
<span></span>
<span><span class="co">#' plot both EIC</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">eic_cystine</span>, main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">eic_cystine</span><span class="op">)</span><span class="op">$</span><span class="va">name</span>, cex.axis <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>     cex.main <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>     col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="fl">80</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">eic_cystine</span><span class="op">)</span><span class="op">$</span><span class="va">rt</span>, col <span class="op">=</span> <span class="st">"red"</span>, lty <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">eic_met</span>, main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">eic_met</span><span class="op">)</span><span class="op">$</span><span class="va">name</span>, cex.axis <span class="op">=</span> <span class="fl">0.8</span>, cex.main <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>     col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="fl">80</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">eic_met</span><span class="op">)</span><span class="op">$</span><span class="va">rt</span>, col <span class="op">=</span> <span class="st">"red"</span>, lty <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, col <span class="op">=</span> <span class="va">col_phenotype</span>, legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">col_phenotype</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>,</span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/unnamed-chunk-4-1.png" width="672"></p>
<figcaption>Figure 9. EIC of cystine and methionine.</figcaption></figure>
</div>
</div>
</div>
<p>We can observe a clear concentration difference between QCs and study samples for the isotope labeled cystine ion. Meanwhile, the labeled methionine internal standard exhibits a discernible signal amidst some noise and a noticeable retention time shift between samples.</p>
</section></section><section class="section level2"><h2 id="data-preprocessing">Data preprocessing<a class="anchor" aria-label="anchor" href="#data-preprocessing"></a>
</h2>
<p>Preprocessing stands as the inaugural step in the analysis of untargeted LC-MS. It is characterized by 3 main stages: chromatographic peak detection, retention time shift correction (alignment) and correspondence which results in features defined. The primary objective of preprocessing is the quantification of signals from ions measured in a sample, addressing any potential retention time drifts between samples, and ensuring alignment of quantified signals across samples within an experiment. The final result of LC-MS data preprocessing is a numeric matrix with abundances of quantified entities in the samples of the experiment.</p>
<section class="level2"><h3 id="chromatographic-peak-detection">Chromatographic peak detection<a class="anchor" aria-label="anchor" href="#chromatographic-peak-detection"></a>
</h3>
<p>The initial preprocessing step involves detecting intensity peaks along the retention time axis, the so called <em>chromatographic peaks</em>. To achieve this, we employ the <code><a href="https://rdrr.io/pkg/xcms/man/findChromPeaks.html" class="external-link">findChromPeaks()</a></code> function within <em>xcms</em>. This function supports various algorithms for peak detection, which can be selected and configured with their respective parameter objects.</p>
<p>The preferred algorithm in this case, <em>CentWave</em>, utilizes continuous wavelet transformation (CWT)-based peak detection <span class="citation" data-cites="tautenhahn_highly_2008">(<a href="#ref-tautenhahn_highly_2008" role="doc-biblioref">Tautenhahn, Böttcher, and Neumann 2008</a>)</span>. This method is known for its effectiveness in handling non-Gaussian shaped chromatographic peaks or peaks with varying retention time widths, which are commonly encountered in HILIC separations.</p>
<p>Below we apply the CentWave algorithm with its default settings on the extracted ion chromatogram for cystine and methionine ions and evaluate its results.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Use default Centwave parameter</span></span>
<span><span class="va">param</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/findChromPeaks-centWave.html" class="external-link">CentWaveParam</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Look at the default parameters</span></span>
<span><span class="va">param</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Object of class:  CentWaveParam
 Parameters:
 - ppm: [1] 25
 - peakwidth: [1] 20 50
 - snthresh: [1] 10
 - prefilter: [1]   3 100
 - mzCenterFun: [1] "wMean"
 - integrate: [1] 1
 - mzdiff: [1] -0.001
 - fitgauss: [1] FALSE
 - noise: [1] 0
 - verboseColumns: [1] FALSE
 - roiList: list()
 - firstBaselineCheck: [1] TRUE
 - roiScales: numeric(0)
 - extendLengthMSW: [1] FALSE
 - verboseBetaColumns: [1] FALSE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Evaluate for Cystine</span></span>
<span><span class="va">cystine_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/findChromPeaks.html" class="external-link">findChromPeaks</a></span><span class="op">(</span><span class="va">eic_cystine</span>, param <span class="op">=</span> <span class="va">param</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">cystine_test</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     rt rtmin rtmax into intb maxo sn row column</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Evaluate for Methionine</span></span>
<span><span class="va">met_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/findChromPeaks.html" class="external-link">findChromPeaks</a></span><span class="op">(</span><span class="va">eic_met</span>, param <span class="op">=</span> <span class="va">param</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">met_test</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     rt rtmin rtmax into intb maxo sn row column</code></pre>
</div>
</div>
<p>While <em>CentWave</em> is a highly performant algorithm, it requires to be costumized to each dataset. This implies that the parameters should be fine-tuned based on the user’s data. The example above serves as a clear motivation for users to familiarize themselves with the various parameters and the need to adapt them to a data set. We will discuss the main parameters that can be easily adjusted to suit the user’s dataset:</p>
<ul>
<li><p><code>peakwidth</code>: Specifies the minimal and maximal expected width of the peaks in the retention time dimension. Highly dependent on the chromatographic settings used.</p></li>
<li><p><code>ppm</code>: The maximal allowed difference of mass peaks’ <em>m/z</em> values (in parts-per-million) in consecutive scans to consider them representing signal from the same ion.</p></li>
<li><p><code>integrate</code>: This parameter defines the integration method. Here, we primarily use <code>integrate = 2</code> because it integrates also signal of a chromatographic peak’s tail and is considered more accurate by the developers.</p></li>
</ul>
<p>To determine <code>peakwidth</code>, we recommend that users refer to previous EICs and estimate the range of peak widths they observe in their dataset. Ideally, examining multiple EICs should be the goal. For this dataset, the peak widths appear to be around 2 to 10 seconds. We do not advise choosing a range that is too wide or too narrow with the <code>peakwidth</code> parameter as it can lead to false positives or negatives.</p>
<p>To determine the <code>ppm</code>, a deeper analysis of the dataset is needed. It is clarified above that <code>ppm</code> depends on the instrument, but users should not necessarily input the vendor-advertised ppm. Here’s how to determine it as accurately as possible:</p>
<p>The following steps involve generating a highly restricted MS area with a single mass peak per spectrum, representing the cystine ion. The <em>m/z</em> of these peaks is then extracted, their absolute difference calculated and finally expressed in ppm.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Restrict the data to signal from cystine in the first sample</span></span>
<span><span class="va">cst</span> <span class="op">&lt;-</span> <span class="va">lcms1</span><span class="op">[</span><span class="fl">1L</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterRt</a></span><span class="op">(</span>rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">208</span>, <span class="fl">218</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterMzRange</a></span><span class="op">(</span>mz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">eic_cystine</span><span class="op">)</span><span class="op">[</span><span class="st">"cystine_13C_15N"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mzmin"</span>, <span class="st">"mzmax"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Show the number of peaks per m/z filtered spectra</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lengths.html" class="external-link">lengths</a></span><span class="op">(</span><span class="va">cst</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Calculate the difference in m/z values between scans</span></span>
<span><span class="va">mz_diff</span> <span class="op">&lt;-</span> <span class="va">cst</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">mz</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Express differences in ppm</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">mz_diff</span> <span class="op">*</span> <span class="fl">1e6</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">mz</a></span><span class="op">(</span><span class="va">cst</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  0.08829605 14.82188728</code></pre>
</div>
</div>
<p>We therefore, choose a value close to the maximum within this range for parameter <code>ppm</code>, i.e., 15 ppm.</p>
<p>We can now perform the chromatographic peak detection with the adapted settings on our EICs. It is important to note that, to properly estimate background noise, sufficient data points outside the chromatographic peak need to be present. This is generally no problem if peak detection is performed on the full LC-MS data set, but for peak detection on EICs the retention time range of the EIC needs to be sufficiently wide. If the function fails to find a peak in an EIC, the initial troubleshooting step should be to increase this range. Additionally, the signal-to-noise threshold <code>snthresh</code> should be reduced for peak detection in EICs, because within the small retention time range, not enough signal is present to properly estimate the background noise. Finally, in case of too few MS1 data points per peaks, setting CentWave’s advanced parameter <code>extendLengthMSW</code> to <code>TRUE</code> can help with peak detection.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Parameters adapted for chromatographic peak detection on EICs.</span></span>
<span><span class="va">param</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/findChromPeaks-centWave.html" class="external-link">CentWaveParam</a></span><span class="op">(</span>peakwidth <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">8</span><span class="op">)</span>, ppm <span class="op">=</span> <span class="fl">15</span>, integrate <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                       snthresh <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Evaluate on the cystine ion</span></span>
<span><span class="va">cystine_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/findChromPeaks.html" class="external-link">findChromPeaks</a></span><span class="op">(</span><span class="va">eic_cystine</span>, param <span class="op">=</span> <span class="va">param</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">cystine_test</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           rt   rtmin   rtmax      into      intb      maxo sn row column
 [1,] 209.251 207.577 212.878  4085.675  2911.376  2157.459  4   1      1
 [2,] 209.251 206.182 213.995 24625.728 19074.407 12907.487  4   1      2
 [3,] 209.252 207.020 214.274 19467.836 14594.041  9996.466  4   1      3
 [4,] 209.251 207.577 212.041  4648.229  3202.617  2458.485  3   1      4
 [5,] 208.974 206.184 213.159 23801.825 18126.978 11300.289  3   1      5
 [6,] 209.250 207.018 213.714 25990.327 21036.768 13650.329  5   1      6
 [7,] 209.252 207.857 212.879  4528.767  3259.039  2445.841  4   1      7
 [8,] 209.252 207.299 213.995 23119.449 17274.140 12153.410  4   1      8
 [9,] 208.972 206.740 212.878 28943.188 23436.119 14451.023  4   1      9
[10,] 209.252 207.578 213.437  4470.552  3065.402  2292.881  4   1     10</code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Evaluate on the methionine ion</span></span>
<span><span class="va">met_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/findChromPeaks.html" class="external-link">findChromPeaks</a></span><span class="op">(</span><span class="va">eic_met</span>, param <span class="op">=</span> <span class="va">param</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">met_test</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           rt   rtmin   rtmax     into     intb      maxo sn row column
 [1,] 159.867 157.913 162.378 20026.61 14715.42 12555.601  4   1      1
 [2,] 160.425 157.077 163.215 16827.76 11843.39  8407.699  3   1      2
 [3,] 160.425 157.356 163.215 18262.45 12881.67  9283.375  3   1      3
 [4,] 159.588 157.635 161.820 20987.72 15424.25 13327.811  4   1      4
 [5,] 160.985 156.799 163.217 16601.72 11968.46 10012.396  4   1      5
 [6,] 160.982 157.634 163.214 17243.24 12389.94  9150.079  4   1      6
 [7,] 159.867 158.193 162.099 21120.10 16202.05 13531.844  3   1      7
 [8,] 160.426 157.356 162.937 18937.40 13739.73 10336.000  3   1      8
 [9,] 160.704 158.472 163.215 17882.21 12299.43  9395.548  3   1      9
[10,] 160.146 157.914 162.379 20275.80 14279.50 12669.821  3   1     10</code></pre>
</div>
</div>
<p>With the customized parameters, a chromatographic peak was detected in each sample. Below, we use the <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> function on the EICs to visualize these results.</p>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Plot test chromatogram</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">cystine_test</span>, main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">cystine_test</span><span class="op">)</span><span class="op">$</span><span class="va">name</span>,</span>
<span>     col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="fl">80</span><span class="op">)</span>,</span>
<span>     peakBg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="fl">40</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">cystine_test</span><span class="op">)</span><span class="op">[</span>, <span class="st">"column"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>     cex.main <span class="op">=</span> <span class="fl">0.8</span>, cex.axis <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">met_test</span>, main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">met_test</span><span class="op">)</span><span class="op">$</span><span class="va">name</span>,</span>
<span>     col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="fl">80</span><span class="op">)</span>,</span>
<span>     peakBg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="fl">40</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">met_test</span><span class="op">)</span><span class="op">[</span>, <span class="st">"column"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>     cex.main <span class="op">=</span> <span class="fl">0.8</span>, cex.axis <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, col <span class="op">=</span> <span class="va">col_phenotype</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">col_phenotype</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, bty <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/unnamed-chunk-6-1.png" width="672"></p>
<figcaption>Figure 11. Chromatographic peak detection on EICs.</figcaption></figure>
</div>
</div>
</div>
<p>We can see a peak seems ot be detected in each sample for both ions. This indicates that our custom settings seem thus to be suitable for our dataset. We now proceed and apply them to the entire dataset, extracting EICs again for the same ions to evaluate and confirm that chromatographic peak detection worked as expected. Note:</p>
<ul>
<li><p>We revert the value for parameter <code>snthresh</code> to its default, because, as mentioned above, background noise estimation is more reliable when performed on the full data set.</p></li>
<li><p>Parameter <code>chunkSize</code> of <code><a href="https://rdrr.io/pkg/xcms/man/findChromPeaks.html" class="external-link">findChromPeaks()</a></code> defines the number of data files that are loaded into memory and processed simultaneously. This parameter thus allows to fine-tune the memory demand as well as performance of the chromatographic peak detection step.</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Using the same settings, but with default snthresh</span></span>
<span><span class="va">param</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/findChromPeaks-centWave.html" class="external-link">CentWaveParam</a></span><span class="op">(</span>peakwidth <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">8</span><span class="op">)</span>, ppm <span class="op">=</span> <span class="fl">15</span>, integrate <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">lcms1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/findChromPeaks.html" class="external-link">findChromPeaks</a></span><span class="op">(</span><span class="va">lcms1</span>, param <span class="op">=</span> <span class="va">param</span>, chunkSize <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Update EIC internal standard object</span></span>
<span><span class="va">eics_is_noprocess</span> <span class="op">&lt;-</span> <span class="va">eic_is</span></span>
<span><span class="va">eic_is</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="va">lcms1</span>,,</span>
<span>                       rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">intern_standard</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rtmin"</span>, <span class="st">"rtmax"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                       mz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">intern_standard</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mzmin"</span>, <span class="st">"mzmax"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Processing chromatographic peaks</code></pre>
</div>
<div class="sourceCode cell-code" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">eic_is</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">eics_is_noprocess</span><span class="op">)</span></span></code></pre></div>
</div>
<p>Below we plot the EICs of the two selected internal standards to evaluate the chromatographic peak detection results.</p>
<div class="cell">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/plot-new-eics-1.png" width="672"></p>
<figcaption>Figure 12. Chromatographic peak detection on EICs after processing.</figcaption></figure>
</div>
</div>
</div>
<p>Peaks seem to have been detected properly in all samples for both ions. This indicates that the peak detection process over the entire dataset was successful.</p>
<section class="level3"><h4 id="refine-identified-chromatographic-peaks">Refine identified chromatographic peaks<a class="anchor" aria-label="anchor" href="#refine-identified-chromatographic-peaks"></a>
</h4>
<p>The identification of chromatographic peaks using the <em>CentWave</em> algorithm can sometimes result in artifacts, such as overlapping or split peaks. To address this issue, the <code><a href="https://rdrr.io/pkg/xcms/man/refineChromPeaks.html" class="external-link">refineChromPeaks()</a></code> function is utilized, in conjunction with <code>MergeNeighboringPeaksParam</code>, which aims at merging such split peaks.</p>
<p>Below we show some examples of <em>CentWave</em> peak detection artifacts. These examples are pre-selected to illustrate the necessity of the next step:</p>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Extract m/z-rt regions for selected peaks</span></span>
<span><span class="va">mz_rt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span>rtmin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">155.2120</span>, <span class="fl">181.71800</span><span class="op">)</span>,</span>
<span>               rtmax <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">201.0710</span>, <span class="fl">228.13500</span><span class="op">)</span>,</span>
<span>               mzmin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">191.0288</span>,  <span class="fl">53.50964</span><span class="op">)</span>,</span>
<span>               mzmax <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">191.0527</span>, <span class="fl">53.53183</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Extract the EICs</span></span>
<span><span class="va">eics</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>, rt <span class="op">=</span> <span class="va">mz_rt</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rtmin"</span>, <span class="st">"rtmax"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                     mz <span class="op">=</span> <span class="va">mz_rt</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mzmin"</span>, <span class="st">"mzmax"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output cell-output-stderr">
<pre><code>Processing chromatographic peaks</code></pre>
</div>
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Plot the EICs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">eics</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/unnamed-chunk-7-1.png" width="672"></p>
<figcaption>Figure 13. Examples of CentWave peak detection artifacts.</figcaption></figure>
</div>
</div>
</div>
<p>In both cases the signal presumably from a single type of ion was split into two separate chromatographic peaks (indicated by the vertical line). The <code>MergeNeigboringPeaksParam</code> allows to combine such split peaks. The parameters for this algorithm are defined below:</p>
<ul>
<li>
<code>expandMz</code>and <code>expandRt</code>: Define which chromatographic peaks should be evaluated for merging.
<ul>
<li>
<code>expandMz</code>: Suggested to be kept relatively small (here at 0.0015) to prevent the merging of isotopes.</li>
<li>
<code>expandRt</code>: Usually set to approximately half the size of the average retention time width used for chromatographic peak detection (in this case, 2.5 seconds).</li>
</ul>
</li>
<li>
<code>minProp</code>: Used to determine whether candidates will be merged. Chromatographic peaks with overlapping <em>m/z</em> ranges (expanded on each side by <code>expandMz</code>) and with a tail-to-head distance in the retention time dimension that is less than <code>2 * expandRt</code>, and for which the signal between them is higher than <code>minProp</code> of the apex intensity of the chromatographic peak with the lower intensity, are merged. Values for this parameter should not be too small to avoid merging closely co-eluting ions, such as isomers.</li>
</ul>
<p>We test these settings below on the EICs with the split peaks.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' set up the parameter</span></span>
<span><span class="va">param</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/refineChromPeaks.html" class="external-link">MergeNeighboringPeaksParam</a></span><span class="op">(</span>expandRt <span class="op">=</span> <span class="fl">2.5</span>, expandMz <span class="op">=</span> <span class="fl">0.0015</span>,</span>
<span>                                    minProp <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Perform the peak refinement on the EICs</span></span>
<span><span class="va">eics</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/refineChromPeaks.html" class="external-link">refineChromPeaks</a></span><span class="op">(</span><span class="va">eics</span>, param <span class="op">=</span> <span class="va">param</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">eics</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/test-merging-1.png" width="672"></p>
<figcaption>Figure 14. Examples of CentWave peak detection artifacts after merging.</figcaption></figure>
</div>
</div>
</div>
<p>We can observe that the artificially split peaks have been appropriately merged. Therefore, we next apply these settings to the entire dataset. After peak merging, column <code>"merged"</code> in the result object’s <code><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeakData()</a></code> data frame can be used to evaluate which chromatographic peaks in the result represent signal from merged, or from the originally identified chromatographic peaks.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Apply on whole dataset</span></span>
<span><span class="va">lcms1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/refineChromPeaks.html" class="external-link">refineChromPeaks</a></span><span class="op">(</span><span class="va">lcms1</span>, param <span class="op">=</span> <span class="va">param</span>, chunkSize <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Reduced from 106714 to 89182 chromatographic peaks.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeakData</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">$</span><span class="va">merged</span> <span class="op">|&gt;</span></span>
<span>                      <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
FALSE  TRUE
79908  9274 </code></pre>
</div>
</div>
<p>Before proceeding with the next preprocessing step it is generally suggested to evaluate the results of the chromatographic peak detection on EICs of e.g. internal standards or other compounds/ions known to be present in the samples. Here we update our EICs objects for the internal standards with the results of the chromatographic peak detection and refinement.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">eics_is_chrompeaks</span> <span class="op">&lt;-</span> <span class="va">eic_is</span></span>
<span></span>
<span><span class="va">eic_is</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="va">lcms1</span>,</span>
<span>                       rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">intern_standard</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rtmin"</span>, <span class="st">"rtmax"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                       mz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">intern_standard</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mzmin"</span>, <span class="st">"mzmax"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">eic_is</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">eics_is_chrompeaks</span><span class="op">)</span></span>
<span></span>
<span><span class="va">eic_cystine</span> <span class="op">&lt;-</span> <span class="va">eic_is</span><span class="op">[</span><span class="st">"cystine_13C_15N"</span>, <span class="op">]</span></span>
<span><span class="va">eic_met</span> <span class="op">&lt;-</span> <span class="va">eic_is</span><span class="op">[</span><span class="st">"methionine_13C_15N"</span>, <span class="op">]</span></span></code></pre></div>
</div>
<p>Additionally, evaluating and comparing the number of identified chromatographic peaks in all samples of a data set can help spotting potentially problematic samples. Below we count the number of chromatographic peaks per sample and show these numbers in a table.</p>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Count the number of peaks per sample and summarize them in a table.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>sample_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">$</span><span class="va">sample_name</span>,</span>
<span>           peak_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">kable</span><span class="op">(</span>format <span class="op">=</span> <span class="st">"pipe"</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<table style="width:100%;" class="table">
<caption>Table 5. Samples and number of identified chromatographic peaks.</caption>
<colgroup>
<col style="width: 18%">
<col style="width: 9%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 9%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 9%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 9%">
</colgroup>
<tbody>
<tr class="odd">
<td style="text-align: left;">sample_name</td>
<td style="text-align: left;">POOL1</td>
<td style="text-align: left;">A</td>
<td style="text-align: left;">B</td>
<td style="text-align: left;">POOL2</td>
<td style="text-align: left;">C</td>
<td style="text-align: left;">D</td>
<td style="text-align: left;">POOL3</td>
<td style="text-align: left;">E</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">POOL4</td>
</tr>
<tr class="even">
<td style="text-align: left;">peak_count</td>
<td style="text-align: left;">9287</td>
<td style="text-align: left;">8986</td>
<td style="text-align: left;">8738</td>
<td style="text-align: left;">9193</td>
<td style="text-align: left;">8351</td>
<td style="text-align: left;">8778</td>
<td style="text-align: left;">9211</td>
<td style="text-align: left;">8787</td>
<td style="text-align: left;">8515</td>
<td style="text-align: left;">9336</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>A similar number of chromatographic peaks was identified within the various samples of the data set.</p>
<p>Additional options to evaluate the results of the chromatographic peak detection can be implemented using the <code><a href="https://rdrr.io/pkg/xcms/man/plotChromPeaks.html" class="external-link">plotChromPeaks()</a></code> function or by summarizing the results using base R commands.</p>
</section></section><section class="level2"><h3 id="retention-time-alignment">Retention time alignment<a class="anchor" aria-label="anchor" href="#retention-time-alignment"></a>
</h3>
<p>Despite using the same chromatographic settings and conditions retention time shifts are unavoidable. Indeed, the performance of the instrument can change over time, for example due to small variations in environmental conditions, such as temperature and pressure. These shifts will be generally small if samples are measured within the same batch/measurement run, but can be considerable if data of an experiment was acquired across a longer time period.</p>
<p>To evaluate the presence of a shift we extract and plot below the BPC from the QC samples.</p>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Get QC samples</span></span>
<span><span class="va">QC_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">$</span><span class="va">phenotype</span> <span class="op">==</span> <span class="st">"QC"</span></span>
<span></span>
<span><span class="co">#' extract BPC</span></span>
<span><span class="va">lcms1</span><span class="op">[</span><span class="va">QC_samples</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span>aggregationFun <span class="op">=</span> <span class="st">"max"</span>, chromPeaks <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span>col <span class="op">=</span> <span class="va">col_phenotype</span><span class="op">[</span><span class="st">"QC"</span><span class="op">]</span>, main <span class="op">=</span> <span class="st">"BPC of QC samples"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/unnamed-chunk-11-1.png" width="672"></p>
<figcaption>Figure 15. BPC of QC samples.</figcaption></figure>
</div>
</div>
</div>
<p>These QC samples representing the same sample (pool) were measured at regular intervals during the measurement run of the experiment and were all measured on the same day. Still, small shifts can be observed, especially in the region between 100 and 150 seconds. To facilitate proper correspondence of signals across samples (and hence definition of the LC-MS features), it is essential to minimize these differences in retention times.</p>
<p>Theoretically, we proceed in two steps: first we select only the QC samples of our dataset and do a first alignment on these, using the so-called <em>anchor peaks</em>. In this way we can assume a linear shift in time, since we are always measuring the same sample in different regular time intervals. Despite having external QCs in our data set, we still use the subset-based alignment assuming retention time shifts to be independent of the different sample matrix (human serum or plasma) and instead are mostly instrument-dependent. Note that it would also be possible to manually specify anchor peaks, respectively their retention times or to align a data set against an external, reference, data set. More information is provided in the vignettes of the <em><a href="https://bioconductor.org/packages/3.20/xcms" class="external-link">xcms</a></em> package.</p>
<p>After calculating how much to adjust the retention time in these samples, we apply this shift also on the study samples.</p>
<p>In <em>xcms</em> retention time alignment can be performed using the <code><a href="https://rdrr.io/pkg/xcms/man/adjustRtime.html" class="external-link">adjustRtime()</a></code> function with an alignment algorithm. For this example we use the <em>PeakGroups</em> method <span class="citation" data-cites="smith_xcms_2006">(<a href="#ref-smith_xcms_2006" role="doc-biblioref">Smith et al. 2006</a>)</span> that performs the alignment by minimizing differences in retention times of a set of <em>anchor peaks</em> in the different samples. This method requires an initial correspondence analysis to match/group chromatographic peaks across samples from which the algorithm then selects the anchor peaks for the alignment.</p>
<p>For the initial correspondence, we use the <em>PeakDensity</em> approach <span class="citation" data-cites="smith_xcms_2006">(<a href="#ref-smith_xcms_2006" role="doc-biblioref">Smith et al. 2006</a>)</span> that groups chromatographic peaks with similar <em>m/z</em> and retention time into LC-MS features. The parameters for this algorithm, that can be configured using the <code>PeakDensityParam</code> object, are <code>sampleGroups</code>, <code>minFraction</code>, <code>binSize</code>, <code>ppm</code> and <code>bw</code>.</p>
<p><code>binSize</code>, <code>ppm</code> and <code>bw</code> allow to specify how similar the chromatographic peaks’ <em>m/z</em> and retention time values need to be to consider them for grouping into a feature.</p>
<ul>
<li><p><code>binSize</code> and <code>ppm</code> define the required similarity of <em>m/z</em> values. Within each <em>m/z</em> bin (defined by <code>binSize</code> and <code>ppm</code>) areas along the retention time axis with a high chromatographic peak density (considering all peaks in all samples) are identified, and chromatographic peaks within these regions are considered for grouping into a feature.</p></li>
<li><p>High density areas are identified using the base R <code><a href="https://rdrr.io/r/stats/density.html" class="external-link">density()</a></code> function, for which <code>bw</code> is a parameter: higher values define wider retention time areas, lower values require chromatographic peaks to have more similar retention times. This parameter can be seen as the black line on the plot below, corresponding to the smoothness of the density curve.</p></li>
</ul>
<p>Whether such candidate peaks get grouped into a feature depends also on parameters <code>sampleGroups</code> and <code>minFraction</code>:</p>
<ul>
<li><p><code>sampleGroups</code> should provide, for each sample, the sample group it belongs to.</p></li>
<li><p><code>minFraction</code> is expected to be a value between 0 and 1 defining the proportion of samples within at least one of the sample groups (defined with <code>sampleGroups</code>) in which a chromatographic peaks was detected to group them into a feature.</p></li>
</ul>
<p>For the initial correspondence, parameters don’t need to be fully optimized. Selection of dataset-specific parameter values is described in more detail in the next section. For our dataset, we use small values for <code>binSize</code> and <code>ppm</code> and, importantly, also for parameter <code>bw</code>, since for our data set an ultra high performance (UHP) LC setup was used. For <code>minFraction</code> we use a high value (0.9) to ensure only features are defined for chromatographic peaks present in almost all samples of one sample group (which can then be used as anchor peaks for the actual alignment). We will base the alignment later on QC samples only and hence define for <code>sampleGroups</code> a binary variable grouping samples either into a study, or QC group.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Initial correspondence analysis</span></span>
<span><span class="va">param</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/groupChromPeaks.html" class="external-link">PeakDensityParam</a></span><span class="op">(</span>sampleGroups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">$</span><span class="va">phenotype</span> <span class="op">==</span> <span class="st">"QC"</span>,</span>
<span>                          minFraction <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>                          binSize <span class="op">=</span> <span class="fl">0.01</span>, ppm <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                          bw <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">lcms1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/groupChromPeaks.html" class="external-link">groupChromPeaks</a></span><span class="op">(</span><span class="va">lcms1</span>, param <span class="op">=</span> <span class="va">param</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/plotChromPeakDensity.html" class="external-link">plotChromPeakDensity</a></span><span class="op">(</span></span>
<span>    <span class="va">eic_cystine</span>, param <span class="op">=</span> <span class="va">param</span>,</span>
<span>    col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="st">"80"</span><span class="op">)</span>,</span>
<span>    peakCol <span class="op">=</span> <span class="va">col_sample</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">eic_cystine</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    peakBg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">eic_cystine</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span>, <span class="fl">20</span><span class="op">)</span>,</span>
<span>    peakPch <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/unnamed-chunk-12-1.png" width="672"></p>
<figcaption>Figure 16. Initial correspondence analysis.</figcaption></figure>
</div>
</div>
</div>
<p><em>PeakGroups</em>-based alignment can next be performed using the <code><a href="https://rdrr.io/pkg/xcms/man/adjustRtime.html" class="external-link">adjustRtime()</a></code> function with a <code>PeakGroupsParam</code> parameter object. The parameters for this algorithm are:</p>
<ul>
<li><p><code>subsetAdjust</code> and <code>subset</code>: Allows for subset alignment. Here we base the retention time alignment on the QC samples, i.e., retention time shifts will be estimated based on these repeatedly measured samples. The resulting adjustment is then applied to the entire data. For data sets in which QC samples (e.g. sample pools) are measured repeatedly, we strongly suggest to use this method. Note also that for subset-based alignment the samples should be ordered by injection index (i.e., in the order in which they were measured during the measurement run).</p></li>
<li><p><code>minFraction</code>: A value between 0 and 1 defining the proportion of samples (of the full data set, or the data subset defined with <code>subset</code>) in which a chromatographic peak has to be identified to use it as <em>anchor peak</em>. This is in contrast to the <code>PeakDensityParam</code> where this parameter was used to define a proportion within a sample group.</p></li>
<li><p><code>span</code>: The <em>PeakGroups</em> method allows, depending on the data, to adjust regions along the retention time axis differently. To enable such local alignments the <em>LOESS</em> function is used and this parameter defines the degree of smoothing of this function. Generally, values between 0.4 and 0.6 are used, however, it is suggested to evaluate alignment results and eventually adapt parameters if the result was not satisfactory.</p></li>
</ul>
<p>Below we perform the alignment of our data set based on retention times of anchor peaks defined in the subset of QC samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Define parameters of choice</span></span>
<span><span class="va">subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">$</span><span class="va">phenotype</span> <span class="op">==</span> <span class="st">"QC"</span><span class="op">)</span></span>
<span><span class="va">param</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/adjustRtime.html" class="external-link">PeakGroupsParam</a></span><span class="op">(</span>minFraction <span class="op">=</span> <span class="fl">0.9</span>, extraPeaks <span class="op">=</span> <span class="fl">50</span>, span <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                         subsetAdjust <span class="op">=</span> <span class="st">"average"</span>,</span>
<span>                         subset <span class="op">=</span> <span class="va">subset</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Perform the alignment</span></span>
<span><span class="va">lcms1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/adjustRtime.html" class="external-link">adjustRtime</a></span><span class="op">(</span><span class="va">lcms1</span>, param <span class="op">=</span> <span class="va">param</span><span class="op">)</span></span></code></pre></div>
</div>
<p>Alignment adjusted the retention times of all spectra in the data set, as well as the retention times of all identified chromatographic peaks.</p>
<p>Once the alignment has been performed, the user should evaluate the results using the <code><a href="https://rdrr.io/pkg/xcms/man/plotAdjustedRtime.html" class="external-link">plotAdjustedRtime()</a></code> function. This function visualizes the difference between adjusted and raw retention time for each sample on the y-axis along the adjusted retention time on the x-axis. Dot points represent the position of the used anchor peak along the retention time axis. For optimal alignment in all areas along the retention time axis, these anchor peaks should be scattered all over the retention time dimension.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Visualize alignment results</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/plotAdjustedRtime.html" class="external-link">plotAdjustedRtime</a></span><span class="op">(</span><span class="va">lcms1</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="fl">80</span><span class="op">)</span>, peakGroupsPch <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, col <span class="op">=</span> <span class="va">col_phenotype</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">col_phenotype</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, bty <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/unnamed-chunk-14-1.png" width="672"></p>
<figcaption>Figure 17. Retention time alignment results.</figcaption></figure>
</div>
</div>
</div>
<p>All samples from the present data set were measured within the same measurement run, resulting in small retention time shifts. Therefore, only little adjustments needed to be performed (shifts of at maximum 1 second as can be seen in the plot above). Generally, the magnitude of adjustment seen in such plots should match the expectation from the analyst.</p>
<p>We can also compare the BPC before and after alignment. To get the original data, i.e. the raw retention times, we can use the <code><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">dropAdjustedRtime()</a></code> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Get data before alignment</span></span>
<span><span class="va">lcms1_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">dropAdjustedRtime</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Apply the adjusted retention time to our dataset</span></span>
<span><span class="va">lcms1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/applyAdjustedRtime.html" class="external-link">applyAdjustedRtime</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Plot the BPC before and after alignment</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span>, mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="va">lcms1_raw</span>, aggregationFun <span class="op">=</span> <span class="st">"max"</span>, chromPeaks <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span>main <span class="op">=</span> <span class="st">"BPC before alignment"</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="fl">80</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, col <span class="op">=</span> <span class="va">col_phenotype</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">col_phenotype</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="va">lcms1</span>, aggregationFun <span class="op">=</span> <span class="st">"max"</span>, chromPeaks <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span>main <span class="op">=</span> <span class="st">"BPC after alignment"</span>,</span>
<span>         col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="fl">80</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, col <span class="op">=</span> <span class="va">col_phenotype</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">col_phenotype</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, bty <span class="op">=</span> <span class="st">"n"</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/bpc-before-and-after1-1.png" width="672"></p>
<figcaption>Figure 18. BPC before and after alignment.</figcaption></figure>
</div>
</div>
</div>
<p>The largest shift can be observed in the retention time range from 120 to 130s. Apart from that retention time range, only little changes can be observed.</p>
<p>We next evaluate the impact of the alignment on the EICs of the selected internal standards. We thus below first extract the ion chromatograms after alignment.</p>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Update the EICs</span></span>
<span><span class="va">eic_is</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="va">lcms1</span>,</span>
<span>                       rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">intern_standard</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rtmin"</span>, <span class="st">"rtmax"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                       mz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">intern_standard</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mzmin"</span>, <span class="st">"mzmax"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">eic_is</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">fdata</span></span>
<span></span>
<span><span class="co">#' Extract the EICs for the test ions</span></span>
<span><span class="va">eic_cystine</span> <span class="op">&lt;-</span> <span class="va">eic_is</span><span class="op">[</span><span class="st">"cystine_13C_15N"</span><span class="op">]</span></span>
<span><span class="va">eic_met</span> <span class="op">&lt;-</span> <span class="va">eic_is</span><span class="op">[</span><span class="st">"methionine_13C_15N"</span><span class="op">]</span></span></code></pre></div>
</details>
</div>
<p>We can now evaluate the alignment effect in our test ions. We will plot the EICs before and after alignment for both the isotope labeled cystine and methionine.</p>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span>, mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">4.5</span>, <span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">old_eic_cystine</span> <span class="op">&lt;-</span> <span class="va">eic_is</span><span class="op">[</span><span class="st">"cystine_13C_15N"</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">old_eic_cystine</span>, main <span class="op">=</span> <span class="st">"Cystine before alignment"</span>, peakType <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>     col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="fl">80</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="va">intern_standard</span><span class="op">[</span><span class="st">"cystine_13C_15N"</span>, <span class="st">"RT"</span><span class="op">]</span>, col <span class="op">=</span> <span class="st">"red"</span>, lty <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="va">old_eic_met</span> <span class="op">&lt;-</span> <span class="va">eic_is</span><span class="op">[</span><span class="st">"methionine_13C_15N"</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">old_eic_met</span>, main <span class="op">=</span> <span class="st">"Methionine before alignment"</span>,</span>
<span>     peakType <span class="op">=</span> <span class="st">"none"</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="fl">80</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="va">intern_standard</span><span class="op">[</span><span class="st">"methionine_13C_15N"</span>, <span class="st">"RT"</span><span class="op">]</span>, col <span class="op">=</span> <span class="st">"red"</span>, lty <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">eic_cystine</span>, main <span class="op">=</span> <span class="st">"Cystine after alignment"</span>, peakType <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>     col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="fl">80</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="va">intern_standard</span><span class="op">[</span><span class="st">"cystine_13C_15N"</span>, <span class="st">"RT"</span><span class="op">]</span>, col <span class="op">=</span> <span class="st">"red"</span>, lty <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, col <span class="op">=</span> <span class="va">col_phenotype</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">col_phenotype</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span>, bty <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">eic_met</span>, main <span class="op">=</span> <span class="st">"Methionine after alignment"</span>,</span>
<span>     peakType <span class="op">=</span> <span class="st">"none"</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="fl">80</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="va">intern_standard</span><span class="op">[</span><span class="st">"methionine_13C_15N"</span>, <span class="st">"RT"</span><span class="op">]</span>, col <span class="op">=</span> <span class="st">"red"</span>, lty <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/specific-ion-before-and-after1-1.png" width="672"></p>
<figcaption>Figure 19. EICs of cystine and methionine before and after alignment.</figcaption></figure>
</div>
</div>
</div>
<p>The non-endogenous cystine ion was already well aligned so the difference is minimal. The methionine ion, however, shows an improvement in alignment.</p>
<p>In addition to a visual inspection of the results, we also evaluate the impact of the alignment by comparing the variance in retention times for internal standards before and after alignment. To this end, we first need to identify chromatographic peaks in each sample with an <em>m/z</em> and retention time close to the expected values for each internal standard. For this we use the <code><a href="https://rdrr.io/pkg/MetaboAnnotation/man/matchValues.html" class="external-link">matchValues()</a></code> function from the <em><a href="https://bioconductor.org/packages/3.20/MetaboAnnotation" class="external-link">MetaboAnnotation</a></em> package <span class="citation" data-cites="rainer_modular_2022">(<a href="#ref-rainer_modular_2022" role="doc-biblioref">Rainer et al. 2022</a>)</span> using the <code>MzRtParam</code> method to identify all chromatographic peaks with similar <em>m/z</em> (+/- 50 ppm) and retention time (+/- 10 seconds) to the internal standard’s values. With parameters <code>mzColname</code> and <code>rtColname</code> we specify the column names in the query (our IS) and target (chromatographic peaks) that contain the <em>m/z</em> and retention time values on which to match entities. We perform this matching separately for each sample. For each internal standard in every sample, we use the <code><a href="https://rdrr.io/pkg/MetaboAnnotation/man/Matched.html" class="external-link">filterMatches()</a></code> function and the <code><a href="https://rdrr.io/pkg/MetaboAnnotation/man/Matched.html" class="external-link">SingleMatchParam()</a></code> parameter to select the chromatographic peak with the highest intensity.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Extract the matrix with all chromatographic peaks and add a column</span></span>
<span><span class="co">#' with the ID of the chromatographic peak</span></span>
<span><span class="va">chrom_peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">chrom_peaks</span><span class="op">$</span><span class="va">peak_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">chrom_peaks</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Define the parameters for the matching and filtering of the matches</span></span>
<span><span class="va">p_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MetaboAnnotation/man/matchValues.html" class="external-link">MzRtParam</a></span><span class="op">(</span>ppm <span class="op">=</span> <span class="fl">50</span>, toleranceRt <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">p_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MetaboAnnotation/man/Matched.html" class="external-link">SingleMatchParam</a></span><span class="op">(</span>duplicates <span class="op">=</span> <span class="st">"top_ranked"</span>, column <span class="op">=</span> <span class="st">"target_maxo"</span>,</span>
<span>                        decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Iterate over samples and identify for each the chromatographic peaks</span></span>
<span><span class="co">#' with similar m/z and retention time than the onse from the internal</span></span>
<span><span class="co">#' standard, and extract among them the ID of the peaks with the</span></span>
<span><span class="co">#' highest intensity.</span></span>
<span><span class="va">intern_standard_peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">tmp</span> <span class="op">&lt;-</span> <span class="va">chrom_peaks</span><span class="op">[</span><span class="va">chrom_peaks</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span> <span class="op">==</span> <span class="va">i</span>, , drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span>    <span class="va">mtch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MetaboAnnotation/man/matchValues.html" class="external-link">matchValues</a></span><span class="op">(</span><span class="va">intern_standard</span>, <span class="va">tmp</span>,</span>
<span>                        mzColname <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mz"</span>, <span class="st">"mz"</span><span class="op">)</span>,</span>
<span>                        rtColname <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RT"</span>, <span class="st">"rt"</span><span class="op">)</span>,</span>
<span>                        param <span class="op">=</span> <span class="va">p_1</span><span class="op">)</span></span>
<span>    <span class="va">mtch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MetaboAnnotation/man/Matched.html" class="external-link">filterMatches</a></span><span class="op">(</span><span class="va">mtch</span>, <span class="va">p_2</span><span class="op">)</span></span>
<span>    <span class="va">mtch</span><span class="op">$</span><span class="va">target_peak_id</span></span>
<span><span class="op">}</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span>what <span class="op">=</span> <span class="va">cbind</span><span class="op">)</span></span></code></pre></div>
</div>
<p>We have now for each internal standard the ID of the chromatographic peak in each sample that most likely represents signal from that ion. We can now extract the retention times for these chromatographic peaks before and after alignment.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Define the index of the selected chromatographic peaks in the</span></span>
<span><span class="co">#' full chromPeaks matrix</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">intern_standard_peaks</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Extract the raw retention times for these</span></span>
<span><span class="va">rt_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">lcms1_raw</span><span class="op">)</span><span class="op">[</span><span class="va">idx</span>, <span class="st">"rt"</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">lcms1_raw</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Extract the adjusted retention times for these</span></span>
<span><span class="va">rt_adj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">[</span><span class="va">idx</span>, <span class="st">"rt"</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">lcms1_raw</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<p>We can now evaluate the impact of the alignment on the retention times of internal standards across the full data set:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>all_raw <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/matrixStats/man/rowSds.html" class="external-link">rowSds</a></span><span class="op">(</span><span class="va">rt_raw</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>     all_adj <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/matrixStats/man/rowSds.html" class="external-link">rowSds</a></span><span class="op">(</span><span class="va">rt_adj</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>     <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/vioplot/man/vioplot.html" class="external-link">vioplot</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/unnamed-chunk-19-1.png" width="672"></p>
<figcaption>Figure 20. Retention time variation of internal standards before and after alignment.</figcaption></figure>
</div>
</div>
</div>
<p>On average, the variation in retention times of internal standards across samples was <em>very</em> slightly reduced by the alignment.</p>
</section><section class="level2"><h3 id="correspondence">Correspondence<a class="anchor" aria-label="anchor" href="#correspondence"></a>
</h3>
<p>We briefly touched on the subject of correspondence before to determine anchor peaks for alignment. Generally, the goal of the correspondence analysis is to identify chromatographic peaks that originate from the same types of ions in all samples of an experiment and to group them into LC-MS <em>features</em>. At this point, proper configuration of parameter <code>bw</code> is crucial. Here we illustrate how sensible choices for this parameter’s value can be made. We use below the <code><a href="https://rdrr.io/pkg/xcms/man/plotChromPeakDensity.html" class="external-link">plotChromPeakDensity()</a></code> function to <em>simulate</em> a correspondence analysis with the default values for <em>PeakGroups</em> on the extracted ion chromatograms of our two selected isotope labeled ions. This plot shows the EIC in the top panel, and the apex position of chromatographic peaks in the different samples (y-axis), along retention time (x-axis) in the lower panel.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Default parameter for the grouping and apply them to the test ions BPC</span></span>
<span><span class="va">param</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/groupChromPeaks.html" class="external-link">PeakDensityParam</a></span><span class="op">(</span>sampleGroups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">$</span><span class="va">phenotype</span>, bw <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span></span>
<span><span class="va">param</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Object of class:  PeakDensityParam
 Parameters:
 - sampleGroups:  [1] "QC"  "CVD" "CTR" "QC"  "CTR" "CVD" "QC"  "CTR" "CVD" "QC"
 - bw: [1] 30
 - minFraction: [1] 0.5
 - minSamples: [1] 1
 - binSize: [1] 0.25
 - maxFeatures: [1] 50
 - ppm: [1] 0</code></pre>
</div>
</div>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/plotChromPeakDensity.html" class="external-link">plotChromPeakDensity</a></span><span class="op">(</span></span>
<span>    <span class="va">eic_cystine</span>, param <span class="op">=</span> <span class="va">param</span>,</span>
<span>    col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="st">"80"</span><span class="op">)</span>,</span>
<span>    peakCol <span class="op">=</span> <span class="va">col_sample</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">eic_cystine</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    peakBg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">eic_cystine</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span>, <span class="fl">20</span><span class="op">)</span>,</span>
<span>    peakPch <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/unnamed-chunk-21-1.png" width="672"></p>
<figcaption>Figure 21. Initial correspondence analysis, Cystine.</figcaption></figure>
</div>
</div>
</div>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/plotChromPeakDensity.html" class="external-link">plotChromPeakDensity</a></span><span class="op">(</span><span class="va">eic_met</span>, param <span class="op">=</span> <span class="va">param</span>,</span>
<span>    col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="st">"80"</span><span class="op">)</span>,</span>
<span>    peakCol <span class="op">=</span> <span class="va">col_sample</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">eic_met</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    peakBg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">eic_met</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span>, <span class="fl">20</span><span class="op">)</span>,</span>
<span>    peakPch <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/unnamed-chunk-22-1.png" width="672"></p>
<figcaption>Figure 22. Initial correspondence analysis, Methionine.</figcaption></figure>
</div>
</div>
</div>
<p>Grouping of peaks depends on the smoothness of the previousl mentionned density curve that can be configured with the parameter <code>bw</code>. As seen above, the smoothness is too high to properly group our features. When looking at the default parameters, we can observe that indeed, the <code>bw</code> parameter is set to <code>bw = 30</code>, which is too high for modern UHPLC-MS setups. We reduce the value of this parameter to 1.8 and evaluate its impact.</p>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Updating parameters</span></span>
<span><span class="va">param</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/groupChromPeaks.html" class="external-link">PeakDensityParam</a></span><span class="op">(</span>sampleGroups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">$</span><span class="va">phenotype</span>, bw <span class="op">=</span> <span class="fl">1.8</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/plotChromPeakDensity.html" class="external-link">plotChromPeakDensity</a></span><span class="op">(</span></span>
<span>    <span class="va">eic_cystine</span>, param <span class="op">=</span> <span class="va">param</span>,</span>
<span>    col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="st">"80"</span><span class="op">)</span>,</span>
<span>    peakCol <span class="op">=</span> <span class="va">col_sample</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">eic_cystine</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    peakBg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">eic_cystine</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span>, <span class="fl">20</span><span class="op">)</span>,</span>
<span>    peakPch <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/unnamed-chunk-23-1.png" width="672"></p>
<figcaption>Figure 23. Correspondence analysis with optimized parameters, Cystine.</figcaption></figure>
</div>
</div>
</div>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/plotChromPeakDensity.html" class="external-link">plotChromPeakDensity</a></span><span class="op">(</span><span class="va">eic_met</span>, param <span class="op">=</span> <span class="va">param</span>,</span>
<span>    col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="st">"80"</span><span class="op">)</span>,</span>
<span>    peakCol <span class="op">=</span> <span class="va">col_sample</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">eic_met</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    peakBg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">eic_met</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span>, <span class="fl">20</span><span class="op">)</span>,</span>
<span>    peakPch <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/unnamed-chunk-24-1.png" width="672"></p>
<figcaption>Figure 24. Correspondence analysis with optimized parameters, Methionine.</figcaption></figure>
</div>
</div>
</div>
<p>We can observe that the peaks are now grouped more accurately into a single feature for each test ion. The other important parameters optimized here are:</p>
<ul>
<li><p><code>binsize</code>: Our data was generated on a high resolution MS instrument, thus we select a low value for this paramete.</p></li>
<li><p><code>ppm</code>: For TOF instruments, it is suggested to use a value for <code>ppm</code> larger than 0 to accommodate the higher measurement error of the instrument for larger <em>m/z</em> values.</p></li>
<li><p><code>minFraction</code>: We set it to <code>minFraction = 0.75</code>, hence defining features only if a chromatographic peak was identified in at least 75% of the samples of one of the sample groups.</p></li>
<li><p><code>sampleGroups</code>: We use the information available in our <code>sampleData</code>’s <code>"phenotype"</code> column.</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Define the settings for the param</span></span>
<span><span class="va">param</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/groupChromPeaks.html" class="external-link">PeakDensityParam</a></span><span class="op">(</span>sampleGroups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">$</span><span class="va">phenotype</span>,</span>
<span>                          minFraction <span class="op">=</span> <span class="fl">0.75</span>, binSize <span class="op">=</span> <span class="fl">0.01</span>, ppm <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                          bw <span class="op">=</span> <span class="fl">1.8</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Apply to whole data</span></span>
<span><span class="va">lcms1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/groupChromPeaks.html" class="external-link">groupChromPeaks</a></span><span class="op">(</span><span class="va">lcms1</span>, param <span class="op">=</span> <span class="va">param</span><span class="op">)</span></span></code></pre></div>
</div>
<p>After correspondence analysis it is suggested to evaluate the results again for selected EICs. Below we extract signal for an <em>m/z</em> similar to that of the isotope labeled methionine for a larger retention time range. Importantly, to show the actual correspondence results, we set <code>simulate = FALSE</code> for the <code><a href="https://rdrr.io/pkg/xcms/man/plotChromPeakDensity.html" class="external-link">plotChromPeakDensity()</a></code> function.</p>
<div class="cell" data-layout-align="center">
<details class="code-fold"><summary>Code</summary><div class="sourceCode cell-code" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Extract chromatogram for an m/z similar to the one of the labeled methionine</span></span>
<span><span class="va">chr_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="va">lcms1</span>,</span>
<span>                         mz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">intern_standard</span><span class="op">[</span><span class="st">"methionine_13C_15N"</span>,</span>
<span>                                                        <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mzmin"</span>, <span class="st">"mzmax"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                         rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">145</span>, <span class="fl">200</span><span class="op">)</span>,</span>
<span>                         aggregationFun <span class="op">=</span> <span class="st">"max"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/plotChromPeakDensity.html" class="external-link">plotChromPeakDensity</a></span><span class="op">(</span></span>
<span>    <span class="va">chr_test</span>, simulate <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="st">"80"</span><span class="op">)</span>,</span>
<span>    peakCol <span class="op">=</span> <span class="va">col_sample</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">chr_test</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    peakBg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">chr_test</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span>, <span class="fl">20</span><span class="op">)</span>,</span>
<span>    peakPch <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/unnamed-chunk-26-1.png" width="672"></p>
<figcaption>Figure 25. Correspondence analysis results, Methionine.</figcaption></figure>
</div>
</div>
</div>
<p>As hoped, signal from the two different ions are now grouped into separate features. Generally, correspondence results should be evaluated on more such extracted chromatograms.</p>
<p>Another interesting information to look a the the distribution of these features along the retention time axis.</p>
<div class="cell" data-tbl-cap="Table 5. Distribution of features along the retention time axis.">
<div class="sourceCode cell-code" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Bin features per RT slices</span></span>
<span><span class="va">vc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">featureDefinitions</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">$</span><span class="va">rtmed</span></span>
<span><span class="va">breaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">vc</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span>, length.out <span class="op">=</span> <span class="fl">15</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">cuts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cut.html" class="external-link">cut</a></span><span class="op">(</span><span class="va">vc</span>, breaks <span class="op">=</span> <span class="va">breaks</span>, include.lowest <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">cuts</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>cuts
   [0,17]   (17,34]   (34,51]   (51,68]   (68,86]  (86,103] (103,120] (120,137]
       15      2608       649        24       132        11        15       105
(137,154] (154,171] (171,188] (188,205] (205,222] (222,240]
     1441      2202       410       869       554        33 </code></pre>
</div>
</div>
<p>The results from the correspondence analysis are now stored, along with the results from the other preprocessing steps, within our <code>XcmsExperiment</code> result object. The correspondence results, i.e., the definition of the LC-MS features, can be extracted using the <code><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">featureDefinitions()</a></code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Definition of the features</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">featureDefinitions</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          mzmed    mzmin    mzmax    rtmed    rtmin    rtmax npeaks CTR CVD QC
FT0001 50.98979 50.98949 50.99038 203.6001 203.1181 204.2331      8   1   3  4
FT0002 51.05904 51.05880 51.05941 191.1675 190.8787 191.5050      9   2   3  4
FT0003 51.98657 51.98631 51.98699 203.1467 202.6406 203.6710      7   0   3  4
FT0004 53.02036 53.02009 53.02043 203.2343 202.5652 204.0901     10   3   3  4
FT0005 53.52080 53.52051 53.52102 203.1936 202.8490 204.0901     10   3   3  4
FT0006 54.01007 54.00988 54.01015 159.2816 158.8499 159.4484      6   1   3  2
            peakidx ms_level
FT0001 7702, 16....        1
FT0002 7176, 16....        1
FT0003 7680, 17....        1
FT0004 7763, 17....        1
FT0005 8353, 17....        1
FT0006 5800, 15....        1</code></pre>
</div>
</div>
<p>This data frame provides the average <em>m/z</em> and retention time (in columns <code>"mzmed"</code> and <code>"rtmed"</code>) that characterize a LC-MS feature. Column, <code>"peakidx"</code> contains the indices of all chromatographic peaks that were assigned to that feature. The actual abundances for these features, which represent also the final preprocessing results, can be extracted with the <code><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-peak-grouping-results.html" class="external-link">featureValues()</a></code> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Extract feature abundances</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-peak-grouping-results.html" class="external-link">featureValues</a></span><span class="op">(</span><span class="va">lcms1</span>, method <span class="op">=</span> <span class="st">"sum"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       MS_QC_POOL_1_POS.mzML MS_A_POS.mzML MS_B_POS.mzML MS_QC_POOL_2_POS.mzML
FT0001              421.6162      689.2422            NA              481.7436
FT0002              710.8078      875.9192            NA              693.6997
FT0003              445.5711      613.4410            NA              497.8866
FT0004            16994.5260    24605.7340     19766.707            17808.0933
FT0005             3284.2664     4526.0531      3521.822             3379.8909
FT0006            10681.7476    10009.6602            NA            10800.5449
       MS_C_POS.mzML MS_D_POS.mzML MS_QC_POOL_3_POS.mzML MS_E_POS.mzML
FT0001            NA      635.2732              439.6086      570.5849
FT0002      781.2416      648.4344              700.9716     1054.0207
FT0003            NA      634.9370              449.0933            NA
FT0004    22780.6683    22873.1061            16965.7762    23432.1252
FT0005     4396.0762     4317.7734             3270.5290     4533.8667
FT0006            NA     7296.4262                    NA     9236.9799
       MS_F_POS.mzML MS_QC_POOL_4_POS.mzML
FT0001      579.9360              437.0340
FT0002      534.4577              711.0361
FT0003      461.0465              232.1075
FT0004    22198.4607            16796.4497
FT0005     4161.0132             3142.2268
FT0006     6817.8785                    NA</code></pre>
</div>
</div>
<p>We can note that some features (e.g. F0003 and F0006) have missing values in some samples. This is expected to a certain degree as not in all samples features, respectively their ions, need to be present. We will address this in the next section.</p>
</section><section class="level2"><h3 id="gap-filling">Gap filling<a class="anchor" aria-label="anchor" href="#gap-filling"></a>
</h3>
<p>The previously observed missing values (<em>NA</em>) could be attributed to various reasons. Although they might represent a genuinely missing value, indicating that an ion (feature) was truly not present in a particular sample, they could also be a result of a failure in the preceding chromatographic peak detection step. It is crucial to be able to recover missing values of the latter category as much as possible to reduce the eventual need for data imputation. We next examine how prevalent missing values are in our present dataset:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Percentage of missing values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-peak-grouping-results.html" class="external-link">featureValues</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-peak-grouping-results.html" class="external-link">featureValues</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 26.41597</code></pre>
</div>
</div>
<p>We can observe a substantial number of missing values values in our dataset. Let’s therefore delve into the process of <em>gap-filling</em>. We first evaluate some example features for which a chromatographic peak was only detected in some samples:</p>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ftidx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-peak-grouping-results.html" class="external-link">featureValues</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">featureDefinitions</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="va">ftidx</span><span class="op">]</span></span>
<span><span class="va">farea</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XcmsExperiment.html" class="external-link">featureArea</a></span><span class="op">(</span><span class="va">lcms1</span>, features <span class="op">=</span> <span class="va">fts</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">chromatogram</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span><span class="op">]</span>,</span>
<span>             rt <span class="op">=</span> <span class="va">farea</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rtmin"</span>, <span class="st">"rtmax"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>             mz <span class="op">=</span> <span class="va">farea</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mzmin"</span>, <span class="st">"mzmax"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span>col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"blue"</span><span class="op">)</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/unnamed-chunk-31-1.png" width="672"></p>
<figcaption>Figure 26. Examples of chromatographic peaks with missing values.</figcaption></figure>
</div>
</div>
</div>
<p>In both instances, a chromatographic peak was only identified in one of the two selected samples (red line), hence a missing value is reported for this feature in those particular samples (blue line). However, in both cases, signal was measured in both samples, thus, reporting a missing value would not be correct in this example. The signal for this feature is very low, which is the most likely reason peak detection failed. To rescue signal in such cases, the <code><a href="https://rdrr.io/pkg/xcms/man/fillChromPeaks.html" class="external-link">fillChromPeaks()</a></code> function can be used with the <code>ChromPeakAreaParam</code> approach. This method defines an <em>m/z</em>-retention time area for each feature based on detected peaks, where the signal for the respective ion is expected. It then integrates all intensities within this area in samples that have missing values for that feature. This is then reported as feature abundance. Below we apply this method using the default parameters.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Fill in the missing values in the whole dataset</span></span>
<span><span class="va">lcms1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/fillChromPeaks.html" class="external-link">fillChromPeaks</a></span><span class="op">(</span><span class="va">lcms1</span>, param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/fillChromPeaks.html" class="external-link">ChromPeakAreaParam</a></span><span class="op">(</span><span class="op">)</span>, chunkSize <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Percentage of missing values after gap-filling</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-peak-grouping-results.html" class="external-link">featureValues</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-peak-grouping-results.html" class="external-link">featureValues</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 5.155492</code></pre>
</div>
</div>
<p>With <code><a href="https://rdrr.io/pkg/xcms/man/fillChromPeaks.html" class="external-link">fillChromPeaks()</a></code> we could thus rescue most of the missing data in the data set. Note that, even if in a sample no ion would be present, in the worst case noise would be integrated, which is expected to be much lower than actual chromatographic peak signal. Let’s look at our previously missing values again:</p>
<div class="cell">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/unnamed-chunk-33-1.png" width="672"></p>
<figcaption>Figure 27. Examples of chromatographic peaks with missing values after gap-filling.</figcaption></figure>
</div>
</div>
</div>
<p>After gap-filling, also in the blue colored sample a chromatographic peak is present and its peak area would be reported as feature abundance for that sample.</p>
<p>To further assess the effectiveness of the gap-filling method for rescuing signals, we can also plot the average signal of features with at least one missing value against the average of filled-in signal. It is advisable to perform this analysis on repeatedly measured samples; in this case, our QC samples will be used.</p>
<p>For this, we extract:</p>
<ul>
<li><p>Feature values from detected chromatographic peaks by setting <code>filled = FALSE</code> in the <code>featuresValues()</code> call.</p></li>
<li><p>The filled-in signal by first extracting both detected and gap-filled abundances and then replace the values for detected chromatographic peaks with <code>NA</code>.</p></li>
</ul>
<p>Then, we calculate the row averages of both of these matrices and plot them against each other.</p>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Get only detected signal in QC samples</span></span>
<span><span class="va">vals_detect</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-peak-grouping-results.html" class="external-link">featureValues</a></span><span class="op">(</span><span class="va">lcms1</span>, filled <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">[</span>, <span class="va">QC_samples</span><span class="op">]</span></span>
<span></span>
<span><span class="co">#' Get detected and filled-in signal</span></span>
<span><span class="va">vals_filled</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-peak-grouping-results.html" class="external-link">featureValues</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">[</span>, <span class="va">QC_samples</span><span class="op">]</span></span>
<span></span>
<span><span class="co">#' Replace detected signal with NA</span></span>
<span><span class="va">vals_filled</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">vals_detect</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span></span>
<span><span class="co">#' Identify features with at least one filled peak</span></span>
<span><span class="va">has_filled</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">vals_detect</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Calculate row averages for features with missing values</span></span>
<span><span class="va">avg_detect</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowMeans.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">vals_detect</span><span class="op">[</span><span class="va">has_filled</span>, <span class="op">]</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">avg_filled</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowMeans.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">vals_filled</span><span class="op">[</span><span class="va">has_filled</span>, <span class="op">]</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Plot the values against each other (in log2 scale)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">avg_detect</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">avg_filled</span><span class="op">)</span>,</span>
<span>     xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">avg_detect</span>, <span class="va">avg_filled</span><span class="op">)</span><span class="op">)</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>     ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">avg_detect</span>, <span class="va">avg_filled</span><span class="op">)</span><span class="op">)</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>     pch <span class="op">=</span> <span class="fl">21</span>, bg <span class="op">=</span> <span class="st">"#00000020"</span>, col <span class="op">=</span> <span class="st">"#00000080"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/Detected-vs-filled-signal1-1.png" width="672"></p>
<figcaption>Figure 28. Detected vs. filled-in signal.</figcaption></figure>
</div>
</div>
</div>
<p>The detected (x-axis) and gap-filled (y-axis) values for QC samples are highly correlated. Especially for higher abundances, the agreement is very high, while for low intensities, as can be expected, differences are higher and trending to below the correlation line. Below we, in addition, fit a linear regression line to the data and summarize its results</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' fit a linear regression line to the data</span></span>
<span><span class="va">l</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html" class="external-link">lm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">avg_filled</span><span class="op">)</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">avg_detect</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">l</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = log2(avg_filled) ~ log2(avg_detect))

Residuals:
    Min      1Q  Median      3Q     Max
-6.8176 -0.3807  0.1725  0.5492  6.7504

Coefficients:
                 Estimate Std. Error t value Pr(&gt;|t|)
(Intercept)      -1.62359    0.11545  -14.06   &lt;2e-16 ***
log2(avg_detect)  1.11763    0.01259   88.75   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.9366 on 2846 degrees of freedom
  (846 observations deleted due to missingness)
Multiple R-squared:  0.7346,    Adjusted R-squared:  0.7345
F-statistic:  7877 on 1 and 2846 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
<p>The linear regression line has a slope of 1.12 and an intercept of -1.62. This indicates that the filled-in signal is on average 1.12 times higher than the detected signal.</p>
</section><section class="level2"><h3 id="filtering-features-missing-values">Filtering Features: Missing values<a class="anchor" aria-label="anchor" href="#filtering-features-missing-values"></a>
</h3>
<p>We now restrict the data set to features for which a chromatographic peak was detected in at least 2/3 of samples of at least one of the study samples groups. This ensures the statistical tests carried out later on the study samples are being performed on <em>reliable</em> signal. Also, with this filter we remove features that were mostly detected in QC samples, but not the study samples. Such filter can be performed with <code><a href="https://rdrr.io/pkg/ProtGenerics/man/filterFeatures.html" class="external-link">filterFeatures()</a></code> function from the <em><a href="https://bioconductor.org/packages/3.20/xcms" class="external-link">xcms</a></em> package with the <code>PercentMissingFilter</code> setting. The parameters of this filer:</p>
<ul>
<li>
<code>threshold</code>: defines the maximal acceptable percentage of samples with missing value(s) in at least one of the sample groups defined by parameter <code>f</code>.</li>
<li>
<code>f</code>: a factor defining the sample groups. By replacing the <code>"QC"</code> sample group with <code>NA</code> in parameter <code>f</code> we exclude the QC samples from the evaluation and consider only the study samples. With <code>threshold = 40</code> we keep only features for which a peak was detected in 2 out of the 3 samples in one of the sample groups.</li>
</ul>
<p>To consider detected chromatographic peaks per sample, we apply the filter on the<code>"raw"</code> assay of our result object, that contains abundance values only for detected chromatographic peaks (prior gap-filling).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Limit features to those with at least two detected peaks in one study group.</span></span>
<span><span class="co">#' Setting the value for QC samples to NA excludes QC samples from the</span></span>
<span><span class="co">#' calculation.</span></span>
<span><span class="va">f</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">$</span><span class="va">phenotype</span></span>
<span><span class="va">f</span><span class="op">[</span><span class="va">f</span> <span class="op">==</span> <span class="st">"QC"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span><span class="va">f</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">f</span><span class="op">)</span></span>
<span><span class="va">lcms1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/filterFeatures.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="va">lcms1</span>, <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/PercentMissingFilter.html" class="external-link">PercentMissingFilter</a></span><span class="op">(</span>f <span class="op">=</span> <span class="va">f</span>, threshold <span class="op">=</span> <span class="fl">40</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>344 features were removed</code></pre>
</div>
</div>
</section><section class="level2"><h3 id="preprocessing-results">Preprocessing results<a class="anchor" aria-label="anchor" href="#preprocessing-results"></a>
</h3>
<p>The final results of the LC-MS data preprocessing are stored within the <code>XcmsExperiment</code> object. This includes the identified chromatographic peaks, the alignment results, as well as the correspondence results. In addition, to guarantee reproducibility, this result object keeps track of all performed processing steps, including the individual parameter objects used to configure these. The <code><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">processHistory()</a></code> function returns a list of the various applied processing steps in chronological order. Below, we extract the information for the first step of the performed preprocessing.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Check first step of the process history</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">processHistory</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Object of class "XProcessHistory"
 type: Peak detection
 date: Wed Dec  4 10:43:03 2024
 info:
 fileIndex: 1,2,3,4,5,6,7,8,9,10
 Parameter class: CentWaveParam
 MS level(s) 1 </code></pre>
</div>
</div>
<p>The <code><a href="https://rdrr.io/pkg/xcms/man/ProcessHistory-class.html" class="external-link">processParam()</a></code> function could then be used to extract the actual parameter class used to configure this processing step.</p>
<p>The final result of the whole LC-MS data preprocessing is a two-dimensional matrix with abundances of the so-called LC-MS features in all samples. Note that at this stage of the analysis features are only characterized by their <em>m/z</em> and retention time and we don’t have yet any information which metabolite a feature could represent.</p>
<p>As we have seen before, such feature matrix can be extracted with the <code><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-peak-grouping-results.html" class="external-link">featureValues()</a></code> function and the corresponding feature characteristics (i.e., their <em>m/z</em> and retention time values) using the <code><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">featureDefinitions()</a></code> function. Thus, these two arrays could be extracted from the <em>xcms</em> result object and used/imported in other analysis packages for further processing. They could for example also be exported to tab delimited text files, and used in an external tool, or used, if also MS2 spectra were available, for feature-based molecular networking in the GNPS analysis environment <span class="citation" data-cites="nothias_feature-based_2020">(<a href="#ref-nothias_feature-based_2020" role="doc-biblioref">Nothias et al. 2020</a>)</span>.</p>
<p>For further processing in R, if no reference or link to the raw MS data is required, it is suggested to extract the <em>xcms</em> preprocessing result using the <code><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">quantify()</a></code> function as a <code>SummarizedExperiment</code> object, Bioconductor’s default container for data from biological assays/experiments. This simplifies integration with other Bioconductor analysis packages. The <code><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">quantify()</a></code> function takes the same parameters than the <code><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-peak-grouping-results.html" class="external-link">featureValues()</a></code> function, thus, with the call below we extract a <code>SummarizedExperiment</code> with only the detected, but not gap-filled, feature abundances:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Extract results as a SummarizedExperiment</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">quantify</a></span><span class="op">(</span><span class="va">lcms1</span>, method <span class="op">=</span> <span class="st">"sum"</span>, filled <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">res</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SummarizedExperiment
dim: 8724 10
metadata(7): '' '' ... '' ''
assays(1): raw
rownames(8724): FT0001 FT0002 ... FT9067 FT9068
rowData names(11): mzmed mzmin ... QC ms_level
colnames(10): MS_QC_POOL_1_POS.mzML MS_A_POS.mzML ... MS_F_POS.mzML
  MS_QC_POOL_4_POS.mzML
colData names(11): sample_name derived_spectra_data_file ... phenotype
  injection_index</code></pre>
</div>
</div>
<p>Sample identifications from the <em>xcms</em> result’s <code><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData()</a></code> are now available as <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> (column, sample annotations) and the <code><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">featureDefinitions()</a></code> as <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code> (row, feature annotations). The feature values have been added as the first <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay()</a></code> in the <code>SummarizedExperiment</code> and even the processing history is available in the object’s <code><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata()</a></code>. A <code>SummarizedExperiment</code> supports multiple assays, all being numeric matrices with the same dimensions. Below we thus add the detected and gap-filled feature abundances as an additional assay to the <code>SummarizedExperiment</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">$</span><span class="va">raw_filled</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-peak-grouping-results.html" class="external-link">featureValues</a></span><span class="op">(</span><span class="va">lcms1</span>, method <span class="op">=</span> <span class="st">"sum"</span>,</span>
<span>                                        filled <span class="op">=</span> <span class="cn">TRUE</span> <span class="op">)</span></span>
<span></span>
<span><span class="co">#' Different assay in the SummarizedExperiment object</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assayNames</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "raw"        "raw_filled"</code></pre>
</div>
</div>
<p>Feature abundances can be extracted with the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay()</a></code> function. Below we extract the first 6 lines of the detected and gap-filled feature abundances:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"raw_filled"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       MS_QC_POOL_1_POS.mzML MS_A_POS.mzML MS_B_POS.mzML MS_QC_POOL_2_POS.mzML
FT0001              421.6162      689.2422      411.3295              481.7436
FT0002              710.8078      875.9192      457.5920              693.6997
FT0003              445.5711      613.4410      277.5022              497.8866
FT0004            16994.5260    24605.7340    19766.7069            17808.0933
FT0005             3284.2664     4526.0531     3521.8221             3379.8909
FT0006            10681.7476    10009.6602     9599.9701            10800.5449
       MS_C_POS.mzML MS_D_POS.mzML MS_QC_POOL_3_POS.mzML MS_E_POS.mzML
FT0001      314.7567      635.2732              439.6086      570.5849
FT0002      781.2416      648.4344              700.9716     1054.0207
FT0003      425.3774      634.9370              449.0933      556.2544
FT0004    22780.6683    22873.1061            16965.7762    23432.1252
FT0005     4396.0762     4317.7734             3270.5290     4533.8667
FT0006     4792.2390     7296.4262             2382.1788     9236.9799
       MS_F_POS.mzML MS_QC_POOL_4_POS.mzML
FT0001      579.9360              437.0340
FT0002      534.4577              711.0361
FT0003      461.0465              232.1075
FT0004    22198.4607            16796.4497
FT0005     4161.0132             3142.2268
FT0006     6817.8785             6911.5439</code></pre>
</div>
</div>
<p>An advantage, in addition to being a container for the full preprocessing results is also the possibility of an easy and intuitive creation of data subsets ensuring data integrity. It would for example be very easy to subset the full data to a selection of features and/or samples:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">14</span>, <span class="fl">3</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SummarizedExperiment
dim: 14 6
metadata(7): '' '' ... '' ''
assays(2): raw raw_filled
rownames(14): FT0001 FT0002 ... FT0013 FT0014
rowData names(11): mzmed mzmin ... QC ms_level
colnames(6): MS_B_POS.mzML MS_QC_POOL_2_POS.mzML ...
  MS_QC_POOL_3_POS.mzML MS_E_POS.mzML
colData names(11): sample_name derived_spectra_data_file ... phenotype
  injection_index</code></pre>
</div>
</div>
<p>Before moving to the next step of the analysis, it is advisable to save the preprocessing results. We have multiple format options to save into, and they can be found in the <code>MsIO</code> package <a href="https://rformassspectrometry.github.io/MsIO/articles/MsIO.html" class="external-link">documentation</a>. Below we will save our <code>XcmsExperiment</code> object into a file format handled by the <a href="https://github.com/ArtifactDB/alabaster.base" class="external-link">alabster framework</a>, which ensures that our object can be easily read from other languages like Python and Javascript as well as loaded easily back into R.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' XcmsExperiment object: </span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MsIO/man/saveMsObject.html" class="external-link">saveMsObject</a></span><span class="op">(</span><span class="va">lcms1</span>,</span>
<span>             <span class="fu"><a href="https://rdrr.io/pkg/MsIO/man/AlabasterParam.html" class="external-link">AlabasterParam</a></span><span class="op">(</span>path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"objects/preprocessed_lcms1"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Below we remove the processHistory of the res object to allow for export, </span></span>
<span><span class="co">#' this need to be fixed later.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/alabaster.base/man/saveObject.html" class="external-link">saveObject</a></span><span class="op">(</span><span class="va">res</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"objects/preprocessed_res"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</section></section><section class="section level2"><h2 id="data-normalization">Data normalization<a class="anchor" aria-label="anchor" href="#data-normalization"></a>
</h2>
<p>After preprocessing, data normalization or scaling might need to be applied to remove any technical variances from the data. While simple approaches like median scaling can be implemented with a few lines of R code, more advanced normalization algorithms are available in packages such as Bioconductor’s <em><a href="https://bioconductor.org/packages/3.20/preprocessCore" class="external-link">preprocessCore</a></em>. The comprehensive workflow “Notame” also propose a very interesting normalization approach adaptable and scalable to the user dataset <span class="citation" data-cites="klavus_notame_2020">(<a href="#ref-klavus_notame_2020" role="doc-biblioref">Klåvus et al. 2020</a>)</span>.</p>
<p>Generally, for LC-MS data, bias can be categorized into three main groups<span class="citation" data-cites="broadhurst_guidelines_2018">(<a href="#ref-broadhurst_guidelines_2018" role="doc-biblioref">Broadhurst et al. 2018</a>)</span>:</p>
<ul>
<li><p>Variances introduced by sample collection and initial processing, which can include differences in sample amounts. This type of bias is expected to be sample-specific and affect all signals in a sample in the same way. Methods like median scaling, LOESS or quantiles normalization can adjust this bias.</p></li>
<li><p>Signal drifts along measurement of samples from an experiment. Reasons for such drifts can be related to aging of the instrumentation used (columns, detector), but also to changes in metabolite abundances and characteristics due to reactions and modifications, such as oxidation. These changes are expected to affect more the samples measured later in a run rather than the ones measured at the beginning. For this reason, this bias can play a major role in large experiments bias can play a major role in large experiments measured over a long time range and is usually considered to affect individual metabolites (or metabolite groups) differently. For adjustment, moving average or linear regression-based approaches can be used. The latter can for example be performed using the <code><a href="https://rdrr.io/pkg/MetaboCoreUtils/man/fit_lm.html" class="external-link">adjust_lm()</a></code> function from the <em><a href="https://bioconductor.org/packages/3.20/MetaboCoreUtils" class="external-link">MetaboCoreUtils</a></em> package.</p></li>
<li><p>Batch-related biases. These comprise any noise that is specific to a larger set of samples, which can be the set of samples measured in one LC-MS measurement run (i.e. from one analysis plate) or samples measured using a specific batch of reagents. This noise is assumed to affect all samples in one batch in the same way and linear modeling-based approaches can be used to adjust for this.</p></li>
</ul>
<p>Unwanted variation can arise from various sources and is highly dependent on the experiment. Therefore, data normalization should be chosen carefully based on experimental design, statistical aims, and the balance of accuracy and precision achieved through the use of auxiliary information.</p>
<p>Sample preparation biases can be evaluated using internal standards, depending however also on when they were added to the sample mixes during sample processing. Repeated measurements of QC samples on the other hand allows to estimate and correct for LC-MS specific biases. Also, proper planning of an experiment, such as measurement of study samples in random order, can largely avoid biases introduced by most of the above mentioned sources of variance.</p>
<p>In this workflow we present some tools to assess data quality and evaluate need for normalization as well as options for normalization. For space reasons we are not able to provide solutions to adjust for all possible sources of variation.</p>
<section class="level2"><h3 id="initial-quality-assessment">Initial quality assessment<a class="anchor" aria-label="anchor" href="#initial-quality-assessment"></a>
</h3>
<p>A principal component analysis (PCA) is a very helpful tool for an initial, unsupervised, visualization of the data that also provides insights into potential quality issues in the data. In order to apply a PCA to the measured feature abundances, we need however to impute (the still present) missing values. We assume that most of these missing values (after the gap-filling step) represent signal which is below detection limit. In such cases, missing values can be replaced with random values sampled from a uniform distribution, ranging from half of the smallest measured value to the smallest measured value for a specific feature. The uniform distribution is defined with two parameters (minimum and maximum) and all values between them have an equal probability of being selected.</p>
<p>Below we impute missing values with this approach and add the resulting data matrix as a new <em>assay</em> to our result object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Impute missing values using an uniform distribution</span></span>
<span><span class="va">na_unidis</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">z</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">na</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">z</span><span class="op">)</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/any.html" class="external-link">any</a></span><span class="op">(</span><span class="va">na</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">min</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">z</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>        <span class="va">z</span><span class="op">[</span><span class="va">na</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">na</span><span class="op">)</span>, min <span class="op">=</span> <span class="va">min</span><span class="op">/</span><span class="fl">2</span>, max <span class="op">=</span> <span class="va">min</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">z</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">#' Row-wise impute missing values and add the data as a new assay</span></span>
<span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"raw_filled"</span><span class="op">)</span>, MARGIN <span class="op">=</span> <span class="fl">1</span>, <span class="va">na_unidis</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">$</span><span class="va">raw_filled_imputed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">tmp</span><span class="op">)</span></span></code></pre></div>
</div>
<section class="level3"><h4 id="principal-component-analysis">Principal Component Analysis<a class="anchor" aria-label="anchor" href="#principal-component-analysis"></a>
</h4>
<p>PCA is a powerful tool for detecting biases in data. As a dimensionality reduction technique, it enables visualization of data in a lower-dimensional space. In the context of LC-MS data, PCA can be used to identify overall biases in batch, sample, injection index, etc. However, it is important to note that PCA is a linear method and may not be able to detect all biases in the data.</p>
<p>Before plotting the PCA, we apply a log2 transform, center and scale the data. The log2 transformation is applied to stabilize the variance while centering to remove dependency on absolute abundances.</p>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Log2 transform and scale data</span></span>
<span><span class="va">vals</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"raw_filled_imputed"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span>center <span class="op">=</span> <span class="cn">TRUE</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Perform the PCA</span></span>
<span><span class="va">pca_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="va">vals</span>, scale <span class="op">=</span> <span class="cn">FALSE</span>, center <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Plot the results</span></span>
<span><span class="va">vals_st</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">vals</span>, phenotype <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">phenotype</span><span class="op">)</span></span>
<span><span class="va">pca_12</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">pca_res</span>, data <span class="op">=</span> <span class="va">vals_st</span> , colour <span class="op">=</span> <span class="st">'phenotype'</span>, scale <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">col_phenotype</span><span class="op">)</span></span>
<span><span class="va">pca_34</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">pca_res</span>, data <span class="op">=</span> <span class="va">vals_st</span>, colour <span class="op">=</span> <span class="st">'phenotype'</span>,</span>
<span>                   x <span class="op">=</span> <span class="fl">3</span>, y <span class="op">=</span> <span class="fl">4</span>, scale <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">col_phenotype</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gridExtra/man/arrangeGrob.html" class="external-link">grid.arrange</a></span><span class="op">(</span><span class="va">pca_12</span>, <span class="va">pca_34</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/unsupervised-checks1-1.png" width="576"></p>
<figcaption>Figure 28. PCA of the data coloured by phenotypes.</figcaption></figure>
</div>
</div>
</div>
<p>The PCA above shows a clear separation of the study samples (plasma) from the QC samples (serum) on the first principal component (PC1). The separation based on phenotype is visible on the third principal component (PC3).</p>
<p>In some cases, it can be a better option to remove the imputed values and evaluate the PCA again. This is especially true if the imputed values are replacing a large proportion of the data.</p>
</section><section class="level3"><h4 id="intensity-evaluation">Intensity evaluation<a class="anchor" aria-label="anchor" href="#intensity-evaluation"></a>
</h4>
<p>Global differences in feature abundances between samples (e.g. due to sample-specific biases) can be evaluated by plotting the distribution of the log2 transformed feature abundances using boxplots of violin plots. Below we show the number of detected chromatographic peaks per sample and the distribution of the log2 transformed feature abundances.</p>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/layout.html" class="external-link">layout</a></span><span class="op">(</span>mat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.8</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">4.5</span>, <span class="fl">0.2</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html" class="external-link">barplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"raw"</span><span class="op">)</span>, MARGIN <span class="op">=</span> <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="fl">80</span><span class="op">)</span>, border <span class="op">=</span> <span class="va">col_sample</span>,</span>
<span>        ylab <span class="op">=</span> <span class="st">"# detected peaks"</span>, xaxt <span class="op">=</span> <span class="st">"n"</span>, space <span class="op">=</span> <span class="fl">0.012</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span>nx <span class="op">=</span> <span class="cn">NA</span>, ny <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html" class="external-link">barplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"raw_filled"</span><span class="op">)</span>, MARGIN <span class="op">=</span> <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="fl">80</span><span class="op">)</span>, border <span class="op">=</span> <span class="va">col_sample</span>,</span>
<span>        ylab <span class="op">=</span> <span class="st">"# detected + filled peaks"</span>, xaxt <span class="op">=</span> <span class="st">"n"</span>,</span>
<span>        space <span class="op">=</span> <span class="fl">0.012</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span>nx <span class="op">=</span> <span class="cn">NA</span>, ny <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/vioplot/man/vioplot.html" class="external-link">vioplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"raw_filled"</span><span class="op">)</span><span class="op">)</span>, xaxt <span class="op">=</span> <span class="st">"n"</span>,</span>
<span>        ylab <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="va">log</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">~</span><span class="va">feature</span><span class="op">~</span><span class="va">abundance</span><span class="op">)</span>,</span>
<span>        col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="fl">80</span><span class="op">)</span>, border <span class="op">=</span> <span class="va">col_sample</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/points.html" class="external-link">points</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/matrixStats/man/rowMedians.html" class="external-link">colMedians</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"raw_filled"</span><span class="op">)</span><span class="op">)</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"b"</span>,</span>
<span>       pch <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span>nx <span class="op">=</span> <span class="cn">NA</span>, ny <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, col <span class="op">=</span> <span class="va">col_phenotype</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">col_phenotype</span><span class="op">)</span>, lty<span class="op">=</span><span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">2</span>, xpd <span class="op">=</span> <span class="cn">TRUE</span>, ncol <span class="op">=</span> <span class="fl">3</span>,</span>
<span>       cex <span class="op">=</span> <span class="fl">0.8</span>,  bty <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/counts1-1.png" width="672"></p>
<figcaption>Figure 29. Number of detected peaks and feature abundances.</figcaption></figure>
</div>
</div>
</div>
<p>The upper part of the plot show that the gap filling steps allowed to rescue a substantial number of <em>NAs</em> and allowed us to have a more consistent number of feature values per sample. This consistency aligns with our asspumption that every sample should have a similar amount of features detected. Additionally we observe that, on average, the signal distribution from the individual samples is very similar.</p>
<p>An alternative way to evaluate differences in abundances between samples are <em>relative log abundance</em> (RLA) plots <span class="citation" data-cites="de_livera_normalizing_2012">(<a href="#ref-de_livera_normalizing_2012" role="doc-biblioref">De Livera et al. 2012</a>)</span>. An RLA value is the abundance of a feature in a sample relative to the median abundance of the same feature across multiple samples. We can discriminate between <em>within group</em> and <em>across group</em> RLAs, depending whether the abundance is compared against samples within the same sample group or across all samples. Within group RLA plots assess the tightness of replicates within groups and should have a median close to zero and low variation around it. When used across groups, they allow to compare behavior between groups. Generally, between-sample differences can be easily spotted using RLA plots. Below we calculate and visualize within group RLA values using the <code><a href="https://rdrr.io/pkg/xcms/man/rla.html" class="external-link">rowRla()</a></code> function from the <em><a href="https://bioconductor.org/packages/3.20/MsCoreUtils" class="external-link">MsCoreUtils</a></em> package defining with parameter <code>f</code> the sample groups.</p>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3.5</span>, <span class="fl">4.5</span>, <span class="fl">2.5</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span><span class="fu">MsCoreUtils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MsCoreUtils/man/rla.html" class="external-link">rowRla</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"raw_filled"</span><span class="op">)</span>,</span>
<span>                            f <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">phenotype</span>, transform <span class="op">=</span> <span class="st">"log2"</span><span class="op">)</span>,</span>
<span>        cex <span class="op">=</span> <span class="fl">0.5</span>, pch <span class="op">=</span> <span class="fl">16</span>,</span>
<span>        col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="fl">80</span><span class="op">)</span>, ylab <span class="op">=</span> <span class="st">"RLA"</span>,</span>
<span>        border <span class="op">=</span> <span class="va">col_sample</span>, boxwex <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        outline <span class="op">=</span> <span class="cn">FALSE</span>, xaxt <span class="op">=</span> <span class="st">"n"</span>, main <span class="op">=</span> <span class="st">"Relative log abundance"</span>,</span>
<span>        cex.main <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/axis.html" class="external-link">axis</a></span><span class="op">(</span>side <span class="op">=</span> <span class="fl">1</span>, at <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">)</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">$</span><span class="va">sample_name</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span>nx <span class="op">=</span> <span class="cn">NA</span>, ny <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>h <span class="op">=</span> <span class="fl">0</span>, lty<span class="op">=</span><span class="fl">3</span>, lwd <span class="op">=</span> <span class="fl">1</span>, col <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, col <span class="op">=</span> <span class="va">col_phenotype</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">col_phenotype</span><span class="op">)</span>, lty<span class="op">=</span><span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">2</span>, xpd <span class="op">=</span> <span class="cn">TRUE</span>, ncol <span class="op">=</span> <span class="fl">3</span>,</span>
<span>       cex <span class="op">=</span> <span class="fl">0.8</span>,  bty <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/rla-plot-raw-and-filled1-1.png" width="672"></p>
<figcaption>Figure 30. RLA plot for the raw data and filled data.</figcaption></figure>
</div>
</div>
</div>
<p>On the RLA plot above, we can observe that the medians for most samples are indeed centered around 0. Exception are two of the <em>CVD</em> samples. Thus, while the distribution of signals across samples is comparable, some differences seem to be present that require a between sample normalization.</p>
</section></section><section class="level2"><h3 id="between-sample-normalisation">Between sample normalisation<a class="anchor" aria-label="anchor" href="#between-sample-normalisation"></a>
</h3>
<p>The previous RLA plot showed that the data had some biases that need to be corrected. Therefore, we will implement between-sample normalization using filled-in features. This process effectively mitigates variations influenced by technical issues, such as differences in sample preparation, processing and injection methods. In this instance, we will employ a commonly used technique known as median scaling <span class="citation" data-cites="de_livera_normalizing_2012">(<a href="#ref-de_livera_normalizing_2012" role="doc-biblioref">De Livera et al. 2012</a>)</span>.</p>
<section class="level3"><h4 id="median-scaling">Median scaling<a class="anchor" aria-label="anchor" href="#median-scaling"></a>
</h4>
<p>This method involves computing the median for each sample, followed by determining the median of these individual sample medians. This ensures consistent median values for each sample throughout the entire data set. Maintaining uniformity in the average total metabolite abundance across all samples is crucial for effective implementation.</p>
<p>This process aims to establish a shared baseline for the central tendency of metabolite abundance, mitigating the impact of sample-specific technical variations. This approach fosters a more robust and comparable analysis of the top features across the data set. The assumption is that normalizing based on the median, known for its lower sensitivity to extreme values, enhances the comparability of top features and ensures a consistent average abundance across samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Compute median and generate normalization factor</span></span>
<span><span class="va">mdns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"raw_filled"</span><span class="op">)</span>, MARGIN <span class="op">=</span> <span class="fl">2</span>,</span>
<span>              <span class="va">median</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span> <span class="op">)</span></span>
<span><span class="va">nf_mdn</span> <span class="op">&lt;-</span> <span class="va">mdns</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">mdns</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' divide dataset by median of median and create a new assay.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">$</span><span class="va">norm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sweep.html" class="external-link">sweep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"raw_filled"</span><span class="op">)</span>, MARGIN <span class="op">=</span> <span class="fl">2</span>, <span class="va">nf_mdn</span>, <span class="st">'/'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">$</span><span class="va">norm_imputed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sweep.html" class="external-link">sweep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"raw_filled_imputed"</span><span class="op">)</span>, MARGIN <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                                  <span class="va">nf_mdn</span>, <span class="st">'/'</span><span class="op">)</span></span></code></pre></div>
</div>
<p>The median scaling is calculated for both imputed and non-imputed data, with each set stored separately within the <code>SummarizedExperiment</code> object. This approach facilitates testing various normalization strategies while maintaining a record of all processing steps undertaken, enabling easy regression to previous stages if necessary.</p>
</section></section><section class="level2"><h3 id="assessing-overall-effectiveness-of-the-normalization-approach">Assessing overall effectiveness of the normalization approach<a class="anchor" aria-label="anchor" href="#assessing-overall-effectiveness-of-the-normalization-approach"></a>
</h3>
<p>It is crucial to evaluate the effectiveness of the normalization process. This can be achieved by comparing the distribution of the log2 transformed feature abundances before and after normalization. Additionally, the RLA plots can be used to assess the tightness of replicates within groups and compare the behavior between groups.</p>
<section class="level3"><h4 id="principal-component-analysis-1">Principal Component Analysis<a class="anchor" aria-label="anchor" href="#principal-component-analysis-1"></a>
</h4>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Data before normalization</span></span>
<span><span class="va">vals_st</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">vals</span>, phenotype <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">phenotype</span><span class="op">)</span></span>
<span><span class="va">pca_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">pca_res</span>, data <span class="op">=</span> <span class="va">vals_st</span>,</span>
<span>                    colour <span class="op">=</span> <span class="st">'phenotype'</span>, scale <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">col_phenotype</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Data after normalization</span></span>
<span><span class="va">vals_norm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"norm"</span><span class="op">)</span>, MARGIN <span class="op">=</span> <span class="fl">1</span>, <span class="va">na_unidis</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span>center <span class="op">=</span> <span class="cn">TRUE</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pca_res_norm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="va">vals_norm</span>, scale <span class="op">=</span> <span class="cn">FALSE</span>, center <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">vals_st_norm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">vals_norm</span>, phenotype <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">phenotype</span><span class="op">)</span></span>
<span><span class="va">pca_adj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">pca_res_norm</span>, data <span class="op">=</span> <span class="va">vals_st_norm</span>,</span>
<span>                    colour <span class="op">=</span> <span class="st">'phenotype'</span>, scale <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">col_phenotype</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gridExtra/man/arrangeGrob.html" class="external-link">grid.arrange</a></span><span class="op">(</span><span class="va">pca_raw</span>, <span class="va">pca_adj</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/unnamed-chunk-43-1.png" width="576"></p>
<figcaption>Figure 31. PC1 and PC2 of the data before and after normalization.</figcaption></figure>
</div>
</div>
</div>
<p>Normalization did not have a large impact on PC1 or PC2, but the separation of the study groups on PC3 seems to be better and difference between QC samples lower after normalization (see below).</p>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">pca_res</span>, data <span class="op">=</span> <span class="va">vals_st</span> ,</span>
<span>                    colour <span class="op">=</span> <span class="st">'phenotype'</span>, x <span class="op">=</span> <span class="fl">3</span>, y <span class="op">=</span> <span class="fl">4</span>, scale <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">col_phenotype</span><span class="op">)</span></span>
<span><span class="va">pca_adj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">pca_res_norm</span>, data <span class="op">=</span> <span class="va">vals_st_norm</span>,</span>
<span>                    colour <span class="op">=</span> <span class="st">'phenotype'</span>, x <span class="op">=</span> <span class="fl">3</span>, y <span class="op">=</span> <span class="fl">4</span>, scale <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">col_phenotype</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/gridExtra/man/arrangeGrob.html" class="external-link">grid.arrange</a></span><span class="op">(</span><span class="va">pca_raw</span>, <span class="va">pca_adj</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/unnamed-chunk-44-1.png" width="576"></p>
<figcaption>Figure 32. PC3 and PC4 of the data before and after normalization.</figcaption></figure>
</div>
</div>
</div>
<p>The PCA plots above show that the normalization process has not changed the overall structure of the data. The separation between the study and QC samples remains the same. This is an expected results as normalization should not correct for biological variance and only technical.</p>
</section><section class="level3"><h4 id="intensity-evaluation-1">Intensity evaluation<a class="anchor" aria-label="anchor" href="#intensity-evaluation-1"></a>
</h4>
<p>We compare the RLA plots before and after between-sample normalization to evaluate its impact on the data.</p>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span>, mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3.5</span>, <span class="fl">4.5</span>, <span class="fl">2.5</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/rla.html" class="external-link">rowRla</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"raw_filled"</span><span class="op">)</span>, group <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">phenotype</span><span class="op">)</span>,</span>
<span>        cex <span class="op">=</span> <span class="fl">0.5</span>, pch <span class="op">=</span> <span class="fl">16</span>, ylab <span class="op">=</span> <span class="st">"RLA"</span>, border <span class="op">=</span> <span class="va">col_sample</span>,</span>
<span>        col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="fl">80</span><span class="op">)</span>, cex.main <span class="op">=</span> <span class="fl">1</span>, outline <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>        xaxt <span class="op">=</span> <span class="st">"n"</span>, main <span class="op">=</span> <span class="st">"Raw data"</span>, boxwex <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span>nx <span class="op">=</span> <span class="cn">NA</span>, ny <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, inset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="op">-</span><span class="fl">0.2</span><span class="op">)</span>, col <span class="op">=</span> <span class="va">col_phenotype</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">col_phenotype</span><span class="op">)</span>, lty<span class="op">=</span><span class="fl">1</span>, lwd <span class="op">=</span> <span class="fl">2</span>, xpd <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>       ncol <span class="op">=</span> <span class="fl">3</span>, cex <span class="op">=</span> <span class="fl">0.7</span>, bty <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>h <span class="op">=</span> <span class="fl">0</span>, lty<span class="op">=</span><span class="fl">3</span>, lwd <span class="op">=</span> <span class="fl">1</span>, col <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/rla.html" class="external-link">rowRla</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"norm"</span><span class="op">)</span>, group <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">phenotype</span><span class="op">)</span>,</span>
<span>        cex <span class="op">=</span> <span class="fl">0.5</span>, pch <span class="op">=</span> <span class="fl">16</span>, ylab <span class="op">=</span> <span class="st">"RLA"</span>, border <span class="op">=</span> <span class="va">col_sample</span>,</span>
<span>        col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span>, <span class="fl">80</span><span class="op">)</span>, boxwex <span class="op">=</span> <span class="fl">1</span>, outline <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>        xaxt <span class="op">=</span> <span class="st">"n"</span>, main <span class="op">=</span> <span class="st">"Normallized data"</span>, cex.main <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/axis.html" class="external-link">axis</a></span><span class="op">(</span>side <span class="op">=</span> <span class="fl">1</span>, at <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">)</span>, labels <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">sample_name</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span>nx <span class="op">=</span> <span class="cn">NA</span>, ny <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>h <span class="op">=</span> <span class="fl">0</span>, lty <span class="op">=</span> <span class="fl">3</span>, lwd <span class="op">=</span> <span class="fl">1</span>, col <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/rla-plot-after-norm2-1.png" width="576"></p>
<figcaption>Figure 33. RLA plot before and after normalization.</figcaption></figure>
</div>
</div>
</div>
<p>The normalization process has effectively centered the data around the median and medians for all samples are now closer to zero.</p>
</section><section class="level3"><h4 id="coefficient-of-variation">Coefficient of variation<a class="anchor" aria-label="anchor" href="#coefficient-of-variation"></a>
</h4>
<p>We next evaluate the coefficient of variation (CV, also referred to as <em>relative standard deviation</em> RSD) for features across samples from either QC or study samples. In QC samples, the CV would represent technical noise, while in study samples it would include also expected biological differences. Thus, normalization should reduce the CV in QC samples, while only slightly reducing the CV in study samples. The CV is calculated below using the <code><a href="https://rdrr.io/pkg/MetaboCoreUtils/man/quality_assessment.html" class="external-link">rowRsd()</a></code> function from the <em><a href="https://bioconductor.org/packages/3.20/MetaboCoreUtils" class="external-link">MetaboCoreUtils</a></em> package. By setting <code>mad = TRUE</code> we use a more robust calculation using the median absolute deviation instead of the standard deviation.</p>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb101"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Calculate the CV values</span></span>
<span><span class="va">index_study</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">phenotype</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CTR"</span>, <span class="st">"CVD"</span><span class="op">)</span></span>
<span><span class="va">index_QC</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">phenotype</span> <span class="op">==</span> <span class="st">"QC"</span></span>
<span></span>
<span><span class="va">sample_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span></span>
<span>    QC_raw <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MetaboCoreUtils/man/quality_assessment.html" class="external-link">rowRsd</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"raw_filled"</span><span class="op">)</span><span class="op">[</span>, <span class="va">index_QC</span><span class="op">]</span>,</span>
<span>                    na.rm <span class="op">=</span> <span class="cn">TRUE</span>, mad <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    QC_norm <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MetaboCoreUtils/man/quality_assessment.html" class="external-link">rowRsd</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"norm"</span><span class="op">)</span><span class="op">[</span>, <span class="va">index_QC</span><span class="op">]</span>,</span>
<span>                     na.rm <span class="op">=</span> <span class="cn">TRUE</span>, mad <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    Study_raw <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MetaboCoreUtils/man/quality_assessment.html" class="external-link">rowRsd</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"raw_filled"</span><span class="op">)</span><span class="op">[</span>, <span class="va">index_study</span><span class="op">]</span>,</span>
<span>                       na.rm <span class="op">=</span> <span class="cn">TRUE</span>, mad <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    Study_norm <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MetaboCoreUtils/man/quality_assessment.html" class="external-link">rowRsd</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"norm"</span><span class="op">)</span><span class="op">[</span>, <span class="va">index_study</span><span class="op">]</span>,</span>
<span>                        na.rm <span class="op">=</span> <span class="cn">TRUE</span>, mad <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Summarize the values across features</span></span>
<span><span class="va">res_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    QC_raw <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">sample_res</span><span class="op">[</span>, <span class="st">"QC_raw"</span><span class="op">]</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    QC_norm <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">sample_res</span><span class="op">[</span>, <span class="st">"QC_norm"</span><span class="op">]</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    Study_raw <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">sample_res</span><span class="op">[</span>, <span class="st">"Study_raw"</span><span class="op">]</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    Study_norm <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">sample_res</span><span class="op">[</span>, <span class="st">"Study_norm"</span><span class="op">]</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">kable</span><span class="op">(</span><span class="va">res_df</span>, format <span class="op">=</span> <span class="st">"pipe"</span><span class="op">)</span></span></code></pre></div>
</details><table class="table">
<caption>Table 6. Distribution of CV values across samples for the raw and normalized data.</caption>
<thead><tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">QC_raw</th>
<th style="text-align: right;">QC_norm</th>
<th style="text-align: right;">Study_raw</th>
<th style="text-align: right;">Study_norm</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">0%</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0008024</td>
<td style="text-align: right;">0.0027477</td>
</tr>
<tr class="even">
<td style="text-align: left;">25%</td>
<td style="text-align: right;">0.0450743</td>
<td style="text-align: right;">0.0446309</td>
<td style="text-align: right;">0.1423902</td>
<td style="text-align: right;">0.1430977</td>
</tr>
<tr class="odd">
<td style="text-align: left;">50%</td>
<td style="text-align: right;">0.0972048</td>
<td style="text-align: right;">0.0970785</td>
<td style="text-align: right;">0.2668930</td>
<td style="text-align: right;">0.2658062</td>
</tr>
<tr class="even">
<td style="text-align: left;">75%</td>
<td style="text-align: right;">0.2003657</td>
<td style="text-align: right;">0.1985980</td>
<td style="text-align: right;">0.4867303</td>
<td style="text-align: right;">0.4825679</td>
</tr>
<tr class="odd">
<td style="text-align: left;">100%</td>
<td style="text-align: right;">1.4635996</td>
<td style="text-align: right;">1.4637193</td>
<td style="text-align: right;">1.4825882</td>
<td style="text-align: right;">1.4825893</td>
</tr>
</tbody>
</table>
</div>
<p>The table above shows the distribution of the CV for both raw and normalized data. The first column highlights the % of the data which is below a given CV value, e.g. 25% of the data has a CV equal or lower than 0.04557 at the <em>QC_raw</em> data. As anticipated, the CV values for the QCs, which reflect technical variance, are lower compared to those for the study samples, which include both technical and biological variance. Overall, minimal disparity exists between the raw and normalized data, which is a positive indication that the normalization process has not introduced bias into the dataset, but also reflects the little differences in average abundances between sample in the raw data.</p>
</section><section class="level3"><h4 id="conclusion-on-normalization">Conclusion on normalization<a class="anchor" aria-label="anchor" href="#conclusion-on-normalization"></a>
</h4>
<p>The overall conclusion of the normalization process is that very little variance was present from the beginning, normalization was however able to center the data around the median (as shown by the RLA plot). Given the simplicity and limited size of our example dataset, we will conclude the normalization process at this stage. For more intricate datasets with diverse biases, a tailored approach would be devised. This could include also approaches to adjust for signal drifts or batch effects. One possible option would be to use a linear-model based approach such as can for example be applied with the <code><a href="https://rdrr.io/pkg/MetaboCoreUtils/man/fit_lm.html" class="external-link">adjust_lm()</a></code> function from the <em><a href="https://bioconductor.org/packages/3.20/MetaboCoreUtils" class="external-link">MetaboCoreUtils</a></em> package.</p>
</section></section></section><section class="section level2"><h2 id="quality-control-feature-prefiltering">Quality control: Feature prefiltering<a class="anchor" aria-label="anchor" href="#quality-control-feature-prefiltering"></a>
</h2>
<p>After normalizing our data we can now pre-filter to clean the data before performing any statistical analysis. In general, pre-filtering of samples and features is performed to remove outliers.</p>
<p>Below we copy the original result object to also keep the unfiltered data for later comparisons.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Number of features before filtering</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 8724</code></pre>
</div>
<div class="sourceCode cell-code" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' keep unfiltered object</span></span>
<span><span class="va">res_unfilt</span> <span class="op">&lt;-</span> <span class="va">res</span></span></code></pre></div>
</div>
<p>Here we will eliminate features that exhibit high variability in our dataset. Repeatedly measured QC samples typically serve as a robust basis for cleansing datasets allowing to identify features with excessively high noise. As in our data set external QC samples were used, i.e. pooled samples from a different collection and using a slightly different sample matrix, their utility for filtering is somewhat limited. For a comprehensive description and guidelines for data filtering in untargeted metabolomic studies, please refer to <span class="citation" data-cites="broadhurst_guidelines_2018">(<a href="#ref-broadhurst_guidelines_2018" role="doc-biblioref">Broadhurst et al. 2018</a>)</span>.</p>
<p>Following the guidelines stated above we decided to still use the QC samples for pre-filtering, on the basis that they represent similar bio-fluids to our study samples, and thus, we anticipate observing relatively similar metabolites affected by similar measurement biases.</p>
<p>We therefore evaluate the <em>dispersion ratio</em> (Dratio) <span class="citation" data-cites="broadhurst_guidelines_2018">(<a href="#ref-broadhurst_guidelines_2018" role="doc-biblioref">Broadhurst et al. 2018</a>)</span> for all features in the data set. We accomplish this task using the same function as above but this time with the <code>DratioFilter</code> parameter. More filters exist for this function and we invite the user to explore them as to decide what is best for their dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Compute and filter based on the Dratio</span></span>
<span><span class="va">filter_dratio</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/DratioFilter.html" class="external-link">DratioFilter</a></span><span class="op">(</span>threshold <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>                              qcIndex <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">phenotype</span> <span class="op">==</span> <span class="st">"QC"</span>,</span>
<span>                              studyIndex <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">phenotype</span> <span class="op">!=</span> <span class="st">"QC"</span>,</span>
<span>                              mad <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/filterFeatures.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="va">res</span>, filter <span class="op">=</span> <span class="va">filter_dratio</span>, assay <span class="op">=</span> <span class="st">"norm_imputed"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>4205 features were removed</code></pre>
</div>
</div>
<p>The Dratio filter is a powerful tool to identify features that exhibit high variability in the data, relating the variance observed in QC samples with that in study samples. By setting a threshold of 0.4, we remove features that have a high degree of variability between the QC and study samples. In this example, any feature in which the deviation at the QC is higher than 40% (<code>threshold = 0.4</code>)of the deviation on the study samples is removed. This filtering step ensures that only features are retained that have considerably lower technical than biological variance.</p>
<p>Note that the <code><a href="https://rdrr.io/pkg/MetaboCoreUtils/man/quality_assessment.html" class="external-link">rowDratio()</a></code> and <code><a href="https://rdrr.io/pkg/MetaboCoreUtils/man/quality_assessment.html" class="external-link">rowRsd()</a></code> functions from the <em><a href="https://bioconductor.org/packages/3.20/MetaboCoreUtils" class="external-link">MetaboCoreUtils</a></em> package could be used to calculate the actual numeric values for the estimates used for filtering, to e.g. to evaluate their distribution in the whole data set or identify data set-dependent threshold values.</p>
<p>Finally, we evaluate the number of features left after the filtering steps and calculate the percentage of features that were removed.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Number of features after analysis</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4519</code></pre>
</div>
<div class="sourceCode cell-code" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Percentage left: end/beginning</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">res_unfilt</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 51.79963</code></pre>
</div>
</div>
<p>The dataset has been reduced from 8724 to 4519 features. We did remove a considerable amount of features but this is expected as we want to focus on the most reliable features for our analysis. For the rest of our analysis we need to separate the QC samples from the study samples. We will store the QC samples in a separate object for later use.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb111"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_qc</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">[</span>, <span class="va">res</span><span class="op">$</span><span class="va">phenotype</span> <span class="op">==</span> <span class="st">"QC"</span><span class="op">]</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">[</span>, <span class="va">res</span><span class="op">$</span><span class="va">phenotype</span> <span class="op">!=</span> <span class="st">"QC"</span><span class="op">]</span></span></code></pre></div>
</div>
<p>We in addition calculate the CV of the QC samples and add that as an additional column to the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code> of our result object. These could be used later to prioritize identified <em>significant</em> features with e.g. low technical noise.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb112"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Calculate the QC's CV and add as feature variable to the data set</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">$</span><span class="va">qc_cv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res_qc</span>, <span class="st">"norm"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>               <span class="fu"><a href="https://rdrr.io/pkg/MetaboCoreUtils/man/quality_assessment.html" class="external-link">rowRsd</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<p>Now that our data set has been preprocessed, normalized and filtered, we can start to evaluate the distribution of the data and estimate the variation due to biology.</p>
</section><section class="section level2"><h2 id="overlapping-features">Overlapping features<a class="anchor" aria-label="anchor" href="#overlapping-features"></a>
</h2>
<p>a possible evalaution is to which features are overlapping between the different sample groups.</p>
<p>Fisrt we can plot the overlapping between the CVD and CTR group.</p>
<p>Before performing complex satistical analysis we can evaluate the data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb113"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span>CVD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"raw_filled"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">10</span>,</span>
<span>                 CTR <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"raw_filled"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<p>In case of a large number of groups a venn diagram can be replace by an upset plot which can handle a larger number of groups, while being still readable. below we compare the overlap between samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb114"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#make upsetplot</span></span></code></pre></div>
</div>
</section><section class="section level2"><h2 id="differential-abundance-analysis">Differential abundance analysis<a class="anchor" aria-label="anchor" href="#differential-abundance-analysis"></a>
</h2>
<p>After normalization and quality control, the next step is to identify features that are differentially abundant between the study groups. This crucial step allows us to identify potential biomarkers or metabolites that are associated with the study groups. There are various approaches and methods available for the identification of such features of interest. In this workflow we use a multiple linear regression analysis to identify features with significantly difference in abundances between the CVD or CTR study group.</p>
<p>Before performing the tests we evaluate similarities between study samples using a PCA (after excluding the QC samples to avoid them influencing the results).</p>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb115"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Define the colors for the plot</span></span>
<span><span class="va">col_sample</span> <span class="op">&lt;-</span> <span class="va">col_phenotype</span><span class="op">[</span><span class="va">res</span><span class="op">$</span><span class="va">phenotype</span><span class="op">]</span></span>
<span></span>
<span><span class="co">#' Log transform and scale the data for PCA analysis</span></span>
<span><span class="va">vals</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"norm_imputed"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span>center <span class="op">=</span> <span class="cn">TRUE</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">pca_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="va">vals</span>, scale <span class="op">=</span> <span class="cn">FALSE</span>, center <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">vals_st</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">vals</span>, phenotype <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">phenotype</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">pca_res</span>, data <span class="op">=</span> <span class="va">vals_st</span> , colour <span class="op">=</span> <span class="st">'phenotype'</span>, scale <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">col_phenotype</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/unnamed-chunk-53-1.png" width="672"></p>
<figcaption>Figure 34. PCA of the data after normalization and quality control.</figcaption></figure>
</div>
</div>
</div>
<p>The samples clearly separate by study group on PCA indicating that there are differences in the metabolite profiles between the two groups. However, what drives the separation on PC1 is not clear. Below we evaluate whether this could be explained by the other available variable of our study, i.e., age:</p>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Add age to the PCA plot</span></span>
<span><span class="va">vals_st</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">vals</span>, age <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">age</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/autoplot.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">pca_res</span>, data <span class="op">=</span> <span class="va">vals_st</span> , colour <span class="op">=</span> <span class="st">'age'</span>, scale <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/unnamed-chunk-54-1.png" width="672"></p>
<figcaption>Figure 35. PCA colored by age of the data after normalization and quality control.</figcaption></figure>
</div>
</div>
</div>
<p>According to the PCA above, PC1 does not seem to be related to age.</p>
<p>Even if there is some variance in the data set we can’t explain at this stage, we proceed with the (supervised) statistical tests to identify features of interest. We compute linear models for each metabolite explaining the observed feature abundance by the available study variables. While we could also use the base R function <code><a href="https://rdrr.io/r/stats/lm.html" class="external-link">lm()</a></code>, we utilize the <code>R Biocpkg("limma")</code> package to conduct the differential abundance analysis: the <em>moderated</em> test statistics <span class="citation" data-cites="smyth_linear_2004">(<a href="#ref-smyth_linear_2004" role="doc-biblioref">Smyth 2004</a>)</span> provided by this package are specifically well suited for experiments with a limited number of replicates. For our tests we use a linear model <code>~ phenotype + age</code>, hence explaining the abundances of one metabolite accounting for the study group assignment and age of each individual. The analysis might benefit from inclusion of a study covariate associated with PC2 or explaining the variance seen in that principal component, but for the present analysis only the participant’s age and disease association was provided.</p>
<p>Below we define the design of the study with the <code><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix()</a></code> function and fit feature-wise linear models to the log2-transformed abundances using the <code><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit()</a></code> function. P-values for significance of association are then calculated using the <code><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes()</a></code> function, that also performs the empirical Bayes-based robust estimation of the standard errors. See also the excellent vignette/user guide of the <em><a href="https://bioconductor.org/packages/3.20/limma" class="external-link">limma</a></em> package for examples and details on the linear model procedure.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb117"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Define the linear model to be applied to the data</span></span>
<span><span class="va">p.cut</span> <span class="op">&lt;-</span> <span class="fl">0.05</span>     <span class="co"># cut-off for significance.</span></span>
<span><span class="va">m.cut</span> <span class="op">&lt;-</span> <span class="fl">0.5</span>      <span class="co"># cut-off for log2 fold change</span></span>
<span></span>
<span><span class="va">age</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">age</span></span>
<span><span class="va">phenotype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">phenotype</span><span class="op">)</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">phenotype</span> <span class="op">+</span> <span class="va">age</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Fit the linear model to the data, explaining metabolite</span></span>
<span><span class="co">#' concentrations by phenotype and age.</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"norm_imputed"</span><span class="op">)</span><span class="op">)</span>, design <span class="op">=</span> <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code></pre></div>
</div>
<p>After the linear models have been fitted, we can now proceed to extract the results. We create a data frame containing the coefficients, raw and adjusted p-values (applying a Benjamini-Hochberg correction, i.e., <code>method = "BH"</code> for improved control of the false discovery rate), the average intensity of signals in CVD and CTR samples, and an indication of whether a feature is deemed significant or not. We consider all metabolites with an adjusted p-value smaller than 0.05 to be significant, but we could also include the (absolute) difference in abundances in the cut-off criteria. At last, we add the differential abundance results to the result object’s <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb118"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Compile a result data frame</span></span>
<span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    coef.CVD <span class="op">=</span> <span class="va">fit</span><span class="op">$</span><span class="va">coefficients</span><span class="op">[</span>, <span class="st">"phenotypeCVD"</span><span class="op">]</span>,</span>
<span>    pvalue.CVD <span class="op">=</span> <span class="va">fit</span><span class="op">$</span><span class="va">p.value</span><span class="op">[</span>, <span class="st">"phenotypeCVD"</span><span class="op">]</span>,</span>
<span>    adjp.CVD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">p.value</span><span class="op">[</span>, <span class="st">"phenotypeCVD"</span><span class="op">]</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span>,</span>
<span>    avg.CVD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowMeans.html" class="external-link">rowMeans</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"norm_imputed"</span><span class="op">)</span><span class="op">[</span>, <span class="va">res</span><span class="op">$</span><span class="va">phenotype</span> <span class="op">==</span> <span class="st">"CVD"</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    avg.CTR <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowMeans.html" class="external-link">rowMeans</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"norm_imputed"</span><span class="op">)</span><span class="op">[</span>, <span class="va">res</span><span class="op">$</span><span class="va">phenotype</span> <span class="op">==</span> <span class="st">"CTR"</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">tmp</span><span class="op">$</span><span class="va">significant.CVD</span> <span class="op">&lt;-</span> <span class="va">tmp</span><span class="op">$</span><span class="va">adjp.CVD</span> <span class="op">&lt;</span> <span class="fl">0.05</span></span>
<span><span class="co">#' Add the results to the object's rowData</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span>, <span class="va">tmp</span><span class="op">)</span></span></code></pre></div>
</div>
<p>We can now proceed to visualize the distribution of the raw and adjusted p-values.</p>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb119"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Plot the distribution of p-values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">$</span><span class="va">pvalue.CVD</span>, breaks <span class="op">=</span> <span class="fl">64</span>, xlab <span class="op">=</span> <span class="st">"p value"</span>,</span>
<span>     main <span class="op">=</span> <span class="st">"Distribution of raw p-values"</span>,</span>
<span>     cex.main <span class="op">=</span> <span class="fl">1</span>, cex.lab <span class="op">=</span> <span class="fl">1</span>, cex.axis <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">$</span><span class="va">adjp.CVD</span>, breaks <span class="op">=</span> <span class="fl">64</span>, xlab <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="va">p</span><span class="op">[</span><span class="va">BH</span><span class="op">]</span><span class="op">~</span><span class="va">value</span><span class="op">)</span>,</span>
<span>     main <span class="op">=</span> <span class="st">"Distribution of adjusted p-values"</span>,</span>
<span>     cex.main <span class="op">=</span> <span class="fl">1</span>, cex.lab <span class="op">=</span> <span class="fl">1</span>, cex.axis <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/histogram-1.png" width="672"></p>
<figcaption>Figure 36. Distribution of raw (left) and adjusted p-values (right).</figcaption></figure>
</div>
</div>
</div>
<p>The histograms above show the distribution of raw and adjusted p-values. Except for an enrichment of small p-values, the raw p-values are (more or less) uniformly distributed, which indicates the absence of any strong systematic biases in the data. The adjusted p-values are more conservative and account for multiple testing; this is important here as we fit a linear model to each feature and therefore perform a large number of tests which leads to a high chance of false positive findings. We do see that some features have very low p-values, indicating that they are likely to be significantly different between the two study groups.</p>
<p>Below we plot the adjusted p-values against the log2 fold change of (average) abundances. This volcano plot will allow us to visualize the features that are significantly different between the two study groups. These are highlighted with a blue color in the plot below.</p>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb120"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Plot volcano plot of the statistical results</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">5</span>, <span class="fl">5</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">$</span><span class="va">coef.CVD</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">$</span><span class="va">adjp.CVD</span><span class="op">)</span>,</span>
<span>     xlab <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="va">log</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">~</span><span class="va">difference</span><span class="op">)</span>,</span>
<span>     ylab <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="op">-</span><span class="va">log</span><span class="op">[</span><span class="fl">10</span><span class="op">]</span><span class="op">~</span><span class="va">p</span><span class="op">[</span><span class="va">BH</span><span class="op">]</span><span class="op">)</span>, pch <span class="op">=</span> <span class="fl">16</span>, col <span class="op">=</span> <span class="st">"#00000060"</span>,</span>
<span>     cex.main <span class="op">=</span> <span class="fl">1.5</span>, cex.lab <span class="op">=</span> <span class="fl">1.5</span>, cex.axis <span class="op">=</span> <span class="fl">1.3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>h <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"#0000ffcc"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/any.html" class="external-link">any</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">$</span><span class="va">significant.CVD</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/graphics/points.html" class="external-link">points</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">$</span><span class="va">coef.CVD</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">$</span><span class="va">significant.CVD</span><span class="op">]</span>,</span>
<span>           <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">$</span><span class="va">adjp.CVD</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">$</span><span class="va">significant.CVD</span><span class="op">]</span><span class="op">)</span>,</span>
<span>           col <span class="op">=</span> <span class="st">"#0000ffcc"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/volcano-1.png" width="672"></p>
<figcaption>Figure 37. Volcano plot showing the analysis results.</figcaption></figure>
</div>
</div>
</div>
<p>The most interesting features are in the top corners in the volcano plot (i.e., features with a large difference in abundance between the groups and a small p-value). All significant features have a negative coefficient (log2 fold change value) indicating that their abundance is lower in CVD samples compared to the CTR samples. These features are listed, along with their average difference in (log2) abundance between the compared groups, their adjusted p-values, their average (log2) abundance in each sample group and their RSD (CV) in QC samples in the table below.</p>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb121"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Table of significant features</span></span>
<span><span class="va">tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">$</span><span class="va">significant.CVD</span>,</span>
<span>                    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mzmed"</span>, <span class="st">"rtmed"</span>, <span class="st">"coef.CVD"</span>, <span class="st">"adjp.CVD"</span>,</span>
<span>                      <span class="st">"avg.CTR"</span>, <span class="st">"avg.CVD"</span>, <span class="st">"qc_cv"</span><span class="op">)</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">tab</span> <span class="op">&lt;-</span> <span class="va">tab</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">tab</span><span class="op">$</span><span class="va">coef.CVD</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="fu">kable</span><span class="op">(</span><span class="va">tab</span>, format <span class="op">=</span> <span class="st">"pipe"</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<table style="width:100%;" class="table">
<caption>Table 7. Features with significant differences in abundances.</caption>
<colgroup>
<col style="width: 9%">
<col style="width: 11%">
<col style="width: 13%">
<col style="width: 13%">
<col style="width: 13%">
<col style="width: 13%">
<col style="width: 13%">
<col style="width: 13%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">mzmed</th>
<th style="text-align: right;">rtmed</th>
<th style="text-align: right;">coef.CVD</th>
<th style="text-align: right;">adjp.CVD</th>
<th style="text-align: right;">avg.CTR</th>
<th style="text-align: right;">avg.CVD</th>
<th style="text-align: right;">qc_cv</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">FT0732</td>
<td style="text-align: right;">182.0749</td>
<td style="text-align: right;">34.83789</td>
<td style="text-align: right;">-8.459808</td>
<td style="text-align: right;">0.0099281</td>
<td style="text-align: right;">12.229286</td>
<td style="text-align: right;">3.859797</td>
<td style="text-align: right;">0.2116471</td>
</tr>
<tr class="even">
<td style="text-align: left;">FT0845</td>
<td style="text-align: right;">195.0877</td>
<td style="text-align: right;">32.65668</td>
<td style="text-align: right;">-6.330400</td>
<td style="text-align: right;">0.0424736</td>
<td style="text-align: right;">16.905340</td>
<td style="text-align: right;">10.454994</td>
<td style="text-align: right;">0.0304712</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FT0565</td>
<td style="text-align: right;">161.0400</td>
<td style="text-align: right;">162.13668</td>
<td style="text-align: right;">-5.512507</td>
<td style="text-align: right;">0.0368547</td>
<td style="text-align: right;">10.287074</td>
<td style="text-align: right;">4.578369</td>
<td style="text-align: right;">0.0360903</td>
</tr>
<tr class="even">
<td style="text-align: left;">FT1171</td>
<td style="text-align: right;">229.1299</td>
<td style="text-align: right;">181.08828</td>
<td style="text-align: right;">-5.411058</td>
<td style="text-align: right;">0.0179212</td>
<td style="text-align: right;">10.721195</td>
<td style="text-align: right;">5.500685</td>
<td style="text-align: right;">0.0706276</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FT0371</td>
<td style="text-align: right;">138.0547</td>
<td style="text-align: right;">148.39599</td>
<td style="text-align: right;">-5.194885</td>
<td style="text-align: right;">0.0179212</td>
<td style="text-align: right;">9.914862</td>
<td style="text-align: right;">4.405343</td>
<td style="text-align: right;">0.5564435</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Below we visualize the EICs of the significant features to evaluate their (raw) signal. We restrict the MS data set to study samples. Parameters</p>
<ul>
<li>
<code>keepFeatures = TRUE</code>: ensures that identified features are retained in the subset object.</li>
<li>
<code>peakBg</code>: defines the (background) color for each individual chromatographic peak in the EIC object.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb122"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Restrict the raw data to study samples.</span></span>
<span><span class="va">lcms1_study</span> <span class="op">&lt;-</span> <span class="va">lcms1</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">$</span><span class="va">phenotype</span> <span class="op">!=</span> <span class="st">"QC"</span>, keepFeatures <span class="op">=</span> <span class="cn">TRUE</span><span class="op">]</span></span>
<span><span class="co">#' Extract EICs for the significant features</span></span>
<span><span class="va">eic_sign</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/featureChromatograms.html" class="external-link">featureChromatograms</a></span><span class="op">(</span></span>
<span>    <span class="va">lcms1_study</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">tab</span><span class="op">)</span>, expandRt <span class="op">=</span> <span class="fl">5</span>, filled <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Plot the EICs.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">eic_sign</span>, col <span class="op">=</span> <span class="va">col_sample</span>,</span>
<span>     peakBg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">eic_sign</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span>, <span class="fl">40</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, col <span class="op">=</span> <span class="va">col_phenotype</span>, legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">col_phenotype</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/unnamed-chunk-57-1.png" width="672"></p>
<figcaption>Figure 38. Extracted ion chromatograms of the significant features.</figcaption></figure>
</div>
</div>
</div>
<p>The EICs for all significant features show a clear and single peak. Their intensities are (as already observed above) much larger in the CTR than in the CVD samples. With the exception of the second feature (second EIC in the top row), the intensities for all significant features are however generally low. This might make it challenging to identify them using an LC-MS/MS setup.</p>
</section><section class="section level2"><h2 id="annotation">Annotation<a class="anchor" aria-label="anchor" href="#annotation"></a>
</h2>
<p>We have now identified the features with significant differences in abundances between the two study groups. These provide information on metabolic pathways that differentiate affected from healthy individuals and might hence also serve as biomarkers. However, at this stage of the analysis we do not know what compounds/metabolites they actually represent. We thus need now to annotate these signals. Annotation can be performed at different level of confidence <span class="citation" data-cites="sumner_proposed_2007">Schymanski et al. (<a href="#ref-schymanski_identifying_2014" role="doc-biblioref">2014</a>)</span>. The lowest level of annotation, with the highest rate of false positive hits, bases on the features <em>m/z</em> ratios. Higher levels of annotations employ fragment spectra (MS2 spectra) of ions of interest requiring this however acquisition of additional data. In this section, we will demonstrate multiple ways to annotate our significant features using functionality provided through Bioconductor packages. Alternative approaches and external software tools, which may be better suited, will also be discussed later in the section.</p>
<section class="level2"><h3 id="ms1-based-annotation">MS1-based annotation<a class="anchor" aria-label="anchor" href="#ms1-based-annotation"></a>
</h3>
<p>Our data set was acquired using an LC-MS setup and our features are thus only characterized by their <em>m/z</em> and retention times. The retention time is LC-setup-specific and, without prior data/knowledge provide only little information on the features’ identity. Modern MS instruments have a high accuracy and the <em>m/z</em> values should therefore be reliable estimates of the compound ion’s mass-to-charge ratio. As first approach, we use the features’ <em>m/z</em> values and match them against reference values, i.e., exact masses of chemical compounds that are provided by a reference database, in our case the MassBank database. The full MassBank data is re-distributed through Bioconductor’s <em><a href="https://bioconductor.org/packages/3.20/AnnotationHub" class="external-link">AnnotationHub</a></em> resource, which simplifies their integration into reproducible R-based analysis workflows.</p>
<p>Below we load the resource, list all available MassBank data sets/releases and load one of them.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb123"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' load reference data</span></span>
<span><span class="va">ah</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">AnnotationHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#' List available MassBank data sets</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">query</a></span><span class="op">(</span><span class="va">ah</span>, <span class="st">"MassBank"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>AnnotationHub with 6 records
# snapshotDate(): 2024-10-28
# $dataprovider: MassBank
# $species: NA
# $rdataclass: CompDb
# additional mcols(): taxonomyid, genome, description,
#   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,
#   rdatapath, sourceurl, sourcetype
# retrieve records with, e.g., 'object[["AH107048"]]'

             title
  AH107048 | MassBank CompDb for release 2021.03
  AH107049 | MassBank CompDb for release 2022.06
  AH111334 | MassBank CompDb for release 2022.12.1
  AH116164 | MassBank CompDb for release 2023.06
  AH116165 | MassBank CompDb for release 2023.09
  AH116166 | MassBank CompDb for release 2023.11  </code></pre>
</div>
<div class="sourceCode cell-code" id="cb125"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Load one MAssBank release</span></span>
<span><span class="va">mb</span> <span class="op">&lt;-</span> <span class="va">ah</span><span class="op">[[</span><span class="st">"AH116166"</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
</div>
<p>The MassBank data is provided as a self-contained SQLite database and data can be queried and accessed through the <em><a href="https://bioconductor.org/packages/3.20/CompoundDb" class="external-link">CompoundDb</a></em> Bioconductor package. Below we use the <code><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">compounds()</a></code> function to extract small compound annotations from the database.</p>
<div class="cell" data-tbl-cap="Table 8. MassBank database extract.">
<div class="sourceCode cell-code" id="cb126"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Extract compound annotations</span></span>
<span><span class="va">cmps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">compounds</a></span><span class="op">(</span><span class="va">mb</span>, columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"name"</span>, <span class="st">"formula"</span>,</span>
<span>                                  <span class="st">"exactmass"</span>, <span class="st">"inchikey"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cmps</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        formula exactmass                    inchikey                 name
1    C27H29NO11  543.1741 AOJJSUZBOXZQNB-UHFFFAOYSA-N           Epirubicin
2      C40H54O4  598.4022 KFNGKYUGHHQDEE-AXWOCEAUSA-N Crassostreaxanthin A
3    C10H24N2O2  204.1838 AEUTYOVWOVBAKS-UWVGGRQHSA-N           Ethambutol
4     C16H27NO5  313.1889 LMFKRLGHEKVMNT-UJDVCPFMSA-N           Heliotrine
5 C20H15Cl3N2OS  435.9971 JLGKQTAYUIMGRK-UHFFFAOYSA-N        Sertaconazole
6      C15H14O5  274.0841 BWNCKEBBYADFPQ-UHFFFAOYSA-N    (R)Semivioxanthin</code></pre>
</div>
</div>
<p>MassBank (as most other small compound annotation databases) provides the (exact) molecular mass of each compound. Since almost all small compounds are neutral in their natural state, they need to be first converted to <em>m/z</em> values to allow matching against the feature’s <em>m/z</em>. To calculate an <em>m/z</em> from a neutral mass, we need to assume which ion (adduct) might be generated from the measured metabolites by the employed electro-spray ionization. In positive polarity, for human serum samples, the most common ions will be protonated (<em>[M+H]+</em>), or bear the addition of sodium (<em>[M+Na]+</em>) or ammonium (<em>[M+H-NH3]+</em>) ions. To match the observed <em>m/z</em> values against reference values from these potential ions we use again the <code><a href="https://rdrr.io/pkg/MetaboAnnotation/man/matchValues.html" class="external-link">matchValues()</a></code> function with the <code>Mass2MzParam</code> approach, that allows to specify the types of expected ions with its <code>adducts</code> parameter and the maximal allowed difference between the compared values using the <code>tolerance</code> and <code>ppm</code> parameters. Below we first prepare the data.frame with the significant features, set up the parameters for the matching and perform the comparison of the query features against the reference database.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb128"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Prepare query data frame</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">$</span><span class="va">feature_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">res_sig</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">$</span><span class="va">significant.CVD</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co">#' Setup parameters for the matching</span></span>
<span><span class="va">param</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MetaboAnnotation/man/matchValues.html" class="external-link">Mass2MzParam</a></span><span class="op">(</span>adducts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"[M+H]+"</span>, <span class="st">"[M+Na]+"</span>, <span class="st">"[M+H-NH3]+"</span><span class="op">)</span>,</span>
<span>                      tolerance <span class="op">=</span> <span class="fl">0</span>, ppm <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Perform the matching.</span></span>
<span><span class="va">mtch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MetaboAnnotation/man/matchValues.html" class="external-link">matchValues</a></span><span class="op">(</span><span class="va">res_sig</span>, <span class="va">cmps</span>, param <span class="op">=</span> <span class="va">param</span>, mzColname <span class="op">=</span> <span class="st">"mzmed"</span><span class="op">)</span></span>
<span><span class="va">mtch</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Object of class Matched
Total number of matches: 43
Number of query objects: 5 (4 matched)
Number of target objects: 25685 (43 matched)</code></pre>
</div>
</div>
<p>The resulting <code>Matched</code> object shows that 4 of our 6 significant features could be matched to ions of compounds from the MassBank database. Below we extract the full result from the <code>Matched</code> object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb130"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Extracting the results</span></span>
<span><span class="va">mtch_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MetaboAnnotation/man/Matched.html" class="external-link">matchedData</a></span><span class="op">(</span><span class="va">mtch</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"feature_id"</span>, <span class="st">"mzmed"</span>, <span class="st">"rtmed"</span>,</span>
<span>                                <span class="st">"adduct"</span>, <span class="st">"ppm_error"</span>,</span>
<span>                                <span class="st">"target_formula"</span>, <span class="st">"target_name"</span>,</span>
<span>                                <span class="st">"target_inchikey"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">mtch_res</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 44 rows and 8 columns
        feature_id     mzmed     rtmed      adduct ppm_error target_formula
       &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;character&gt; &lt;numeric&gt;    &lt;character&gt;
FT0371      FT0371   138.055   148.396      [M+H]+   2.08055        C7H7NO2
FT0371      FT0371   138.055   148.396      [M+H]+   1.93568        C7H7NO2
FT0371      FT0371   138.055   148.396      [M+H]+   2.08055        C7H7NO2
FT0371      FT0371   138.055   148.396      [M+H]+   1.93568        C7H7NO2
FT0371      FT0371   138.055   148.396      [M+H]+   1.93568        C7H7NO2
...            ...       ...       ...         ...       ...            ...
FT0845      FT0845   195.088   32.6567      [M+H]+ 0.1639884      C8H10N4O2
FT0845      FT0845   195.088   32.6567      [M+H]+ 0.1867474      C8H10N4O2
FT0845      FT0845   195.088   32.6567      [M+H]+ 0.0614704      C8H10N4O2
FT0845      FT0845   195.088   32.6567      [M+H]+ 0.1639884      C8H10N4O2
FT1171      FT1171   229.130  181.0883     [M+Na]+ 3.0770838      C12H18N2O
         target_name target_inchikey
         &lt;character&gt;     &lt;character&gt;
FT0371 Benzohydro...   VDEUYMSGMP...
FT0371 Trigonelli...   WWNNZCOKKK...
FT0371 Salicylami...   SKZKKFZAGN...
FT0371 4-Aminoben...   ALYNCZNDIQ...
FT0371 Anthranili...   RWZYAGGXGH...
...              ...             ...
FT0845      CAFFEINE   RYYVLZVUVI...
FT0845      Caffeine   RYYVLZVUVI...
FT0845      caffeine   RYYVLZVUVI...
FT0845 1,3,7-TRIM...   RYYVLZVUVI...
FT1171 Isoproturo...   PUIYMUZLKQ...</code></pre>
</div>
</div>
<p>Thus, in total 43 ions of compounds in MassBank were matched to our significant features based on the specified tolerance settings. Many compounds, while having a different structure and thus function/chemical property, have an identical chemical formula and thus mass. Matching exclusively on the <em>m/z</em> of features will hence result in many potentially false positive hits and is thus considered to provide only low confidence annotation. An additional complication is that some annotation resources, like MassBank, being community maintained, contain a large amount of redundant information. To reduce redundancy in the result table we below iterate over the hits of a feature and keep only matches to unique compounds (identified by their INCHIKEY). The INCHI or INCHIKEY combine information from compound’s chemical formula and structure, so while different compounds can share the same chemical formula, they should have a different structure and thus INCHI.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb132"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mtch_res</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span></span>
<span><span class="co">#' Keep only info on features that machted - create a utility function for that</span></span>
<span><span class="va">mtch_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">mtch_res</span>, <span class="va">mtch_res</span><span class="op">$</span><span class="va">feature_id</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">x</span><span class="op">$</span><span class="va">target_inchikey</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">z</span><span class="op">)</span> <span class="op">{</span></span>
<span>            <span class="va">z</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.min.html" class="external-link">which.min</a></span><span class="op">(</span><span class="va">z</span><span class="op">$</span><span class="va">ppm_error</span><span class="op">)</span>, <span class="op">]</span></span>
<span>        <span class="op">}</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span>recursive <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span>what <span class="op">=</span> <span class="va">rbind</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Display the results</span></span>
<span><span class="fu">kable</span><span class="op">(</span><span class="va">mtch_res</span>, format <span class="op">=</span> <span class="st">"pipe"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<table class="table">
<caption>Table 9. MS1 annotation results.</caption>
<colgroup>
<col style="width: 8%">
<col style="width: 6%">
<col style="width: 7%">
<col style="width: 6%">
<col style="width: 7%">
<col style="width: 11%">
<col style="width: 31%">
<col style="width: 21%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;">feature_id</th>
<th style="text-align: right;">mzmed</th>
<th style="text-align: right;">rtmed</th>
<th style="text-align: left;">adduct</th>
<th style="text-align: right;">ppm_error</th>
<th style="text-align: left;">target_formula</th>
<th style="text-align: left;">target_name</th>
<th style="text-align: left;">target_inchikey</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">FT0371</td>
<td style="text-align: right;">138.0547</td>
<td style="text-align: right;">148.39599</td>
<td style="text-align: left;">[M+H]+</td>
<td style="text-align: right;">1.9356815</td>
<td style="text-align: left;">C7H7NO2</td>
<td style="text-align: left;">4-Aminobenzoic acid</td>
<td style="text-align: left;">ALYNCZNDIQEVRV-UHFFFAOYSA-N</td>
</tr>
<tr class="even">
<td style="text-align: left;">FT0371</td>
<td style="text-align: right;">138.0547</td>
<td style="text-align: right;">148.39599</td>
<td style="text-align: left;">[M+H]+</td>
<td style="text-align: right;">1.9356815</td>
<td style="text-align: left;">C7H7NO2</td>
<td style="text-align: left;">2-PYRIDINECARBOXYLIC ACID METHYL ESTER</td>
<td style="text-align: left;">NMMIHXMBOZYNET-UHFFFAOYSA-N</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FT0371</td>
<td style="text-align: right;">138.0547</td>
<td style="text-align: right;">148.39599</td>
<td style="text-align: left;">[M+H]+</td>
<td style="text-align: right;">1.9356815</td>
<td style="text-align: left;">C7H7NO2</td>
<td style="text-align: left;">4-PYRIDINECARBOXYLIC ACID METHYL ESTER</td>
<td style="text-align: left;">OLXYLDUSSBULGU-UHFFFAOYSA-N</td>
</tr>
<tr class="even">
<td style="text-align: left;">FT0371</td>
<td style="text-align: right;">138.0547</td>
<td style="text-align: right;">148.39599</td>
<td style="text-align: left;">[M+H]+</td>
<td style="text-align: right;">1.9356815</td>
<td style="text-align: left;">C7H7NO2</td>
<td style="text-align: left;">SALICYLALDOXIME</td>
<td style="text-align: left;">ORIHZIZPTZTNCU-VMPITWQZSA-N</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FT0371</td>
<td style="text-align: right;">138.0547</td>
<td style="text-align: right;">148.39599</td>
<td style="text-align: left;">[M+H]+</td>
<td style="text-align: right;">1.9356815</td>
<td style="text-align: left;">C7H7NO2</td>
<td style="text-align: left;">ORTHO NITROTOLUENE</td>
<td style="text-align: left;">PLAZTCDQAHEYBI-UHFFFAOYSA-N</td>
</tr>
<tr class="even">
<td style="text-align: left;">FT0371</td>
<td style="text-align: right;">138.0547</td>
<td style="text-align: right;">148.39599</td>
<td style="text-align: left;">[M+H]+</td>
<td style="text-align: right;">1.9356815</td>
<td style="text-align: left;">C7H7NO2</td>
<td style="text-align: left;">4-PYRIDYL ACETATE</td>
<td style="text-align: left;">PTZQTYADHHANGW-UHFFFAOYSA-N</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FT0371</td>
<td style="text-align: right;">138.0547</td>
<td style="text-align: right;">148.39599</td>
<td style="text-align: left;">[M+H]+</td>
<td style="text-align: right;">1.9246374</td>
<td style="text-align: left;">C7H7NO2</td>
<td style="text-align: left;">Salicylaldehyde oxime</td>
<td style="text-align: left;">PUNVNOQWQQPNES-UHFFFAOYSA-N</td>
</tr>
<tr class="even">
<td style="text-align: left;">FT0371</td>
<td style="text-align: right;">138.0547</td>
<td style="text-align: right;">148.39599</td>
<td style="text-align: left;">[M+H]+</td>
<td style="text-align: right;">1.9356815</td>
<td style="text-align: left;">C7H7NO2</td>
<td style="text-align: left;">META NITROTOLUENE</td>
<td style="text-align: left;">QZYHIOPPLUPUJF-UHFFFAOYSA-N</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FT0371</td>
<td style="text-align: right;">138.0547</td>
<td style="text-align: right;">148.39599</td>
<td style="text-align: left;">[M+H]+</td>
<td style="text-align: right;">1.9356815</td>
<td style="text-align: left;">C7H7NO2</td>
<td style="text-align: left;">Anthranilic acid</td>
<td style="text-align: left;">RWZYAGGXGHYGMB-UHFFFAOYSA-N</td>
</tr>
<tr class="even">
<td style="text-align: left;">FT0371</td>
<td style="text-align: right;">138.0547</td>
<td style="text-align: right;">148.39599</td>
<td style="text-align: left;">[M+H]+</td>
<td style="text-align: right;">1.9356815</td>
<td style="text-align: left;">C7H7NO2</td>
<td style="text-align: left;">2-HYDROXYBENZAMIDE</td>
<td style="text-align: left;">SKZKKFZAGNVIMN-UHFFFAOYSA-N</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FT0371</td>
<td style="text-align: right;">138.0547</td>
<td style="text-align: right;">148.39599</td>
<td style="text-align: left;">[M+H]+</td>
<td style="text-align: right;">1.9246374</td>
<td style="text-align: left;">C7H7NO2</td>
<td style="text-align: left;">N-Hydroxybenzamide</td>
<td style="text-align: left;">VDEUYMSGMPQMIK-UHFFFAOYSA-N</td>
</tr>
<tr class="even">
<td style="text-align: left;">FT0371</td>
<td style="text-align: right;">138.0547</td>
<td style="text-align: right;">148.39599</td>
<td style="text-align: left;">[M+H]+</td>
<td style="text-align: right;">1.9356815</td>
<td style="text-align: left;">C7H7NO2</td>
<td style="text-align: left;">3-Pyridylacetic acid</td>
<td style="text-align: left;">WGNUNYPERJMVRM-UHFFFAOYSA-N</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FT0371</td>
<td style="text-align: right;">138.0547</td>
<td style="text-align: right;">148.39599</td>
<td style="text-align: left;">[M+H]+</td>
<td style="text-align: right;">1.9356815</td>
<td style="text-align: left;">C7H7NO2</td>
<td style="text-align: left;">Trigonelline</td>
<td style="text-align: left;">WWNNZCOKKKDOPX-UHFFFAOYSA-N</td>
</tr>
<tr class="even">
<td style="text-align: left;">FT0371</td>
<td style="text-align: right;">138.0547</td>
<td style="text-align: right;">148.39599</td>
<td style="text-align: left;">[M+H]+</td>
<td style="text-align: right;">1.9356815</td>
<td style="text-align: left;">C7H7NO2</td>
<td style="text-align: left;">META-AMINOBENZOIC ACID</td>
<td style="text-align: left;">XFDUHJPVQKIXHO-UHFFFAOYSA-N</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FT0371</td>
<td style="text-align: right;">138.0547</td>
<td style="text-align: right;">148.39599</td>
<td style="text-align: left;">[M+H]+</td>
<td style="text-align: right;">1.9356815</td>
<td style="text-align: left;">C7H7NO2</td>
<td style="text-align: left;">3-PYRIDINECARBOXYLIC ACID METHYL ESTER</td>
<td style="text-align: left;">YNBADRVTZLEFNH-UHFFFAOYSA-N</td>
</tr>
<tr class="even">
<td style="text-align: left;">FT0371</td>
<td style="text-align: right;">138.0547</td>
<td style="text-align: right;">148.39599</td>
<td style="text-align: left;">[M+H]+</td>
<td style="text-align: right;">1.9356815</td>
<td style="text-align: left;">C7H7NO2</td>
<td style="text-align: left;">4-NITROTOLUENE</td>
<td style="text-align: left;">ZPTVNYMJQHSSEA-UHFFFAOYSA-N</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FT0565</td>
<td style="text-align: right;">161.0400</td>
<td style="text-align: right;">162.13668</td>
<td style="text-align: left;">[M+Na]+</td>
<td style="text-align: right;">3.1247452</td>
<td style="text-align: left;">C8H10S</td>
<td style="text-align: left;">BENZYL METHYL SULFIDE</td>
<td style="text-align: left;">OFQPKKGMNWASPN-UHFFFAOYSA-N</td>
</tr>
<tr class="even">
<td style="text-align: left;">FT0845</td>
<td style="text-align: right;">195.0877</td>
<td style="text-align: right;">32.65668</td>
<td style="text-align: left;">[M+H]+</td>
<td style="text-align: right;">0.0614704</td>
<td style="text-align: left;">C8H10N4O2</td>
<td style="text-align: left;">Isocaffeine</td>
<td style="text-align: left;">LPHGQDQBBGAPDZ-UHFFFAOYSA-N</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FT0845</td>
<td style="text-align: right;">195.0877</td>
<td style="text-align: right;">32.65668</td>
<td style="text-align: left;">[M+H]+</td>
<td style="text-align: right;">0.0614704</td>
<td style="text-align: left;">C8H10N4O2</td>
<td style="text-align: left;">Caffeine</td>
<td style="text-align: left;">RYYVLZVUVIJVGH-UHFFFAOYSA-N</td>
</tr>
<tr class="even">
<td style="text-align: left;">FT0845</td>
<td style="text-align: right;">195.0877</td>
<td style="text-align: right;">32.65668</td>
<td style="text-align: left;">[M+H]+</td>
<td style="text-align: right;">0.1865772</td>
<td style="text-align: left;">C8H10N4O2</td>
<td style="text-align: left;">1,3-Benzenedicarboxylic acid, dihydrazide</td>
<td style="text-align: left;">UTTHLMXOSUFZCQ-UHFFFAOYSA-N</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FT1171</td>
<td style="text-align: right;">229.1299</td>
<td style="text-align: right;">181.08828</td>
<td style="text-align: left;">[M+Na]+</td>
<td style="text-align: right;">3.0770838</td>
<td style="text-align: left;">C12H18N2O</td>
<td style="text-align: left;">Isoproturon</td>
<td style="text-align: left;">PUIYMUZLKQOUOZ-UHFFFAOYSA-N</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The table above shows the results of the MS1-based annotation process. We can see that four of our significant features were matched. The matches seem to be pretty accurate with low ppm errors. The deduplication performed above considerably reduced the number of hits for each feature, but the first still matches ions of a large number of compounds (all with the same chemical formula).</p>
<p>Considering both features’ <em>m/z</em> and retention times in the MS1-based annotation will increase the annotation confidence, but requires additional data, such as recording of the retention time of thepure standard for a compound on the same LC setup. An alternative approach which might provide better inside on annotations and help to choose between different annotations for a same feature is to evaluate certain chemical properties of the possible matches. For instance, the LogP value, available in several databases such HMDB, provides an insight on a given compound’s polarity. As this property highly affects the interaction of the analyte with the column, it usually also directly affects separation. Therefore, a comparison between an analyte’s retention time and polarity can help to rule out some possible misidentifications.</p>
<p>While being of low confidence, MS1-based annotation can provide first candidate annotations that could be confirmed or rejected by additional analyses.</p>
</section><section class="level2"><h3 id="ms2-based-annotation">MS2-based annotation<a class="anchor" aria-label="anchor" href="#ms2-based-annotation"></a>
</h3>
<p>MS1 annotation is a fast and efficient method to annotate features and therefore give a first insight into the compounds that are significantly different between the two study groups. However, it is not always the most accurate. MS2 data can provide a higher level of confidence in the annotation process as it provides, through the observed fragmentation pattern, information on the structure of the compound.</p>
<p>MS2 data can be generated through LC-MS/MS measurement in which MS2 spectra are recorded for ions either in a data dependent acquisition (DDA) or data independent acquisition (DIA) mode. Generally, it is advised to include some LC-MS/MS runs of QC samples or randomly selected study samples already during the acquisition of the MS1 data that is used for quantification of the signals. As an alternative, or in addition, a post-hoc LC-MS/MS acquisition can be performed to generate the MS2 data needed for annotation. For the present experiment, a separate LC-MS/MS measurement was conducted on QC samples and selected study samples to generate such data using an <em>inclusion list</em> of pre-selected ions. These represent features found to be significantly different between CVD and CTR samples in an initial analysis of the full experiment. We use a subset of this second LC-MS/MS data set to show how such data can be used for MS2-based annotation. In our differential abundance analysis we found features with significantly higher abundances in CTR samples. Consequently, we will utilize the MS2 data obtained from the CTR samples to annotate these significant features.</p>
<p>Below we load the LC-MS/MS data from the experiment and restrict it to data acquired from the CTR sample.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb133"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Load form the MetaboLights Database</span></span>
<span><span class="va">param</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MsIO/man/MetaboLightsParam.html" class="external-link">MetaboLightsParam</a></span><span class="op">(</span>mtblsId <span class="op">=</span> <span class="st">"MTBLS8735"</span>,</span>
<span>                           assayName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"a_MTBLS8735_LC-MSMS_positive_"</span>,</span>
<span>                           <span class="st">"hilic_metabolite_profiling.txt"</span><span class="op">)</span>,</span>
<span>                           filePattern <span class="op">=</span> <span class="st">".mzML"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">lcms2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MsIO/man/saveMsObject.html" class="external-link">readMsObject</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">MsExperiment</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                     <span class="va">param</span>,</span>
<span>                     keepOntology <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                     keepProtocol <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                     simplify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb134"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#adjust sampleData</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms2</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample_name"</span>, <span class="st">"derived_spectra_data_file"</span>,</span>
<span>                                <span class="st">"metabolite_asssignment_file"</span>,</span>
<span>                                <span class="st">"source_name"</span>,</span>
<span>                                <span class="st">"organism"</span>,</span>
<span>                                <span class="st">"blood_sample_type"</span>,</span>
<span>                                <span class="st">"sample_type"</span>, <span class="st">"age"</span>, <span class="st">"unit"</span>, <span class="st">"phenotype"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># filter samples to keep MSMS data from CTR samples:</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms2</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms2</span><span class="op">)</span><span class="op">$</span><span class="va">phenotype</span> <span class="op">==</span> <span class="st">"CTR"</span>, <span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms2</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"MSMS"</span>, <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms2</span><span class="op">)</span><span class="op">$</span><span class="va">derived_spectra_data_file</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Add fragmentation data information (from filenames)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms2</span><span class="op">)</span><span class="op">$</span><span class="va">fragmentation_mode</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CE20"</span>, <span class="st">"CE30"</span>, <span class="st">"CES"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#let's look at the updated sample data</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms2</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"derived_spectra_data_file"</span>,</span>
<span>                     <span class="st">"phenotype"</span>, <span class="st">"sample_name"</span>, <span class="st">"age"</span><span class="op">)</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">kable</span><span class="op">(</span>format <span class="op">=</span> <span class="st">"pipe"</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<table class="table">
<caption>Table 10. Samples from the LC-MS/MS data set.</caption>
<thead><tr class="header">
<th style="text-align: left;">derived_spectra_data_file</th>
<th style="text-align: left;">phenotype</th>
<th style="text-align: left;">sample_name</th>
<th style="text-align: right;">age</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">FILES/MSMS_2_E_CE20_POS.mzML</td>
<td style="text-align: left;">CTR</td>
<td style="text-align: left;">E</td>
<td style="text-align: right;">66</td>
</tr>
<tr class="even">
<td style="text-align: left;">FILES/MSMS_2_E_CE30_POS.mzML</td>
<td style="text-align: left;">CTR</td>
<td style="text-align: left;">E</td>
<td style="text-align: right;">66</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FILES/MSMS_2_E_CES_POS.mzML</td>
<td style="text-align: left;">CTR</td>
<td style="text-align: left;">E</td>
<td style="text-align: right;">66</td>
</tr>
</tbody>
</table>
</div>
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb135"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Filter the data to the same RT range as the LC-MS run</span></span>
<span><span class="va">lcms2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterRt</a></span><span class="op">(</span><span class="va">lcms2</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">240</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</details>
</div>
<p>In total we have 3 LC-MS/MS data files for the control samples, each with a different collision energy to fragment the ions. Below we show the number of MS1 and MS2 spectra for each of the files.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb136"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' check the number of spectra per ms level</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">lcms2</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">msLevel</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">spectraSampleIndex</a></span><span class="op">(</span><span class="va">lcms2</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">table</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span>what <span class="op">=</span> <span class="va">cbind</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    1    2    3    4   5    6    7    8   9   10   11   12
1 825  186  186  186 825  186  186  186 825  185  186  185
2 825 3121 3118 3124 825 3123 3118 3120 825 3117 3117 3116</code></pre>
</div>
</div>
<p>Compared to the number of MS2 spectra, far less MS1 spectra were acquired. The configuration of the MS instrument was set to ensure that the ions specified in the <em>inclusion list</em> were selected for fragmentation, even if their intensity might be very low. With this setting, however, most of the recorded MS2 spectra will represent only noise. The plot below shows the location of precursor ions in the <em>m/z</em> - retention time plane for the three files.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb138"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/plotPrecursorIons.html" class="external-link">plotPrecursorIons</a></span><span class="op">(</span><span class="va">lcms2</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/unnamed-chunk-66-1.png" width="480"></p>
</figure>
</div>
</div>
</div>
<p>We can see MS2 spectra being recorded for the <em>m/z</em> of interest along the full retention time range, even if the actual ions will only be eluting within certain retention time windows.</p>
<p>We next extract the <code>Spectra</code> object with the MS data from the data object and assign a new spectra variable with the employed collision energy, which we extract from the data object <code>sampleData</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb139"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ms2_ctr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">lcms2</span><span class="op">)</span></span>
<span><span class="va">ms2_ctr</span><span class="op">$</span><span class="va">collision_energy</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms2</span><span class="op">)</span><span class="op">$</span><span class="va">fragmentation_mode</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">spectraSampleIndex</a></span><span class="op">(</span><span class="va">lcms2</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
</div>
<p>We next filter the MS data by first restricting to MS2 spectra and then removing mass peaks from each spectrum with an intensity that is lower than 5% of the highest intensity for that spectrum, assuming that these very low intensity peaks represent background signal.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb140"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Remove low intensity peaks</span></span>
<span><span class="va">low_int</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">x</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">x</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">*</span> <span class="fl">0.05</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">ms2_ctr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterMsLevel</a></span><span class="op">(</span><span class="va">ms2_ctr</span>, <span class="fl">2L</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterIntensity</a></span><span class="op">(</span>intensity <span class="op">=</span> <span class="va">low_int</span><span class="op">)</span></span></code></pre></div>
</div>
<p>We next remove also mass peaks with an <em>m/z</em> value greater or equal to the precursor <em>m/z</em> of the ion. This puts, for the later matching against reference spectra, more weight on the fragmentation pattern of ions and avoids hits based only on the precursor <em>m/z</em> peak (and hence a similar mass of the compared compounds). At last, we restrict the data to spectra with at least two fragment peaks and scale their intensities such that their sum is 1 for each spectrum. While similarity calculations are not affected by this scaling, it makes visual comparison of fragment spectra easier to read.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb141"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Remove precursor peaks and restrict to spectra with a minimum</span></span>
<span><span class="co">#' number of peaks</span></span>
<span><span class="va">ms2_ctr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/filterMsLevel.html" class="external-link">filterPrecursorPeaks</a></span><span class="op">(</span><span class="va">ms2_ctr</span>, ppm <span class="op">=</span> <span class="fl">50</span>, mz <span class="op">=</span> <span class="st">"&gt;="</span><span class="op">)</span></span>
<span><span class="va">ms2_ctr</span> <span class="op">&lt;-</span> <span class="va">ms2_ctr</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/lengths.html" class="external-link">lengths</a></span><span class="op">(</span><span class="va">ms2_ctr</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/addProcessing.html" class="external-link">scalePeaks</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<p>Finally, also to speed up the later comparison of the spectra against the reference database, we load the full MS data into memory (by changing the <em>backend</em> to <code>MsBackendMemory</code>) and <em>apply</em> all processing steps performed on this data so far. Keeping the MS data in memory has performance benefits, but is generally not suggested for large data sets. To evaluate the impact for the present data set we print in addition the size of the data object before and after changing the backend.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb142"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Size of the object before loading into memory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">object.size</a></span><span class="op">(</span><span class="va">ms2_ctr</span><span class="op">)</span>, units <span class="op">=</span> <span class="st">"MB"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>5.1 Mb</code></pre>
</div>
<div class="sourceCode cell-code" id="cb144"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Load the MS data subset into memory</span></span>
<span><span class="va">ms2_ctr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/backendInitialize.html" class="external-link">setBackend</a></span><span class="op">(</span><span class="va">ms2_ctr</span>, <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/MsBackend.html" class="external-link">MsBackendMemory</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ms2_ctr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/addProcessing.html" class="external-link">applyProcessing</a></span><span class="op">(</span><span class="va">ms2_ctr</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Size of the object after loading into memory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">object.size</a></span><span class="op">(</span><span class="va">ms2_ctr</span><span class="op">)</span>, units <span class="op">=</span> <span class="st">"MB"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>18.2 Mb</code></pre>
</div>
</div>
<p>There is thus only a moderate increase in memory demand after loading the MS data into memory (also because we filtered and cleaned the MS2 data).</p>
<p>While we could proceed and match all these experimental MS2 spectra against reference fragment spectra, for our workflow we aim to annotate the features found significant in the differential abundance analysis. The goal is thus to identify the MS2 spectra from the second (LC-MS/MS) run that could represent fragments of the ions of features from the data in the first (LC-MS) run. Our approach is to match MS2 spectra against the significant features determined earlier based on their precursor <em>m/z</em> and retention time (given an acceptable tolerance) to the feature’s <em>m/z</em> and retention time. This can be easily done using the <code><a href="https://rdrr.io/pkg/xcms/man/XcmsExperiment.html" class="external-link">featureArea()</a></code> function that effectively considers the actual <em>m/z</em> and retention time ranges of the features’ chromatographic peaks and therefore increase the chance of finding a correct match. This however also assumes that retention times between the first and second run don’t differ by much. Alternatively, we would need to align the retention times of the second LC-MS/MS data set to those of the first.</p>
<p>Below we first extract the <em>feature area</em>, i.e., the <em>m/z</em> and retention time ranges, for the significant features.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb146"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Define the m/z and retention time ranges for the significant features</span></span>
<span><span class="va">target</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XcmsExperiment.html" class="external-link">featureArea</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">res_sig</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">target</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          mzmin    mzmax     rtmin     rtmax
FT0371 138.0544 138.0552 146.32270 152.86115
FT0565 161.0391 161.0407 159.00234 164.30799
FT0732 182.0726 182.0756  32.71242  42.28755
FT0845 195.0799 195.0887  30.73235  35.67337
FT1171 229.1282 229.1335 178.01450 183.35303</code></pre>
</div>
</div>
<p>We next identify the fragment spectra with their precursor <em>m/z</em> and retention times within these ranges. We use the <code><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterRanges()</a></code> function that allows to filter a <code>Spectra</code> object using multiple ranges simultaneously. We apply this function separately to each feature (row in the above matrix) to extract the MS2 spectra representing fragmentation information of the presumed feature’s ions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb148"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Identify for each feature MS2 spectra with their precursor m/z and</span></span>
<span><span class="co">#' retention time within the feature's m/z and retention time range</span></span>
<span><span class="va">ms2_ctr_fts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">target</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rtmin"</span>, <span class="st">"rtmax"</span>, <span class="st">"mzmin"</span>, <span class="st">"mzmax"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                     MARGIN <span class="op">=</span> <span class="fl">1</span>, FUN <span class="op">=</span> <span class="va">filterRanges</span>, object <span class="op">=</span> <span class="va">ms2_ctr</span>,</span>
<span>                     spectraVariables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rtime"</span>, <span class="st">"precursorMz"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lengths.html" class="external-link">lengths</a></span><span class="op">(</span><span class="va">ms2_ctr_fts</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>FT0371 FT0565 FT0732 FT0845 FT1171
    38     36    135     68     38 </code></pre>
</div>
</div>
<p>The result from this <code><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply()</a></code> call is a <code>list</code> of <code>Spectra</code>, each element representing the result for one feature. With the exception of the last feature, multiple MS2 spectra could be identified. We next combine the <code>list</code> of <code>Spectra</code> into a single <code>Spectra</code> object using the <code><a href="https://rdrr.io/pkg/Spectra/man/combineSpectra.html" class="external-link">concatenateSpectra()</a></code> function and add an additional spectra variable containing the respective feature identifier.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb150"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">l</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lengths.html" class="external-link">lengths</a></span><span class="op">(</span><span class="va">ms2_ctr_fts</span><span class="op">)</span></span>
<span><span class="co">#' Combine the individual Spectra objects</span></span>
<span><span class="va">ms2_ctr_fts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/combineSpectra.html" class="external-link">concatenateSpectra</a></span><span class="op">(</span><span class="va">ms2_ctr_fts</span><span class="op">)</span></span>
<span><span class="co">#' Assign the feature identifier to each MS2 spectrum</span></span>
<span><span class="va">ms2_ctr_fts</span><span class="op">$</span><span class="va">feature_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">res_sig</span><span class="op">)</span>, <span class="va">l</span><span class="op">)</span></span></code></pre></div>
</div>
<p>We have now a <code>Spectra</code> object with fragment spectra for the significant features from our differential expression analysis.</p>
<p>We next build our reference data which we need to process the same way as our <em>query</em> spectra. We extract all fragment spectra from the MassBank database, restrict to positive polarity data (since our experiment was acquired in positive polarity) and perform the same processing to the fragment spectra from the MassBank database.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb151"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ms2_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/Spectra.html" class="external-link">Spectra</a></span><span class="op">(</span><span class="va">mb</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterPolarity</a></span><span class="op">(</span><span class="fl">1L</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterIntensity</a></span><span class="op">(</span>intensity <span class="op">=</span> <span class="va">low_int</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/filterMsLevel.html" class="external-link">filterPrecursorPeaks</a></span><span class="op">(</span>ppm <span class="op">=</span> <span class="fl">50</span>, mz <span class="op">=</span> <span class="st">"&gt;="</span><span class="op">)</span></span>
<span><span class="va">ms2_ref</span> <span class="op">&lt;-</span> <span class="va">ms2_ref</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/lengths.html" class="external-link">lengths</a></span><span class="op">(</span><span class="va">ms2_ref</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/addProcessing.html" class="external-link">scalePeaks</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<p>Note that here we could switch to a <code>MsBackendMemory</code> backend hence loading the full data from the reference database into memory. This could have a positive impact on the performance of the subsequent spectra matching, however it would also increase the memory demand of the present analysis.</p>
<p>Now that both the <code>Spectra</code> object for the second run and the database spectra have been prepared, we can proceed with the matching process. We use the <code><a href="https://rdrr.io/pkg/MetaboAnnotation/man/matchSpectra.html" class="external-link">matchSpectra()</a></code> function from the <em><a href="https://bioconductor.org/packages/3.20/MetaboAnnotation" class="external-link">MetaboAnnotation</a></em> package with the <code>CompareSpectraParam</code> to define the settings for the matching. With the following parameters:</p>
<ul>
<li>
<code>requirePrecursor = TRUE</code>: Limits spectra similarity calculations to fragment spectra with a similar precursor <em>m/z</em>.</li>
<li>
<code>tolerance</code> and <code>ppm</code>: Defines the acceptable difference between compared <em>m/z</em> values. These relaxed tolerance settings ensure that we find matches even if reference spectra were acquired on instruments with a lower accuracy.</li>
<li>
<code>THRESHFUN</code>: Defines which matches to report. Here, we keep all matches resulting in a spectra similarity score (calculated with the normalized dot product <span class="citation" data-cites="stein_optimization_1994">(<a href="#ref-stein_optimization_1994" role="doc-biblioref">Stein and Scott 1994</a>)</span>, the default similarity function) larger than 0.6.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb152"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html" class="external-link">register</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html" class="external-link">SerialParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#' Define the settings for the spectra matching.</span></span>
<span><span class="va">prm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MetaboAnnotation/man/CompareSpectraParam.html" class="external-link">CompareSpectraParam</a></span><span class="op">(</span>ppm <span class="op">=</span> <span class="fl">40</span>, tolerance <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                           requirePrecursor <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                           THRESHFUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">x</span> <span class="op">&gt;=</span> <span class="fl">0.6</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ms2_mtch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MetaboAnnotation/man/matchSpectra.html" class="external-link">matchSpectra</a></span><span class="op">(</span><span class="va">ms2_ctr_fts</span>, <span class="va">ms2_ref</span>, param <span class="op">=</span> <span class="va">prm</span><span class="op">)</span></span>
<span><span class="va">ms2_mtch</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Object of class MatchedSpectra
Total number of matches: 214
Number of query objects: 315 (16 matched)
Number of target objects: 69561 (21 matched)</code></pre>
</div>
</div>
<p>Thus, from the total 315 query MS2 spectra, only 16 could be matched to (at least) one reference fragment spectrum.</p>
<p>Below we restrict the results to these matching spectra and extract all metadata of the query and target spectra as well as the similarity score (the complete list of available metadata information can be listed with the <code><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames()</a></code> function).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb154"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Keep only query spectra with matching reference spectra</span></span>
<span><span class="va">ms2_mtch</span> <span class="op">&lt;-</span> <span class="va">ms2_mtch</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/MetaboAnnotation/man/Matched.html" class="external-link">whichQuery</a></span><span class="op">(</span><span class="va">ms2_mtch</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co">#' Extract the results</span></span>
<span><span class="va">ms2_mtch_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MetaboAnnotation/man/Matched.html" class="external-link">matchedData</a></span><span class="op">(</span><span class="va">ms2_mtch</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">ms2_mtch_res</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 214</code></pre>
</div>
</div>
<p>Now, we have query-target pairs with a spectra similarity higher than 0.6. Similar to the MS1-based annotation also this result table contains redundant information: we have multiple fragment spectra per feature and also MassBank contains several fragment spectra for each compound, measured using differing collision energies or MS instruments, by different laboratories. Below we thus iterate over the feature-compound pairs and select for each the one with the highest score. As an identifier for a compound, we use its fragment spectra’s INCHI-key, since compound names in MassBank do not have accepted consensus/controlled vocabularies.</p>
<div class="sourceCode cell-code" id="cb156"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' - split the result per feature</span></span>
<span><span class="co">#' - select for each feature the best matching result for each compound</span></span>
<span><span class="co">#' - combine the result again into a data frame</span></span>
<span><span class="va">ms2_mtch_res</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">ms2_mtch_res</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span>f <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">ms2_mtch_res</span><span class="op">$</span><span class="va">feature_id</span>, <span class="va">ms2_mtch_res</span><span class="op">$</span><span class="va">target_inchikey</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">z</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">z</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">z</span><span class="op">$</span><span class="va">score</span><span class="op">)</span>, <span class="op">]</span></span>
<span>    <span class="op">}</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span>what <span class="op">=</span> <span class="va">rbind</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' List the best matching feature-compound pair</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pander/man/pandoc.table.return.html" class="external-link">pandoc.table</a></span><span class="op">(</span><span class="va">ms2_mtch_res</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"feature_id"</span>, <span class="st">"target_name"</span>, <span class="st">"score"</span>,</span>
<span>                              <span class="st">"target_inchikey"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>             style <span class="op">=</span> <span class="st">"rmarkdown"</span>,</span>
<span>             caption <span class="op">=</span> <span class="st">"Table 9.MS2 annotation results."</span>,</span>
<span>             split.table <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<caption>Table 9.MS2 annotation results.</caption>
<colgroup>
<col style="width: 15%">
<col style="width: 35%">
<col style="width: 10%">
<col style="width: 38%">
</colgroup>
<thead><tr class="header">
<th style="text-align: center;">feature_id</th>
<th style="text-align: center;">target_name</th>
<th style="text-align: center;">score</th>
<th style="text-align: center;">target_inchikey</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">FT0371</td>
<td style="text-align: center;">3-Dimethylaminophenol</td>
<td style="text-align: center;">0.6274</td>
<td style="text-align: center;">MESJRHHDBDCQTH-UHFFFAOYSA-N</td>
</tr>
<tr class="even">
<td style="text-align: center;">FT0371</td>
<td style="text-align: center;">2-Methoxy-5-methylaniline</td>
<td style="text-align: center;">0.6149</td>
<td style="text-align: center;">WXWCDTXEKCVRRO-UHFFFAOYSA-N</td>
</tr>
<tr class="odd">
<td style="text-align: center;">FT0845</td>
<td style="text-align: center;">Isocaffeine</td>
<td style="text-align: center;">0.8048</td>
<td style="text-align: center;">LPHGQDQBBGAPDZ-UHFFFAOYSA-N</td>
</tr>
<tr class="even">
<td style="text-align: center;">FT0845</td>
<td style="text-align: center;">Caffeine</td>
<td style="text-align: center;">0.9205</td>
<td style="text-align: center;">NA</td>
</tr>
<tr class="odd">
<td style="text-align: center;">FT0845</td>
<td style="text-align: center;">Caffeine</td>
<td style="text-align: center;">0.9998</td>
<td style="text-align: center;">RYYVLZVUVIJVGH-UHFFFAOYSA-N</td>
</tr>
</tbody>
</table>
<p>Thus, from the 5 significant features, only one could be annotated to a compound based on the MS2-based approach. There could be many reasons for the failure to find matches for the other features. Although MS2 spectra were selected for each feature, most appear to only represent noise, as for most features, in the LC-MS/MS run, no or only very low MS1 signal was recorded, indicating that in that selected sample the original compound might not (or no longer) be present. Also, most reference databases contain predominantly fragment spectra for protonated (<em>[M+H]+</em>) ions of compounds, while some of the features might represent signal from other types of ions which would result in a different fragmentation pattern. Finally, fragment spectra of compounds of interest might also simply not be present in the used reference database.</p>
<p>Thus, combining the information from the MS1- and MS2 based annotation we can only annotate one feature with a considerable confidence. The feature with the <em>m/z</em> of 195.0879 and a retention time of 32 seconds seems to be an ion of caffeine. While the result is somewhat disappointing it also clearly shows the importance of proper experimental planning and the need to control for all potential confounding factors. For the present experiment, no disease-specific biomarker could be identified, but a life-style property of individuals suffering from this disease: coffee consumption was probably contraindicated to patients of the CVD group to reduce the risk of heart arrhythmia.</p>
<p>We below plot the EIC for this feature highlighting the retention time at which the highest scoring MS2 spectra was recorded and create a mirror plot comparing this MS2 spectra to the reference fragment spectra for caffeine.</p>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb157"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">col_sample</span> <span class="op">&lt;-</span> <span class="va">col_phenotype</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">lcms1</span><span class="op">)</span><span class="op">$</span><span class="va">phenotype</span><span class="op">]</span></span>
<span><span class="co">#' Extract and plot EIC for the annotated feature</span></span>
<span><span class="va">eic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/xcms/man/featureChromatograms.html" class="external-link">featureChromatograms</a></span><span class="op">(</span><span class="va">lcms1</span>, features <span class="op">=</span> <span class="va">ms2_mtch_res</span><span class="op">$</span><span class="va">feature_id</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">eic</span>, col <span class="op">=</span> <span class="va">col_sample</span>, peakCol <span class="op">=</span> <span class="va">col_sample</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">eic</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>     peakBg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">col_sample</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/xcms/man/XCMSnExp-class.html" class="external-link">chromPeaks</a></span><span class="op">(</span><span class="va">eic</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span>, <span class="fl">20</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, col <span class="op">=</span> <span class="va">col_phenotype</span>, legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">col_phenotype</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Identify the best matching query-target spectra pair</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">ms2_mtch_res</span><span class="op">$</span><span class="va">score</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' Indicate the retention time of the MS2 spectrum in the EIC plot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="va">ms2_mtch_res</span><span class="op">$</span><span class="va">rtime</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/unnamed-chunk-77-1.png" width="480"></p>
<figcaption>Figure 39. Extracted ion chromatograms and MS2 spectra for the annotated feature.</figcaption></figure>
</div>
</div>
</div>
<div class="cell">
<details class="code-fold"><summary>Show the code</summary><div class="sourceCode cell-code" id="cb158"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' Get the index of the MS2 spectrum in the query object</span></span>
<span><span class="va">query_idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">query</a></span><span class="op">(</span><span class="va">ms2_mtch</span><span class="op">)</span><span class="op">$</span><span class="va">.original_query_index</span> <span class="op">==</span></span>
<span>                                  <span class="va">ms2_mtch_res</span><span class="op">$</span><span class="va">.original_query_index</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">query_ms2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">query</a></span><span class="op">(</span><span class="va">ms2_mtch</span><span class="op">)</span><span class="op">[</span><span class="va">query_idx</span><span class="op">]</span></span>
<span><span class="co">#' Get the index of the MS2 spectrum in the target object</span></span>
<span><span class="va">target_idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/MetaboAnnotation/man/Matched.html" class="external-link">target</a></span><span class="op">(</span><span class="va">ms2_mtch</span><span class="op">)</span><span class="op">$</span><span class="va">spectrum_id</span> <span class="op">==</span></span>
<span>                                    <span class="va">ms2_mtch_res</span><span class="op">$</span><span class="va">target_spectrum_id</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">target_ms2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MetaboAnnotation/man/Matched.html" class="external-link">target</a></span><span class="op">(</span><span class="va">ms2_mtch</span><span class="op">)</span><span class="op">[</span><span class="va">target_idx</span><span class="op">]</span></span>
<span></span>
<span><span class="co">#' Create a mirror plot comparing the two best matching spectra</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html" class="external-link">plotSpectraMirror</a></span><span class="op">(</span><span class="va">query_ms2</span>, <span class="va">target_ms2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topleft"</span>,</span>
<span>       legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"precursor m/z: "</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/format.html" class="external-link">format</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">precursorMz</a></span><span class="op">(</span><span class="va">query_ms2</span><span class="op">)</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="a-end-to-end-untargeted-metabolomics_files/figure-html/unnamed-chunk-78-1.png" width="672"></p>
<figcaption>Figure 40. MS2 spectra for the annotated feature.</figcaption></figure>
</div>
</div>
</div>
<p>The plot above clearly shows the higher signal for this feature in CTR compared to CVD samples. The QC samples exhibit a lower but highly consistent signal, suggesting the absence of strong technical noise or biases in the raw data of this experiment. The vertical line indicates the retention time of the fragment spectrum with the best match against a reference spectrum. It has to be noted that, since the fragment spectra were measured in a separate LC-MS/MS experiment, this should only be considered as an indication of the approximate retention time in which ions were fragmented in the second experiment. The fragment spectrum for this feature, shown in the upper panel of the right plot above is highly similar to the reference spectrum for caffeine from MassBank (shown in the lower panel). In addition to their matching precursor <em>m/z</em>, the same two fragments (same <em>m/z</em> and intensity) are present in both spectra.</p>
<p>We can also extract additional metadata for the matching reference spectrum, such as the used collision energy, fragmentation mode, instrument type, instrument as well as the ion (adduct) that was fragmented.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb159"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectraData</a></span><span class="op">(</span><span class="va">target_ms2</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"collisionEnergy_text"</span>, <span class="st">"fragmentation_mode"</span>,</span>
<span>                          <span class="st">"instrument_type"</span>, <span class="st">"instrument"</span>, <span class="st">"adduct"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  collisionEnergy_text fragmentation_mode instrument_type
1         55 (nominal)                HCD     LC-ESI-ITFT
                         instrument adduct
1 LTQ Orbitrap XL Thermo Scientific [M+H]+</code></pre>
</div>
</div>
</section><section class="level2"><h3 id="external-tools-or-alternative-annotation-approaches">External tools or alternative annotation approaches<a class="anchor" aria-label="anchor" href="#external-tools-or-alternative-annotation-approaches"></a>
</h3>
<p>The present workflow highlights how annotation could be performed within R using packages from the Bioconductor project, but there are also other excellent external softwares that could be used as an alternative, such as SIRIUS <span class="citation" data-cites="duhrkop_sirius_2019">(<a href="#ref-duhrkop_sirius_2019" role="doc-biblioref">Dührkop et al. 2019</a>)</span>, mummichog <span class="citation" data-cites="li_predicting_2013">(<a href="#ref-li_predicting_2013" role="doc-biblioref">Li et al. 2013</a>)</span> or GNPS <span class="citation" data-cites="nothias_feature-based_2020">(<a href="#ref-nothias_feature-based_2020" role="doc-biblioref">Nothias et al. 2020</a>)</span> among others. To use these, the data would need to be exported in a format supported by these. For MS2 spectra, the data could easily be exported in the required MGF file format using the <em><a href="https://bioconductor.org/packages/3.20/MsBackendMgf" class="external-link">MsBackendMgf</a></em> Bioconductor package. Integration of <em>xcms</em> into feature-based molecular networking with GNPS is described in the <a href="https://ccms-ucsd.github.io/GNPSDocumentation/featurebasedmolecularnetworking-with-xcms3/" class="external-link">GNPS documentation</a>.</p>
<p>In alternative, or in addition, evidence for the potential matching chemical formula of a feature could be derived by evaluating the isotope pattern of its full MS1 scan. This could provide information on the isotope composition. Also for this, various functions such as the <code><a href="https://rdrr.io/pkg/MetaboCoreUtils/man/isotopologues.html" class="external-link">isotopologues()</a></code> from the or <em><a href="https://bioconductor.org/packages/3.20/MetaboCoreUtils" class="external-link">MetaboCoreUtils</a></em> package or the functionality of the <em>envipat</em> R package <span class="citation" data-cites="loos_accelerated_2015">(<a href="#ref-loos_accelerated_2015" role="doc-biblioref">Loos et al. 2015</a>)</span> could be used.</p>
</section></section><section class="section level2"><h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<p>In this tutorial, we describe an end-to-end workflow for LC-MS-based untargeted metabolomics experiments, conducted entirely within R using packages from the Bioconductor project or base R functionality. While other excellent software exists to perform similar analyses, the power of an R-based workflow lies in its adaptability to individual data sets or research questions and its ability to build reproducible workflows and documentation.</p>
<p>Due to space restrictions we don’t provide a comprehensive listing of methodologies for the individual analysis steps. More advanced options or approaches would be available, e.g., for normalization of the data, which will however also be heavily dependent on the size and properties of the analyzed data set, as well as for annotation of the features.</p>
<p>As a result, we found in the present analysis a set of features with significant abundance differences between the compared groups. From these we could however only reliably annotate a single feature, that related to the lifestyle of individuals rather then to the pathological properties of the investigated disease. This low proportion of annotated signals is however not uncommon for untargeted metabolomics experiments and reflects the need for more comprehensive and reliable reference annotation libraries.</p>
</section><section class="section level2"><h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb161"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.4.2 (2024-10-31)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 24.04.1 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods
[8] base

other attached packages:
 [1] MetaboAnnotation_1.10.1     CompoundDb_1.10.0
 [3] AnnotationFilter_1.30.0     AnnotationHub_3.14.0
 [5] BiocFileCache_2.14.0        dbplyr_2.5.0
 [7] gridExtra_2.3               ggfortify_0.4.17
 [9] ggplot2_3.5.1               vioplot_0.5.0
[11] zoo_1.8-12                  sm_2.2-6.0
[13] pheatmap_1.0.12             RColorBrewer_1.1-3
[15] pander_0.6.5                limma_3.62.1
[17] MetaboCoreUtils_1.14.0      xcms_4.5.1
[19] MsBackendMetaboLights_1.0.0 Spectra_1.16.0
[21] BiocParallel_1.40.0         alabaster.se_1.6.0
[23] alabaster.base_1.6.1        SummarizedExperiment_1.36.0
[25] Biobase_2.66.0              GenomicRanges_1.58.0
[27] GenomeInfoDb_1.42.1         IRanges_2.40.0
[29] S4Vectors_0.44.0            BiocGenerics_0.52.0
[31] MatrixGenerics_1.18.0       matrixStats_1.4.1
[33] MsIO_0.0.8                  MsExperiment_1.8.0
[35] ProtGenerics_1.38.0         readxl_1.4.3
[37] BiocStyle_2.34.0            quarto_1.4.4
[39] knitr_1.49

loaded via a namespace (and not attached):
  [1] later_1.4.1                 bitops_1.0-9
  [3] filelock_1.0.3              tibble_3.2.1
  [5] cellranger_1.1.0            preprocessCore_1.68.0
  [7] XML_3.99-0.17               lifecycle_1.0.4
  [9] doParallel_1.0.17           processx_3.8.4
 [11] lattice_0.22-6              MASS_7.3-61
 [13] MultiAssayExperiment_1.32.0 magrittr_2.0.3
 [15] rmarkdown_2.29              yaml_2.3.10
 [17] MsCoreUtils_1.18.0          DBI_1.2.3
 [19] abind_1.4-8                 zlibbioc_1.52.0
 [21] purrr_1.0.2                 RCurl_1.98-1.16
 [23] rappdirs_0.3.3              GenomeInfoDbData_1.2.13
 [25] MSnbase_2.32.0              ncdf4_1.23
 [27] codetools_0.2-20            DelayedArray_0.32.0
 [29] xml2_1.3.6                  DT_0.33
 [31] tidyselect_1.2.1            farver_2.1.2
 [33] UCSC.utils_1.2.0            base64enc_0.1-3
 [35] jsonlite_1.8.9              iterators_1.0.14
 [37] foreach_1.5.2               tools_4.4.2
 [39] progress_1.2.3              Rcpp_1.0.13-1
 [41] glue_1.8.0                  SparseArray_1.6.0
 [43] xfun_0.49                   dplyr_1.1.4
 [45] HDF5Array_1.34.0            withr_3.0.2
 [47] BiocManager_1.30.25         fastmap_1.2.0
 [49] rhdf5filters_1.18.0         fansi_1.0.6
 [51] digest_0.6.37               mime_0.12
 [53] R6_2.5.1                    colorspace_2.1-1
 [55] rsvg_2.6.1                  RSQLite_2.3.8
 [57] utf8_1.2.4                  tidyr_1.3.1
 [59] generics_0.1.3              prettyunits_1.2.0
 [61] PSMatch_1.10.0              httr_1.4.7
 [63] htmlwidgets_1.6.4           S4Arrays_1.6.0
 [65] pkgconfig_2.0.3             gtable_0.3.6
 [67] blob_1.2.4                  impute_1.80.0
 [69] MassSpecWavelet_1.72.0      XVector_0.46.0
 [71] htmltools_0.5.8.1           MALDIquant_1.22.3
 [73] clue_0.3-66                 scales_1.3.0
 [75] alabaster.matrix_1.6.1      png_0.1-8
 [77] rstudioapi_0.17.1           reshape2_1.4.4
 [79] rjson_0.2.23                curl_6.0.1
 [81] cachem_1.1.0                rhdf5_2.50.0
 [83] stringr_1.5.1               BiocVersion_3.20.0
 [85] parallel_4.4.2              AnnotationDbi_1.68.0
 [87] mzID_1.44.0                 vsn_3.74.0
 [89] pillar_1.9.0                grid_4.4.2
 [91] alabaster.schemas_1.6.0     vctrs_0.6.5
 [93] MsFeatures_1.14.0           pcaMethods_1.98.0
 [95] cluster_2.1.6               evaluate_1.0.1
 [97] cli_3.6.3                   compiler_4.4.2
 [99] rlang_1.1.4                 crayon_1.5.3
[101] labeling_0.4.3              QFeatures_1.16.0
[103] ChemmineR_3.58.0            ps_1.8.1
[105] affy_1.84.0                 plyr_1.8.9
[107] fs_1.6.5                    stringi_1.8.4
[109] munsell_0.5.1               Biostrings_2.74.0
[111] lazyeval_0.2.2              Matrix_1.7-1
[113] hms_1.1.3                   bit64_4.5.2
[115] Rhdf5lib_1.28.0             KEGGREST_1.46.0
[117] statmod_1.5.0               alabaster.ranges_1.6.0
[119] mzR_2.40.0                  igraph_2.1.1
[121] memoise_2.0.1               affyio_1.76.0
[123] bit_4.5.0                  </code></pre>
</div>
</div>
</section><section class="section level2"><h2 id="aknowledgment">Aknowledgment<a class="anchor" aria-label="anchor" href="#aknowledgment"></a>
</h2>
<p>Thanks to Steffen Neumann for his continuous work to develop and maintain the xcms software…</p>
</section><section class="section level2"><h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-broadhurst_guidelines_2018" class="csl-entry" role="listitem">
Broadhurst, David, Royston Goodacre, Stacey N. Reinke, Julia Kuligowski, Ian D. Wilson, Matthew R. Lewis, and Warwick B. Dunn. 2018. <span>“Guidelines and Considerations for the Use of System Suitability and Quality Control Samples in Mass Spectrometry Assays Applied in Untargeted Clinical Metabolomic Studies.”</span> <em>Metabolomics : Official Journal of the Metabolomic Society</em> 14 (6): 72. <a href="https://doi.org/10.1007/s11306-018-1367-3" class="external-link">https://doi.org/10.1007/s11306-018-1367-3</a>.
</div>
<div id="ref-chambers_cross-platform_2012" class="csl-entry" role="listitem">
Chambers, Matthew C, Brendan Maclean, Robert Burke, Dario Amodei, Daniel L Ruderman, Steffen Neumann, Laurent Gatto, et al. 2012. <span>“A Cross-Platform Toolkit for Mass Spectrometry and Proteomics.”</span> <em>Nature Biotechnology</em> 30 (10): 918–20. <a href="https://doi.org/10.1038/nbt.2377" class="external-link">https://doi.org/10.1038/nbt.2377</a>.
</div>
<div id="ref-de_livera_normalizing_2012" class="csl-entry" role="listitem">
De Livera, Alysha M., Daniel A. Dias, David De Souza, Thusitha Rupasinghe, James Pyke, Dedreia Tull, Ute Roessner, Malcolm McConville, and Terence P. Speed. 2012. <span>“Normalizing and Integrating Metabolomics Data.”</span> <em>Analytical Chemistry</em> 84 (24): 10768–76. <a href="https://doi.org/10.1021/ac302748b" class="external-link">https://doi.org/10.1021/ac302748b</a>.
</div>
<div id="ref-duhrkop_sirius_2019" class="csl-entry" role="listitem">
Dührkop, Kai, Markus Fleischauer, Marcus Ludwig, Alexander A. Aksenov, Alexey V. Melnik, Marvin Meusel, Pieter C. Dorrestein, Juho Rousu, and Sebastian Böcker. 2019. <span>“<span>SIRIUS</span> 4: A Rapid Tool for Turning Tandem Mass Spectra into Metabolite Structure Information.”</span> <em>Nature Methods</em> 16 (4): 299–302. <a href="https://doi.org/10.1038/s41592-019-0344-8" class="external-link">https://doi.org/10.1038/s41592-019-0344-8</a>.
</div>
<div id="ref-klavus_notame_2020" class="csl-entry" role="listitem">
Klåvus, Anton, Marietta Kokla, Stefania Noerman, Ville M. Koistinen, Marjo Tuomainen, Iman Zarei, Topi Meuronen, et al. 2020. <span>“<span>‘<span>Notame</span>’</span>: <span>Workflow</span> for <span>Non</span>-<span>Targeted</span> <span>LC</span>–<span>MS</span> <span>Metabolic</span> <span>Profiling</span>.”</span> <em>Metabolites</em> 10 (4): 135. <a href="https://doi.org/10.3390/metabo10040135" class="external-link">https://doi.org/10.3390/metabo10040135</a>.
</div>
<div id="ref-li_predicting_2013" class="csl-entry" role="listitem">
Li, Shuzhao, Youngja Park, Sai Duraisingham, Frederick H. Strobel, Nooruddin Khan, Quinlyn A. Soltow, Dean P. Jones, and Bali Pulendran. 2013. <span>“Predicting Network Activity from High Throughput Metabolomics.”</span> <em>PLoS Computational Biology</em> 9 (7): e1003123. <a href="https://doi.org/10.1371/journal.pcbi.1003123" class="external-link">https://doi.org/10.1371/journal.pcbi.1003123</a>.
</div>
<div id="ref-loos_accelerated_2015" class="csl-entry" role="listitem">
Loos, Martin, Christian Gerber, Francesco Corona, Juliane Hollender, and Heinz Singer. 2015. <span>“Accelerated <span>Isotope</span> <span>Fine</span> <span>Structure</span> <span>Calculation</span> <span>Using</span> <span>Pruned</span> <span>Transition</span> <span>Trees</span>.”</span> <em>Analytical Chemistry</em> 87 (11): 5738–44. <a href="https://doi.org/10.1021/acs.analchem.5b00941" class="external-link">https://doi.org/10.1021/acs.analchem.5b00941</a>.
</div>
<div id="ref-nothias_feature-based_2020" class="csl-entry" role="listitem">
Nothias, Louis-Félix, Daniel Petras, Robin Schmid, Kai Dührkop, Johannes Rainer, Abinesh Sarvepalli, Ivan Protsyuk, et al. 2020. <span>“Feature-Based Molecular Networking in the <span>GNPS</span> Analysis Environment.”</span> <em>Nature Methods</em> 17 (9): 905–8. <a href="https://doi.org/10.1038/s41592-020-0933-6" class="external-link">https://doi.org/10.1038/s41592-020-0933-6</a>.
</div>
<div id="ref-rainer_modular_2022" class="csl-entry" role="listitem">
Rainer, Johannes, Andrea Vicini, Liesa Salzer, Jan Stanstrup, Josep M. Badia, Steffen Neumann, Michael A. Stravs, et al. 2022. <span>“A <span>Modular</span> and <span>Expandable</span> <span>Ecosystem</span> for <span>Metabolomics</span> <span>Data</span> <span>Annotation</span> in <span>R</span>.”</span> <em>Metabolites</em> 12 (2): 173. <a href="https://doi.org/10.3390/metabo12020173" class="external-link">https://doi.org/10.3390/metabo12020173</a>.
</div>
<div id="ref-schymanski_identifying_2014" class="csl-entry" role="listitem">
Schymanski, Emma L., Junho Jeon, Rebekka Gulde, Kathrin Fenner, Matthias Ruff, Heinz P. Singer, and Juliane Hollender. 2014. <span>“Identifying <span>Small</span> <span>Molecules</span> via <span>High</span> <span>Resolution</span> <span>Mass</span> <span>Spectrometry</span>: <span>Communicating</span> <span>Confidence</span>.”</span> <em>Environmental Science &amp; Technology</em> 48 (4): 2097–98. <a href="https://doi.org/10.1021/es5002105" class="external-link">https://doi.org/10.1021/es5002105</a>.
</div>
<div id="ref-smith_xcms_2006" class="csl-entry" role="listitem">
Smith, Colin A., Elizabeth J. Want, Grace O&amp;apos;Maille, Ruben Abagyan, and Gary Siuzdak. 2006. <span>“<span>XCMS</span>: Processing Mass Spectrometry Data for Metabolite Profiling Using Nonlinear Peak Alignment, Matching, and Identification.”</span> <em>Analytical Chemistry</em> 78 (3): 779–87. <a href="https://doi.org/10.1021/ac051437y" class="external-link">https://doi.org/10.1021/ac051437y</a>.
</div>
<div id="ref-smyth_linear_2004" class="csl-entry" role="listitem">
Smyth, Gordon K. 2004. <span>“Linear Models and Empirical <span>Bayes</span> Methods for Assessing Differential Expression in Microarray Experiments.”</span> <em>Statistical Applications in Genetics and Molecular Biology</em> 3: Art. 3, 29 pp. (electronic).
</div>
<div id="ref-stein_optimization_1994" class="csl-entry" role="listitem">
Stein, Stephen E., and Donald R. Scott. 1994. <span>“Optimization and Testing of Mass Spectral Library Search Algorithms for Compound Identification.”</span> <em>Journal of the American Society for Mass Spectrometry</em> 5 (9): 859–66. <a href="https://doi.org/10.1021/jasms.8b00613" class="external-link">https://doi.org/10.1021/jasms.8b00613</a>.
</div>
<div id="ref-sumner_proposed_2007" class="csl-entry" role="listitem">
Sumner, Lloyd W., Alexander Amberg, Dave Barrett, Michael H. Beale, Richard Beger, Clare A. Daykin, Teresa W.-M. Fan, et al. 2007. <span>“Proposed Minimum Reporting Standards for Chemical Analysis <span>Chemical</span> <span>Analysis</span> <span>Working</span> <span>Group</span> (<span>CAWG</span>) <span>Metabolomics</span> <span>Standards</span> <span>Initiative</span> (<span>MSI</span>).”</span> <em>Metabolomics : Official Journal of the Metabolomic Society</em> 3 (3): 211–21. <a href="https://doi.org/10.1007/s11306-007-0082-2" class="external-link">https://doi.org/10.1007/s11306-007-0082-2</a>.
</div>
<div id="ref-tautenhahn_highly_2008" class="csl-entry" role="listitem">
Tautenhahn, Ralf, Christoph Böttcher, and Steffen Neumann. 2008. <span>“Highly Sensitive Feature Detection for High Resolution <span>LC</span>/<span>MS</span>.”</span> <em>BMC Bioinformatics</em> 9 (1): 504. <a href="https://doi.org/10.1186/1471-2105-9-504" class="external-link">https://doi.org/10.1186/1471-2105-9-504</a>.
</div>
</div>
</section><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'light-border',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config);
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script></main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Philippine Louail, Johannes Rainer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
