<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Normalization and feature selection with the notame package • Metabonaut</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><script src="notame_normalization_and_feature_selection_files/libs/quarto-html/popper.min.js"></script><script src="notame_normalization_and_feature_selection_files/libs/quarto-html/tippy.umd.min.js"></script><link href="notame_normalization_and_feature_selection_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="notame_normalization_and_feature_selection_files/libs/quarto-html/light-border.css" rel="stylesheet">
<!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Normalization and feature selection with the notame package">
<meta property="og:image" content="https://rformassspectrometry.github.io/Metabonaut/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Metabonaut</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.4.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/install_v0.html">Installation instructions</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-workflows" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Workflows</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-workflows">
<li><a class="dropdown-item" href="../articles/a-end-to-end-untargeted-metabolomics.html">End-to-end workflow for untargeted metabolomics data analysis in R</a></li>
    <li><a class="dropdown-item" href="../articles/notame_normalization_feature_selection.html">Quality control and Feature selection with Notame</a></li>
    <li><a class="dropdown-item" href="../articles/dataset-investigation.html">Dataset investigation: What to do when you get your data</a></li>
    <li><a class="dropdown-item" href="../articles/alignment-to-external-dataset.html">Seamless Alignment: Merging New data with Existing Preprocessed Datasets</a></li>
    <li><a class="dropdown-item" href="../articles/SpectriPy-tutorial-metabonaut.html">LC-MS/MS Data Annotation using R and Python</a></li>
    <li><a class="dropdown-item" href="../articles/large-scale-analysis.html">Large Scale Data Preprocessing with xcms</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/rformassspectrometry/Metabonaut/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-quarto">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Normalization and feature selection with the notame package</h1>


      <p class="date">2025-12-12</p>

      <small class="dont-index">Source: <a href="https://github.com/rformassspectrometry/Metabonaut/blob/devel/vignettes/notame_normalization_and_feature_selection.qmd" class="external-link"><code>vignettes/notame_normalization_and_feature_selection.qmd</code></a></small>
      <div class="d-none name"><code></code></div>
    </div>






    <section class="section level2"><h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Herein, we will perform normalization and feature selection using the <em>notame</em> R packages, originally developed in parallel with a protocol article published in the “Metabolomics Data Processing and Data Analysis—Current Best Practices” special issue of the Metabolites journal <span class="citation" data-cites="klavus_notame_2020">(<a href="#ref-klavus_notame_2020" role="doc-biblioref">Klåvus et al. 2020</a>)</span>. The main outcome is identifying interesting features for laborious downstream steps relating to biological context, such as annotation and pathway analysis, which fall outside the purview of <em>notame</em>. The associated protocol article presents a sequence of steps that is easily adopted by new practitioners. We will not follow the protocol article exactly and instead do an analysis which meshes with the trunk of the end-to- end Metabonaut workflow.</p>
<p>The normalization process is based on QC samples so the sample size here is limiting, with only four QC samples. The QC samples were also pooled from another sample collection, whereas typically, the QC samples would be pooled from the same sample collection. These limitations will become evident below, so we make an effort to explain the basic things one would want to look out for.</p>
</section><section class="section level2"><h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<p>Let’s prepare by attaching packages and loading the preprocessed data as a SummarizedExperiment object returned from <code>MSExperiment::quantify()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">alabaster.se</span><span class="op">)</span> <span class="co"># Portable loading and saving of SummarizedExperiment</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span> <span class="co"># Parallelization</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span> <span class="co"># Combinine plots in pane</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/limma/" class="external-link">limma</a></span><span class="op">)</span> <span class="co"># Differential expression analysis</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hanhineva-lab/notame" class="external-link">notame</a></span><span class="op">)</span> <span class="co"># Preprocessing</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hanhineva-lab/notameViz" class="external-link">notameViz</a></span><span class="op">)</span> <span class="co"># Visualization</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hanhineva-lab/notameStats" class="external-link">notameStats</a></span><span class="op">)</span> <span class="co"># Feature selection</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span> <span class="co"># The SummarizedExperiment container</span></span>
<span></span>
<span><span class="co"># Set up parallel processing using 2 cores</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">.Platform</span><span class="op">$</span><span class="va">OS.type</span> <span class="op">==</span> <span class="st">"unix"</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html" class="external-link">register</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html" class="external-link">register</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"preprocessed_res"</span>, package <span class="op">=</span> <span class="st">"Metabonaut"</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/alabaster.base/man/readObject.html" class="external-link">readObject</a></span><span class="op">(</span><span class="va">dir</span><span class="op">)</span></span>
<span><span class="va">se</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SummarizedExperiment
dim: 9068 10
metadata(0):
assays(2): raw raw_filled
rownames(9068): FT0001 FT0002 ... FT9067 FT9068
rowData names(11): mzmed mzmin ... QC ms_level
colnames(10): MS_QC_POOL_1_POS.mzML MS_A_POS.mzML ... MS_F_POS.mzML
  MS_QC_POOL_4_POS.mzML
colData names(11): sample_name derived_spectra_data_file ... phenotype
  injection_index</code></pre>
</div>
</div>
<p>The <code>SummarizedExperiment</code> container supports several assays. In <em>notame</em>, you need to specify the assay using the <code>assay.type</code> parameter if using multiple assays. Herein, we have two assays: “raw” contains the detected peaks without gap-filling, whereas “raw_filled” contains the gap-filled peak table. For this demo to go smoothly, we prepare the SummarizedExperiment object for all the functions at once, mostly by renaming and creating columns.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Rename columns of sample data</span></span>
<span><span class="va">rename_ind</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span></span>
<span>                    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample_name"</span>, <span class="st">"sample_type"</span>, <span class="st">"injection_index"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="va">rename_ind</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Sample_ID"</span>, <span class="st">"QC"</span>, <span class="st">"Injection_order"</span><span class="op">)</span></span>
<span><span class="co"># Change rownames to be identical with "Sample_ID" column</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">Sample_ID</span></span>
<span><span class="co"># Change name of pooled samples to "QC" in the "QC" column</span></span>
<span><span class="va">se</span><span class="op">$</span><span class="va">QC</span><span class="op">[</span><span class="va">se</span><span class="op">$</span><span class="va">QC</span> <span class="op">==</span> <span class="st">"pool"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"QC"</span></span>
<span><span class="co"># Convert phenotype column to factor</span></span>
<span><span class="va">se</span><span class="op">$</span><span class="va">phenotype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">se</span><span class="op">$</span><span class="va">phenotype</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># In feature data, create Feature_ID column from rownames</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span>Feature_ID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Create "Split" column with analytical mode</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">Split</span> <span class="op">&lt;-</span> <span class="st">"HILIC_pos"</span></span>
<span><span class="co"># Create "Flag" column for flagging low-quality features</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">Flag</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span></span>
<span><span class="co"># Convert assay from DelayedMatrix to regular matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"raw_filled"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"raw_filled"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"raw"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"raw"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</section><section class="section level2"><h2 id="normalization">Normalization<a class="anchor" aria-label="anchor" href="#normalization"></a>
</h2>
<p>Features with a low detection rate are flagged, as they are deemed too unreliable not only for statistical analysis, but also for the normalization process which relies heavily on QC samples. Flagged features are automatically excluded in most functions in <em>notame</em>, but can be included using <code>all_features = TRUE</code>. We set the detection rate threshold for QC samples at 70% <span class="citation" data-cites="broadhurst_guidelines_2018">(<a href="#ref-broadhurst_guidelines_2018" role="doc-biblioref">Broadhurst et al. 2018</a>)</span>, plus a within-group threshold of 80%. The within-group threshold must be met in at least one study group.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://hanhineva-lab.github.io/notame/reference/flag_detection.html" class="external-link">flag_detection</a></span><span class="op">(</span><span class="va">se</span>, qc_limit <span class="op">=</span> <span class="fl">0.7</span>, group_limit <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>                     group <span class="op">=</span> <span class="st">"phenotype"</span>, assay.type <span class="op">=</span> <span class="st">"raw_filled"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>INFO [2025-12-12 13:16:36]
11% of features flagged for low detection rate</code></pre>
</div>
</div>
<p>Next, we correct for drift using a cubic spline, relating each features’ abundance in QC samples to injection order <span class="citation" data-cites="kirwan_characterising_2013">(<a href="#ref-kirwan_characterising_2013" role="doc-biblioref">Kirwan et al. 2013</a>)</span>. Drift correction is performed in log-space since the log transformed data better follows assumptions of cubic spline regression. The value for the smoothing parameter is, by default, optimized using leave-one-out cross-validation to avoid overfitting. The abundances are corrected by adding the mean of a feature’s abundance in the QC samples and subtracting the predicted fit for each feature.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://hanhineva-lab.github.io/notame/reference/correct_drift.html" class="external-link">correct_drift</a></span><span class="op">(</span><span class="va">se</span>, assay.type <span class="op">=</span> <span class="st">"raw_filled"</span>, name <span class="op">=</span> <span class="st">"drift_norm"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>INFO [2025-12-12 13:16:36] Starting drift correction
INFO [2025-12-12 13:16:44] Recomputing quality metrics for drift corrected data
INFO [2025-12-12 13:16:46] Drift correction performed
INFO [2025-12-12 13:16:46] Inspecting drift correction results
INFO [2025-12-12 13:16:46] Original quality metrics missing, recomputing
INFO [2025-12-12 13:16:48] Drift correction results inspected: Drift_corrected: 94%,  Missing_QCS: 6%</code></pre>
</div>
</div>
<p>Brief notes about drift correction are stored in the <code>DC_note</code> column of feature data. For example, it is noted if drift correction couldn’t be performed because because at least four QC samples with values are needed for fitting the cubic spline.</p>
<p>Next, we apply probabilistic quotient normalization (PQN) to reduce unwanted variation from differential dilution of samples <span class="citation" data-cites="dieterle_probabilistic_2006">(<a href="#ref-dieterle_probabilistic_2006" role="doc-biblioref">Dieterle et al. 2006</a>)</span>. The central challenge here is to reduce unwanted variation from dilution, whether biological or experimental, while accounting for the possibility of biologically meaningful variation in total feature abundances. In probabilistic quotient normalization, the most probable dilution factor is determined for each sample as the median of quotients calculated for each feature relative to the median of each feature’s abundance in reference samples. The sample abundances are then divided by the dilution factor. Below, we use quality features in QC samples as reference, although the literature suggests that the choice of reference samples isn’t critical <span class="citation" data-cites="dieterle_probabilistic_2006">(<a href="#ref-dieterle_probabilistic_2006" role="doc-biblioref">Dieterle et al. 2006</a>)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://hanhineva-lab.github.io/notame/reference/pqn_normalization.html" class="external-link">pqn_normalization</a></span><span class="op">(</span><span class="va">se</span>, ref <span class="op">=</span> <span class="st">"qc"</span>, method <span class="op">=</span> <span class="st">"median"</span>,</span>
<span>                        assay.type <span class="op">=</span> <span class="st">"drift_norm"</span>, name <span class="op">=</span> <span class="st">"dil_norm"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>INFO [2025-12-12 13:16:48] Starting PQN normalization
INFO [2025-12-12 13:16:48] Using median of qc samples as reference spectrum</code></pre>
</div>
</div>
<p>Next, let’s visualize the data before and after normalization for drift and dilution using the <em>notameViz</em> package. Typically, the data would be assessed visually after each step of the normalization process, using the <code><a href="https://rdrr.io/pkg/notameViz/man/save_QC_plots.html" class="external-link">save_QC_plots()</a></code> wrapper to save relevant plots. Herein, the data is visualized side-by-side after normalization for both drift and dilution for conciseness. The are many parameters for customizing the visualizations, but here we stick to the basics of coloring, shaping and ordering the samples to best represent what is of interest.</p>
<p>Typically, drift in the biological samples would be evidenced by an overabundance of low p-values in histograms relating each features’ abundance to injection order. After drift correction, the distributions would be expected to be more uniform for all samples and biological samples. In the case of QC samples, an overabundance of high p-values is expected given that the QC samples are identical.</p>
<p>The QC samples behave as expected, but it seems that the small number of samples does not allow us to fit linear models relating abundance to injection order properly for all samples and biological samples, as evidenced by the overabundance of high p-values in the top histograms.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/notameViz/man/plot_injection_lm.html" class="external-link">plot_injection_lm</a></span><span class="op">(</span><span class="va">se</span>, assay.type <span class="op">=</span> <span class="st">"raw_filled"</span><span class="op">)</span>,</span>
<span>          <span class="fu"><a href="https://rdrr.io/pkg/notameViz/man/plot_injection_lm.html" class="external-link">plot_injection_lm</a></span><span class="op">(</span><span class="va">se</span>, assay.type <span class="op">=</span> <span class="st">"dil_norm"</span><span class="op">)</span>,</span>
<span>          labels <span class="op">=</span> <span class="st">"AUTO"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="notame_normalization_and_feature_selection_files/figure-html/unnamed-chunk-6-1.png" style="width:100.0%"></p>
<figcaption>Figure 1. P-values from linear regression models relating each feature to injection order. The dashed red lines represent the expected uniform distribution. A) Before normalization, featuring all samples, biological samples and QC samples. B) After normalization, featuring all samples, biological samples and QC samples</figcaption></figure>
</div>
</div>
</div>
<p>Sample boxplots are useful for inspecting the distributions of feature abundances. One would expect the QC samples to be more alike after normalization for drift and dilution. Depending on the instrumentation and experimental details, drift may be apparent as a global trend or the effect of drift is cancelled out.</p>
<p>In this case, the distributions of feature abundances are only slightly shifted after normalization. The QC samples appear not to have been affected by PQN. No obvious differences can be seen within study groups or between study groups either. PQN generally decreased the abundances of the biological samples to levels more similar to the QC samples. This comes from a dilution factor of over one for each of the biological samples, reflecting the different source and processing of QC samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/notameViz/man/plot_sample_boxplots.html" class="external-link">plot_sample_boxplots</a></span><span class="op">(</span><span class="va">se</span>, order_by <span class="op">=</span> <span class="st">"Injection_order"</span>,</span>
<span>                               fill_by <span class="op">=</span> <span class="st">"phenotype"</span>,</span>
<span>                               assay.type <span class="op">=</span> <span class="st">"raw_filled"</span><span class="op">)</span>,</span>
<span>          <span class="fu"><a href="https://rdrr.io/pkg/notameViz/man/plot_sample_boxplots.html" class="external-link">plot_sample_boxplots</a></span><span class="op">(</span><span class="va">se</span>, order_by <span class="op">=</span> <span class="st">"Injection_order"</span>,</span>
<span>                               fill_by <span class="op">=</span> <span class="st">"phenotype"</span>, assay.type <span class="op">=</span> <span class="st">"dil_norm"</span><span class="op">)</span>,</span>
<span>          labels <span class="op">=</span> <span class="st">"AUTO"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="notame_normalization_and_feature_selection_files/figure-html/unnamed-chunk-7-1.png" style="width:100.0%"></p>
<figcaption>Figure 2. Boxplots representing the distributions of feature abundances in each sample by injection order, featuring the median as a black line, the interquartile range as a box and the 1.5x the interquartile range as whiskers. A) Before normalization. B) After normalization.</figcaption></figure>
</div>
</div>
</div>
<p>For the remaining visualizations which all involve Euclidean distances, the data was autoscaled. Conservative interpretation focusing on the QC samples is enough to assess the normalization process, but intuitions about the data at large can be formed based on the biological samples.</p>
<p>In distance density plots, one would generally expect a single prominent density peak for QC samples and greater separation between QC samples and biological samples after normalization, which also is the case herein. The separation between QC samples and biological samples has not increased, although in this case it may be an artifact from the sample size affecting the kernel density estimate.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/notameViz/man/plot_dist_density.html" class="external-link">plot_dist_density</a></span><span class="op">(</span><span class="va">se</span>, assay.type <span class="op">=</span> <span class="st">"raw_filled"</span>, title <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span>,</span>
<span>          <span class="fu"><a href="https://rdrr.io/pkg/notameViz/man/plot_dist_density.html" class="external-link">plot_dist_density</a></span><span class="op">(</span><span class="va">se</span>, assay.type <span class="op">=</span> <span class="st">"dil_norm"</span>, title <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span>,</span>
<span>          labels <span class="op">=</span> <span class="st">"AUTO"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="notame_normalization_and_feature_selection_files/figure-html/unnamed-chunk-8-1.png" style="width:100.0%"></p>
<figcaption>Figure 3. Density plot of Euclidean distances between samples. A) Before normalization. B) After normalization.</figcaption></figure>
</div>
</div>
</div>
<p>PCA is useful for visualizing the similarity of individual samples and groups as it represents the data along axes which, consecutively, maximally explains the variation. PCA is especially useful in an exploratory sense, allowing us to relate covariates to the grouping of samples. Typically, the QC samples will group more tightly after normalization and the overall structure is preserved. The variation explained by the first principal components can also be expected to increase as normalization should reduce the masking effect of unwanted variation.</p>
<p>As expected, the overall structure is intact and the QC samples group more tightly, indicating that we have reduced unwanted variation. The CVD group seems to group more tightly after normalization; the opposite can be said for the CTR group. Such local differences could be investigated further using t-SNE and coloring for injection order. There is little difference in the variance explained by the first two components.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/notameViz/man/plot_pca.html" class="external-link">plot_pca</a></span><span class="op">(</span><span class="va">se</span>, color <span class="op">=</span> <span class="st">"Injection_order"</span>,</span>
<span>                   shape <span class="op">=</span> <span class="st">"phenotype"</span>, assay.type <span class="op">=</span> <span class="st">"raw_filled"</span><span class="op">)</span>,</span>
<span>          <span class="fu"><a href="https://rdrr.io/pkg/notameViz/man/plot_pca.html" class="external-link">plot_pca</a></span><span class="op">(</span><span class="va">se</span>, color <span class="op">=</span> <span class="st">"Injection_order"</span>,</span>
<span>                   shape <span class="op">=</span> <span class="st">"phenotype"</span>, assay.type <span class="op">=</span> <span class="st">"dil_norm"</span><span class="op">)</span>,</span>
<span>          labels <span class="op">=</span> <span class="st">"AUTO"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="notame_normalization_and_feature_selection_files/figure-html/unnamed-chunk-9-1.png" style="width:100.0%"></p>
<figcaption>Figure 4. PCA plots of samples, shape by study group and color by injection order. A) Before normalization. B) After normalization.</figcaption></figure>
</div>
</div>
</div>
<p>PCA and hierarchical clustering using Ward’s criterion are complementary unsupervised methods operating in Euclidean space. While PCA focuses on explaining the variation with maximally reduced dimensionality, hierarchical clustering allows observation of relationships between samples at higher resolution by minimizing within-cluster variation in the clustering of samples. Tighter clustering of QC samples is expected. Hierarchical clustering is generally more sensitive to noise than PCA.</p>
<p>In the present case, the QC samples appear more dissimilar after normalization, so we may actually have ended up increasing unwanted variation.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/notameViz/man/plot_dendrogram.html" class="external-link">plot_dendrogram</a></span><span class="op">(</span><span class="va">se</span>, color <span class="op">=</span> <span class="st">"phenotype"</span>, assay.type <span class="op">=</span> <span class="st">"raw_filled"</span>,</span>
<span>                          title <span class="op">=</span> <span class="cn">NULL</span>, subtitle <span class="op">=</span> <span class="st">""</span><span class="op">)</span>,</span>
<span>          <span class="fu"><a href="https://rdrr.io/pkg/notameViz/man/plot_dendrogram.html" class="external-link">plot_dendrogram</a></span><span class="op">(</span><span class="va">se</span>, color <span class="op">=</span> <span class="st">"phenotype"</span>, assay.type <span class="op">=</span> <span class="st">"dil_norm"</span>,</span>
<span>                          title <span class="op">=</span> <span class="cn">NULL</span>, subtitle <span class="op">=</span> <span class="st">""</span><span class="op">)</span>,</span>
<span>          labels <span class="op">=</span> <span class="st">"AUTO"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="notame_normalization_and_feature_selection_files/figure-html/unnamed-chunk-10-1.png" style="width:100.0%"></p>
<figcaption>Figure 5. Dendrograms of hierarchical sample clusters using Ward’s criterion on Euclidean distances between samples. A) Before normalization. B) After normalization.</figcaption></figure>
</div>
</div>
</div>
<p>Heatmaps give insight into the actual distances between samples, where the distances between QC samples should be smaller after normalization. Organizing the heatmap as per hierachical sample clusters using Ward’s criterion facilitates interpretation. The QC samples should appear as a darker, more uniform square.</p>
<p>In unison with the density plots, we can observe less distance between QC samples, although the less uniform QC sample block suggests that all QC samples did not respond equally to the normalization process.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/notameViz/man/plot_sample_heatmap.html" class="external-link">plot_sample_heatmap</a></span><span class="op">(</span><span class="va">se</span>, group <span class="op">=</span> <span class="st">"phenotype"</span>,</span>
<span>                              assay.type <span class="op">=</span> <span class="st">"raw_filled"</span>,</span>
<span>                              title <span class="op">=</span> <span class="cn">NULL</span>, subtitle <span class="op">=</span> <span class="st">""</span><span class="op">)</span>,</span>
<span>          <span class="fu"><a href="https://rdrr.io/pkg/notameViz/man/plot_sample_heatmap.html" class="external-link">plot_sample_heatmap</a></span><span class="op">(</span><span class="va">se</span>, group <span class="op">=</span> <span class="st">"phenotype"</span>, assay.type <span class="op">=</span> <span class="st">"dil_norm"</span>,</span>
<span>                              title <span class="op">=</span> <span class="cn">NULL</span>, subtitle <span class="op">=</span> <span class="st">""</span><span class="op">)</span>,</span>
<span>          labels <span class="op">=</span> <span class="st">"AUTO"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="notame_normalization_and_feature_selection_files/figure-html/unnamed-chunk-11-1.png" style="width:100.0%"></p>
<figcaption>Figure 6. Heatmaps of Euclidean distances between samples, grouped by hierarchical clusters using Ward’s criterion. A) Before normalization. B) After normalization.</figcaption></figure>
</div>
</div>
</div>
</section><section class="section level2"><h2 id="feature-prefiltering">Feature prefiltering<a class="anchor" aria-label="anchor" href="#feature-prefiltering"></a>
</h2>
<p>We flag low-quality features, excluding them from downstream steps. In addition to the D-ratio and detection rate in the trunk of the workflow, we flag features by their relative standard deviation in QC samples. The non- parametric, robust versions of the D-ratio and relative standard deviation are used<span class="citation" data-cites="broadhurst_guidelines_2018">(<a href="#ref-broadhurst_guidelines_2018" role="doc-biblioref">Broadhurst et al. 2018</a>)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://hanhineva-lab.github.io/notame/reference/flag_quality.html" class="external-link">flag_quality</a></span><span class="op">(</span><span class="va">se</span>, assay.type <span class="op">=</span> <span class="st">"dil_norm"</span>,</span>
<span>                   condition <span class="op">=</span> <span class="st">"RSD_r &lt; 0.2 &amp; D_ratio_r &lt; 0.4"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>INFO [2025-12-12 13:23:48]
39% of features flagged for low quality</code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://hanhineva-lab.github.io/notame/reference/flag_report.html" class="external-link">flag_report</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<table style="width:100%;" class="table">
<colgroup>
<col style="width: 12%">
<col style="width: 6%">
<col style="width: 25%">
<col style="width: 21%">
<col style="width: 15%">
<col style="width: 7%">
<col style="width: 10%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;">Split</th>
<th style="text-align: right;">Kept</th>
<th style="text-align: right;">Low_group_detection</th>
<th style="text-align: right;">Low_qc_detection</th>
<th style="text-align: right;">Low_quality</th>
<th style="text-align: right;">Total</th>
<th style="text-align: right;">Flagged</th>
</tr></thead>
<tbody><tr class="odd">
<td style="text-align: left;">HILIC_pos</td>
<td style="text-align: right;">4498</td>
<td style="text-align: right;">716</td>
<td style="text-align: right;">303</td>
<td style="text-align: right;">3551</td>
<td style="text-align: right;">9068</td>
<td style="text-align: right;">4570</td>
</tr></tbody>
</table>
</div>
</div>
<p>Finally, we impute missing values. Using random forest imputation, we can accommodate the possibility of missing values arising not only from the limit of detection, but also because of issues in gap-filling. We drop QC samples so as not to bias the imputation. The strict thresholds for quality metrics and detection rate allow us to be more confident in our imputed values.</p>
<p>The data is now ready for feature selection with differential abundance analysis. It’s handy to save pretreated data in a separate object and continue with a single assay for the remainder of the analysis. We also drop flagged features, although low-quality features can be needed in annotation when searching for specific ions or fragments of known molecules.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2024</span><span class="op">)</span></span>
<span><span class="va">pretreated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://hanhineva-lab.github.io/notame/reference/impute_rf.html" class="external-link">impute_rf</a></span><span class="op">(</span><span class="fu"><a href="https://hanhineva-lab.github.io/notame/reference/drop_qcs.html" class="external-link">drop_qcs</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span>, assay.type <span class="op">=</span> <span class="st">"dil_norm"</span>,</span>
<span>                        name <span class="op">=</span> <span class="st">"imputed"</span><span class="op">)</span></span>
<span><span class="va">base</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://hanhineva-lab.github.io/notame/reference/drop_flagged.html" class="external-link">drop_flagged</a></span><span class="op">(</span><span class="va">pretreated</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">base</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">base</span><span class="op">)</span><span class="op">)</span> <span class="op">!=</span> <span class="st">"imputed"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>INFO [2025-12-12 13:23:48]
Starting random forest imputation at 2025-12-12 13:23:48.177866
INFO [2025-12-12 13:23:50] Out-of-bag error in random forest imputation: 0.006
INFO [2025-12-12 13:23:50] Random forest imputation finished at 2025-12-12 13:23:50.272116 </code></pre>
</div>
</div>
<p>The out-of-bag error is promising. Now the data is ready for feature selection.</p>
</section><section class="section level2"><h2 id="differential-abundance-analysis">Differential abundance analysis<a class="anchor" aria-label="anchor" href="#differential-abundance-analysis"></a>
</h2>
<p>Let’s get a final look at the data with PCA, this time without QC samples and coloring by the only available clinical covariate, age.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/notameViz/man/plot_pca.html" class="external-link">plot_pca</a></span><span class="op">(</span><span class="va">base</span>, color <span class="op">=</span> <span class="st">"age"</span>, shape <span class="op">=</span> <span class="st">"phenotype"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="notame_normalization_and_feature_selection_files/figure-html/unnamed-chunk-14-1.png" style="width:100.0%"></p>
<figcaption>Figure 7. PCA plot of biological samples, shape by study group and color by age.</figcaption></figure>
</div>
</div>
</div>
<p>We’ll use linear models to find interesting features, adjust for false positives from multiple testing using the false discovery rate approach and plot the results in histograms to assess validity of the tests and get a feel for the results. The formula interface is used in most univariate statistics functions in <em>notameStats</em> for flexibility. By default, linear models are fit for each feature but correction for multiple testing is only applied for quality features. Here we also use <code><a href="https://hanhineva-lab.github.io/notame/reference/join_rowData.html" class="external-link">join_rowData()</a></code> to add the results to the object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">base</span>, <span class="st">"log2"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">base</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">lm_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/notameStats/man/perform_lm.html" class="external-link">perform_lm</a></span><span class="op">(</span><span class="va">base</span>, assay.type <span class="op">=</span> <span class="st">"log2"</span>,</span>
<span>                         formula_char <span class="op">=</span> <span class="st">"Feature ~ phenotype + age"</span><span class="op">)</span></span>
<span><span class="va">base</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://hanhineva-lab.github.io/notame/reference/join_rowData.html" class="external-link">join_rowData</a></span><span class="op">(</span><span class="va">base</span>, <span class="va">lm_results</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Feature_ID"</span>, <span class="st">"phenotypeCVD.p.value"</span>,</span>
<span>                                          <span class="st">"phenotypeCVD.p.value_FDR"</span>,</span>
<span>                                          <span class="st">"phenotypeCVD.estimate"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/notameViz/man/plot_p_histogram.html" class="external-link">plot_p_histogram</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">lm_results</span><span class="op">$</span><span class="va">phenotypeCVD.p.value</span><span class="op">)</span><span class="op">)</span>,</span>
<span>          <span class="fu"><a href="https://rdrr.io/pkg/notameViz/man/plot_p_histogram.html" class="external-link">plot_p_histogram</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">lm_results</span><span class="op">$</span><span class="va">phenotypeCVD.p.value_FDR</span><span class="op">)</span><span class="op">)</span>,</span>
<span>          labels <span class="op">=</span> <span class="st">"AUTO"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="notame_normalization_and_feature_selection_files/figure-html/unnamed-chunk-15-1.png" style="width:100.0%"></p>
<figcaption>Figure 8. Linear regression p-value histograms with abundance and age as independent variables. A) p-values. B) FDR-adjusted p-values (q-values).</figcaption></figure>
</div>
</div>
</div>
<p>The p-value histogram for looks promising; it is a relatively uniform distribution with an overabundance of low p-values. However, there are no significant features after correction for multiple testing. For demonstration purposes, we’ll consider the feature “FT0845”, which was identified as caffeine in the trunk of the workflow, most interesting.</p>
<p>After differential abundance analysis or feature selection using supervised learning, for example, the number of significant features or a ranking cutoff allows for manual inspection of feature-wise plots. <em>notameViz</em> includes a variety of feature-wise visualizations adaptable to a variety of study designs. For feature “FT0845”, the abundance of CTR samples is reduced after normalization.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/notameViz/man/save_beeswarm_plots.html" class="external-link">save_beeswarm_plots</a></span><span class="op">(</span><span class="va">pretreated</span><span class="op">[</span><span class="st">"FT0845"</span>, <span class="op">]</span>, x <span class="op">=</span> <span class="st">"phenotype"</span>,</span>
<span>                              color <span class="op">=</span> <span class="st">"phenotype"</span>, save <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                              assay.type <span class="op">=</span> <span class="st">"raw_filled"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>          <span class="fu"><a href="https://rdrr.io/pkg/notameViz/man/save_beeswarm_plots.html" class="external-link">save_beeswarm_plots</a></span><span class="op">(</span><span class="va">base</span><span class="op">[</span><span class="st">"FT0845"</span>, <span class="op">]</span>, x <span class="op">=</span> <span class="st">"phenotype"</span>,</span>
<span>                              color <span class="op">=</span> <span class="st">"phenotype"</span>, save <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                              assay.type <span class="op">=</span> <span class="st">"imputed"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>          labels <span class="op">=</span> <span class="st">"AUTO"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="notame_normalization_and_feature_selection_files/figure-html/unnamed-chunk-16-1.png" style="width:100.0%"></p>
<figcaption>Figure 10. Beeswarm plots for the lowest p-value feature. A) Before normalization. B) After normalization.</figcaption></figure>
</div>
</div>
</div>
<p>The results are also visualized with a variety of comprehensive visualizations. The volcano plot below shows that feature “FT0845” is distinct in having a low p-value and the second-largest fold-change. Manhattan plots and cloud plots could also be used to inspect how interesting features relate to m/z and retention time. We often co-visualize results from differential abundance analysis with a ranking of features from supervised learning for a combined perspective.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/notameViz/man/volcano_plot.html" class="external-link">volcano_plot</a></span><span class="op">(</span><span class="va">base</span>, x <span class="op">=</span> <span class="st">"phenotypeCVD.estimate"</span>,</span>
<span>             p <span class="op">=</span> <span class="st">"phenotypeCVD.p.value"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">base</span><span class="op">)</span><span class="op">[</span><span class="st">"FT0845"</span>, <span class="op">]</span>,</span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="st">"FT0845"</span><span class="op">)</span>,</span>
<span>            vjust <span class="op">=</span> <span class="op">-</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure><p><img src="notame_normalization_and_feature_selection_files/figure-html/unnamed-chunk-17-1.png" style="width:100.0%"></p>
<figcaption>Figure 11. Volcano plot of p-values from limma (negative log10 scale) related to fold-change (log2 scale).</figcaption></figure>
</div>
</div>
</div>
</section><section class="section level2"><h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>Caffeine is a mild diuretic and can concentrate blood in high doses, so a difference in biological dilution between the study groups can be expected. The overall structure of the data seems to be intact after normalization. On the other hand, the QC samples were from a different source so the most probable dilution factors may not have been determined accurately for the biological samples and drift correction likely suffered as well.</p>
</section><section class="section level2"><h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.5.2 (2025-10-31)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 24.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C

time zone: Etc/UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods
[8] base

other attached packages:
 [1] notameStats_1.0.0           notameViz_1.0.0
 [3] notame_1.0.0                ggplot2_4.0.1
 [5] limma_3.66.0                cowplot_1.2.0
 [7] BiocParallel_1.44.0         alabaster.se_1.10.0
 [9] alabaster.base_1.10.0       SummarizedExperiment_1.40.0
[11] Biobase_2.70.0              GenomicRanges_1.62.1
[13] Seqinfo_1.0.0               IRanges_2.44.0
[15] S4Vectors_0.48.0            BiocGenerics_0.56.0
[17] generics_0.1.4              MatrixGenerics_1.22.0
[19] matrixStats_1.5.0           quarto_1.5.1.9002
[21] knitr_1.50

loaded via a namespace (and not attached):
 [1] tidyselect_1.2.1         viridisLite_0.4.2        vipor_0.4.7
 [4] dplyr_1.1.4              farver_2.1.2             S7_0.2.1
 [7] fastmap_1.2.0            digest_0.6.39            lifecycle_1.0.4
[10] alabaster.matrix_1.10.0  statmod_1.5.1            processx_3.8.6
[13] magrittr_2.0.4           compiler_4.5.2           rngtools_1.5.2
[16] rlang_1.1.6              tools_4.5.2              yaml_2.3.12
[19] lambda.r_1.2.4           doRNG_1.8.6.2            S4Arrays_1.10.1
[22] labeling_0.4.3           DelayedArray_0.36.0      RColorBrewer_1.1-3
[25] abind_1.4-8              HDF5Array_1.38.0         withr_3.0.2
[28] purrr_1.2.0              itertools_0.1-3          grid_4.5.2
[31] Rhdf5lib_1.32.0          iterators_1.0.14         scales_1.4.0
[34] MASS_7.3-65              cli_3.6.5                rmarkdown_2.30
[37] rstudioapi_0.17.1        ggbeeswarm_0.7.3         rhdf5_2.54.1
[40] stringr_1.6.0            parallel_4.5.2           formatR_1.14
[43] XVector_0.50.0           alabaster.schemas_1.10.0 vctrs_0.6.5
[46] Matrix_1.7-4             jsonlite_2.0.0           beeswarm_0.4.0
[49] alabaster.ranges_1.10.0  h5mread_1.2.1            foreach_1.5.2
[52] tidyr_1.3.1              ggdendro_0.2.0           missForest_1.6.1
[55] glue_1.8.0               codetools_0.2-20         ps_1.9.1
[58] stringi_1.8.7            gtable_0.3.6             futile.logger_1.4.3
[61] later_1.4.4              tibble_3.3.0             pillar_1.11.1
[64] pcaMethods_2.2.0         htmltools_0.5.9          rhdf5filters_1.22.0
[67] randomForest_4.7-1.2     R6_2.6.1                 Rdpack_2.6.4
[70] evaluate_1.0.5           lattice_0.22-7           rbibutils_2.4
[73] futile.options_1.0.1     Rcpp_1.1.0               SparseArray_1.10.6
[76] ranger_0.17.0            xfun_0.54                pkgconfig_2.0.3         </code></pre>
</div>
</div>
</section><section class="section level2"><h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-broadhurst_guidelines_2018" class="csl-entry" role="listitem">
Broadhurst, David, Royston Goodacre, Stacey N Reinke, Julia Kuligowski, Ian D Wilson, Matthew R Lewis, and Warwick B Dunn. 2018. <span>“Guidelines and Considerations for the Use of System Suitability and Quality Control Samples in Mass Spectrometry Assays Applied in Untargeted Clinical Metabolomic Studies.”</span> <em>Metabolomics</em> 14: 1–17.
</div>
<div id="ref-dieterle_probabilistic_2006" class="csl-entry" role="listitem">
Dieterle, Frank, Alfred Ross, Götz Schlotterbeck, and Hans Senn. 2006. <span>“Probabilistic Quotient Normalization as Robust Method to Account for Dilution of Complex Biological Mixtures. Application in 1H NMR Metabonomics.”</span> <em>Analytical Chemistry</em> 78 (13): 4281–90.
</div>
<div id="ref-kirwan_characterising_2013" class="csl-entry" role="listitem">
Kirwan, JA, DI Broadhurst, RL Davidson, and MR Viant. 2013. <span>“Characterising and Correcting Batch Variation in an Automated Direct Infusion Mass Spectrometry (DIMS) Metabolomics Workflow.”</span> <em>Analytical and Bioanalytical Chemistry</em> 405: 5147–57.
</div>
<div id="ref-klavus_notame_2020" class="csl-entry" role="listitem">
Klåvus, Anton, Marietta Kokla, Stefania Noerman, Ville M Koistinen, Marjo Tuomainen, Iman Zarei, Topi Meuronen, et al. 2020. <span>“<span>‘Notame’</span>: Workflow for Non-Targeted LC–MS Metabolic Profiling.”</span> <em>Metabolites</em> 10 (4): 135.
</div>
</div>
</section><script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'light-border',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config);
    }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script></main>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Philippine Louail, Johannes Rainer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.9000.</p>
</div>

    </footer>
</div>





  </body>
</html>
